---
title: "scDART-seq_vignette"
output: 
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
abstract: |
      This is the vignette to show the analysis on the scDART-seq data of SMART-seq2.You can not direct run the code. This is just provide for reading the process of the coding step by step.\n <center> **Table of content** </center>
vignette: >
  % \VignetteIndexEntry{scDART-seq_vignette}
  % \VignetteEngine{knitr::rmarkdown}
  % \VignetteEncoding{UTF-8}
  % \VignetteDepends{pROC} 
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
#library(SigRM)
```


# the data and code can be access from:
https://doi.org/10.6084/m9.figshare.25815862

# Extraction of m6A signals from scDART-seq data 
The raw data was directly downloaded from GEO and aligned to hg38 reference genome using hisat2. The candidate m6A sites (corresponding to mutated C within the DRACH motifs) were detected using Samtools (default setting). The m6A and non-modification signals (number of mutated and unmutated sequences with respect to each candidate sites) were counted using GenomicAlignments R package. The Information of a total of 2371331 sites was extracted for 1382 cells, including 991 testing cells and 391 negative control which has mutated YTH domain. 



```{r,eval=FALSE}
###########################################################
####                         scDART_seq                ####
###########################################################
# data download
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE180954 

# alignment
(nohup trim_galore --paired -o /home/share/bowen/scDART/data/YTHcell_',i,' /home/share/bowen/scDART/data/YTHcell_',i,'/',sra,'_1.fastq',' /home/share/bowen/scDART/data/YTHcell_',i,'/',sra,'_2.fastq)&
  
(nohup STAR --runThreadN 10 --runMode genomeGenerate --genomeDir /home/share/bowen/genome/hg38/star_indx/ --genomeFastaFiles /home/share/bowen/genome/hg38/Homo_sapiens.GRCh38.dna.primary_assembly.fa --sjdbGTFfile /home/share/bowen/genome/hg38/Homo_sapiens.GRCh38.109.gtf --sjdbOverhang 150 > /home/share/bowen/genome/hg38/star_indx/build_index.out)&   

(nohup STAR --runThreadN 10 --outSAMtype BAM Unsorted --readFilesIn /home/share/bowen/scDART/data/YTHcell_1/SRR15268889_1_val_1.fq /home/share/bowen/scDART/data/YTHcell_1/SRR15268889_2_val_2.fq --genomeDir /home/share/bowen/genome/hg38/star_indx --sjdbGTFfile /home/share/bowen/genome/hg38/Homo_sapiens.GRCh38.109.gtf --outFileNamePrefix /home/share/bowen/scDART/data/YTHcell_1/bam/SRR15268889 --outSAMattributes All --outSAMunmapped Within --sjdbOverhang 150 > /home/share/bowen/scDART/data/YTHcell_1/bam/SRR15268889.log)& 
  
(nohup STAR --runThreadN 10 --outSAMtype BAM Unsorted --readFilesIn /home/share/bowen/scDART/data/YTHmutcell_1/SRR15268462_1_val_1.fq /home/share/bowen/scDART/data/YTHmutcell_1/SRR15268462_2_val_2.fq --genomeDir /home/share/bowen/genome/hg38/star_indx --sjdbGTFfile /home/share/bowen/genome/hg38/Homo_sapiens.GRCh38.109.gtf --outFileNamePrefix /home/share/bowen/scDART/data/YTHmutcell_1/bam/SRR15268462 --outSAMattributes All --outSAMunmapped Within --sjdbOverhang 150 > /home/share/bowen/scDART/data/YTHmutcell_1/bam/SRR15268462.log)&  

# variant calling
samtools sort -o /home/share/bowen/scDART/data/YTHcell_1/bam/SRR15268889_sorted.bam /home/share/bowen/scDART/data/YTHcell_1/bam/SRR15268889Aligned.out.bam

bcftools mpileup -f /home/share/bowen/genome/hg38/Homo_sapiens.GRCh38.dna.primary_assembly.fa /home/share/bowen/scDART/data/YTHcell_1/bam/SRR15268889_sorted.bam | bcftools call -mv -Oz -o /home/share/bowen/scDART/data/YTHcell_1/bam/SRR15268889_calls.vcf 

# find relevant mutation
data <- read.vcfR(SRR15268889_calls.vcf)

data2 <- as.data.frame(data@fix)

indx <- which(data2$REF == 'A' | data2$REF == 'G' | data2$REF == 'C' | data2$REF == 'T')

data2 <- data2[indx,]

indx <- which(data2$ALT == 'A' | data2$ALT == 'G' | data2$ALT == 'C' | data2$ALT == 'T')

data2 <- data2[indx,]

gr <- GRanges(seqnames = paste0('chr',data2$CHROM),
              IRanges(start = as.numeric(data2$POS),
                      width = 1),
              strand = '+',
              ref = data2$REF,
              alt = data2$ALT,
              quality = as.numeric(data2$QUAL),
              info = data2$INFO)

gr <- keepStandardChromosomes(gr,pruning.mode = "coarse")

ref <- as.character(DNAStringSet(Views(Hsapiens,gr)))

indx <- which(gr$ref != ref)

if(length(indx) != 0){
  gr <- gr[-indx]
}
  
saveRDS(gr,'/home/share/bowen/scDART/data/YTHcell_1/bam/SNP.rds')  
  
# filter mutation
transcript <- transcripts(TxDb.Hsapiens.UCSC.hg38.knownGene)

transcript_plus <- transcript[which(strand(transcript)== "+")]

transcript_minus <- transcript[which(strand(transcript)== "-")]

gr <- readRDS('/home/share/bowen/scDART/data/YTHcell_1/bam/SNP.rds')

gr_all <- GRanges()

# 1.	Coordinate在 (+) strand: ACX
gr_plus <- gr
olp <- findOverlaps(gr_plus,transcript_plus)
q <- queryHits(olp)
s <- subjectHits(olp)
gr_plus <- gr_plus[q]
gr_plus$transcript_overlapped <- transcript_plus[s]$tx_name
gr_plus$transcript_strand <- '+'
gr_plus <- unique(gr_plus)

gr_plus$seq <- as.character(DNAStringSet(Views(Hsapiens,gr_plus+1)))
indx <- which(gr_plus$seq == 'ACA'|gr_plus$seq == 'ACG'|gr_plus$seq == 'ACC'|gr_plus$seq == 'ACT')
gr_plus <- gr_plus[indx]
indx <- which(gr_plus$alt == 'T')
gr_plus <- gr_plus[indx]
gr_plus$sample <- list3[i]

# 2.	Coordinate在 (-) strand: XGT
gr_minus <- gr
strand(gr_minus) <- '-'

olp <- findOverlaps(gr_minus,transcript_minus)
q <- queryHits(olp)
s <- subjectHits(olp)
gr_minus <- gr_minus[q]
gr_minus$transcript_overlapped <- transcript_minus[s]$tx_name
gr_minus$transcript_strand <- '-'
gr_minus <- unique(gr_minus)

strand(gr_minus) <- '+'
gr_minus$seq <- as.character(DNAStringSet(Views(Hsapiens,gr_minus+1)))
indx <- which(gr_minus$seq == 'AGT'|gr_minus$seq == 'CGT'|gr_minus$seq == 'GGT'|gr_minus$seq == 'TGT')
gr_minus <- gr_minus[indx]
indx <- which(gr_minus$alt == 'A')
gr_minus <- gr_minus[indx]
gr_minus$sample <- list3[i]

gr <- c(gr_plus,gr_minus)
saveRDS(gr,paste0(list2[i],'SNP_filtered.rds'))
gr_all <- unique(c(gr_all,gr))
saveRDS(gr_all,'/home/share/bowen/scDART/testSNP_all.rds')

## coverage
bamfilePath <- '/home/share/bowen/scDART/data/YTHcell_1/bam/SRR15268889_sorted.bam'

flag <- scanBamFlag(isNotPassingQualityControls=FALSE,
                    isDuplicate=FALSE)

param <- ScanBamParam(flag=flag, what=c("seq","qual"),which=gr_all)

alns <- readGAlignments(bamfilePath, param=param)

alns <- keepStandardChromosomes(alns,pruning.mode = "coarse")

qseq <- mcols(alns)$seq

qual <- mcols(alns)$qual

seqlevels(alns) <- seqlevels(gr_all)

nucl_piles <- pileLettersAt(qseq, seqnames(alns), start(alns),
                            cigar(alns), gr_all)

frequency <- alphabetFrequency(nucl_piles, baseOnly=TRUE)

# gene expression
stringtie /home/share/bowen/scDART/data/YTHcell_1/bam/SRR15268889_sorted.bam -G /home/share/bowen/genome/hg38/Homo_sapiens.GRCh38.109.gtf -e -A /home/share/bowen/scDART/data/YTHcell_1/bam/gene_abund.tab


```




There are three files that can be obtained:

* gene_expression: a list with Gene.ID, Gene.Name, Reference, Strand, Start, End, Coverage, FPKM, and TPM.
* frequency_all: a list with the frequency of A, C, T, G, and others for 2,371,331 sites.
* SNP_all: a Grange object with detailed information for the 2,371,331 sites.


# Pre-process


We need to calculate the methylation read counts and unmethylated read counts base the frequency_all and SNP_all file
## SMART-seq

### Import the data

There are two data:

* frequency_all is the data contains the read counts of the A C G T and others for each sites and cell (row means sites, column means cell)
* testSNP_all is the data contains the information of the SNP for each sites and cells
```{r,eval=FALSE}


frequency_all <- readRDS("/home/share/haozhe/SIGMR_single_cell/data/SMART/origin/frequency_all.rds")


testSNP_all <- readRDS("/home/share/haozhe/SIGMR_single_cell/data/SMART/origin/SNP_all.rds")
```



We need to consolidata the two data for the following research.

### Origin data

Firstly, see the data. 

* All the data in the frequency_all are useful.
* testSNP_all data have the transcript strand and seq are useful.
```{r,eval=FALSE}
head(frequency_all[[1]])
testSNP_all
```

### Combine the frequency_all and testSNP_all
In the following, I put the information of the strand and the seq information from testSNP_all data into the frequency_all data. Base on that, I calculate the number of methylation reads, unmethylation reads, total reads, error rate. I also put this information into the frequency_all data and delete the not important information.



```{r,eval=FALSE}

library(dplyr)
# get the number of cell number of sites

num_cell <- length(frequency_all)
num_sites <- length(testSNP_all)

# get the index of the + - STAND 

strand <- testSNP_all$transcript_strand
strand_plus <- which(strand=="+")
strand_minus <- which(strand=="-")

error_rate_list <- list()

error_reads_list <- list()
methylation_rate <- list()

site_length_nomatch <-  which(lapply(frequency_all, function(x) dim(x)[1])!=length(testSNP_all))


for (i in setdiff((1:num_cell),site_length_nomatch)){
  
  # change to data frame and add strand column
  
  frequency_all[[i]] <-  data.frame(frequency_all[[i]])
  
  # get methylated reads 
  
  methylated_reads <- rep(0, num_sites)
  methylated_reads[strand_plus] <- frequency_all[[i]][strand_plus,4]
  methylated_reads[strand_minus] <-frequency_all[[i]][strand_minus,1]
  
  # get unmethylated reads 
  
  unmethylated_reads <- rep(0, num_sites)
  unmethylated_reads[strand_plus] <- frequency_all[[i]][strand_plus,2]
  unmethylated_reads[strand_minus] <- frequency_all[[i]][strand_minus,3]
  
  
 
  # get error reads 
  
  error_reads <- rep(0, num_sites)
  error_reads[strand_plus] <- frequency_all[[i]][strand_plus,1]+frequency_all[[i]][strand_plus,3]+
                              frequency_all[[i]][strand_plus,5]
  error_reads[strand_minus] <- frequency_all[[i]][strand_minus,2]+frequency_all[[i]][strand_minus,4]+
                              frequency_all[[i]][strand_minus,5]
  
  
  
    frequency_all[[i]] <- frequency_all[[i]] %>% 
    mutate(methylated_reads=methylated_reads)%>% 
    mutate(unmethylated_reads=unmethylated_reads)%>%
    mutate(error_reads=error_reads)%>% 
     select(methylated_reads,unmethylated_reads,error_reads)
 
  error_reads_list[[i]] <-  frequency_all[[i]]$error_reads
  
  frequency_all[[i]] <- frequency_all[[i]] %>%
  mutate(methylated_reads=methylated_reads)%>%
   mutate(unmethylated_reads=unmethylated_reads)%>%
   mutate(error_rate=error_reads/(methylated_reads+unmethylated_reads+error_reads))%>%
    select(methylated_reads,unmethylated_reads,error_rate)

  error_rate_list[[i]] <-  frequency_all[[i]]$error_rate
  frequency_all[[i]] <- frequency_all[[i]] %>% select(methylated_reads,unmethylated_reads)
  
  methylation_rate[[i]] <- frequency_all[[i]]$methylated_reads/(frequency_all[[i]]$methylated_reads+frequency_all[[i]]$unmethylated_reads)
}

if (length(site_length_nomatch)==0){
}else{
  
  frequency_all <- frequency_all[-site_length_nomatch]
  error_reads_list <- error_reads_list[-site_length_nomatch]
  error_rate_list <- error_rate_list[-site_length_nomatch]


}


setwd("/home/share/haozhe/SIGMR_single_cell/data/SMART/preprocess1")
saveRDS(error_rate_list,"error_rate_list_73_treatment.rds")
saveRDS(frequency_all,"reads_counts_73_treatment.rds")
saveRDS(error_reads_list,"error_reads_list_73_treatment.rds")
saveRDS(methylation_rate,"methylation_rate.rds")


```




### visulisation the methylation rate
```{r,eval=FALSE}
library(ggplot2)
methylation_rate <- readRDS("~/SIGMR_single_cell/data/SMART/preprocess2/methylation_rate.rds")

# the discard cell is the cell do not match with the expression file or cell with error
discard_cell <- readRDS("~/SIGMR_single_cell/data/SMART/preprocess2/discard_cell.rds")
methylation_rate <- methylation_rate[-discard_cell]
num_controlcell <- length(which(grepl("YTHmutcell",names(methylation_rate))))
num_testcell <- length(which(grepl("YTHcell",names(methylation_rate))))


set.seed(10)
index_control <- sample(which(grepl("YTHmutcell",names(methylation_rate))),3)
index_test <- sample(which(grepl("YTHmutcell",names(methylation_rate))),3)
# see the density of the methylation_rate from control cell and test cell


# see the histogram of the number of the sites methylation rate >=0.5
methylation_rate_high_num <- as.numeric(lapply(methylation_rate, function(x) length(which(x>=0.3))))

data_methylation_rate_high_num <- data.frame(cbind(methylation_rate_high_num,c(rep("control",num_controlcell),rep("test",num_testcell))))

colnames(data_methylation_rate_high_num) <- c("num","group")
data_methylation_rate_high_num$num <- as.numeric(data_methylation_rate_high_num$num)

ggplot() +
  geom_histogram(data=data_methylation_rate_high_num[1:num_controlcell,],alpha=0.8, aes(x=num,y=after_stat(density)),bins=100, fill="#69b3a2")+
    geom_label(aes(x=20000, y=3e-4, label="control"), color="#69b3a2") +
  # Bottom
geom_histogram(data=data_methylation_rate_high_num[-(1:num_controlcell),], aes(x = num, y = -after_stat(density)),bins=100, fill= "#404080")+
    geom_label(aes(x=20000, y=-1.5e-4, label="test"), color="#404080")+
coord_cartesian(xlim = c(0,22000))

```












```{r,eval=FALSE}

error_reads_list_1 <- readRDS("/home/share/haozhe/SIGMR_single_cell/data/SMART/preprocess1/error_reads_list_73_control.rds")

error_reads_list_2 <- readRDS("/home/share/haozhe/SIGMR_single_cell/data/SMART/preprocess1/error_reads_list_73_treatment.rds")


error_reads_list_3 <- readRDS("/home/share/haozhe/SIGMR_single_cell/data/SMART/preprocess1/error_reads_list_74_treatment.rds")

error_reads_list <- c(error_reads_list_1,error_reads_list_2,error_reads_list_3)
rm(error_reads_list_1,error_reads_list_2,error_reads_list_3)

```






### Discard Sites and final output

#### find the sites
```{r,eval=FALSE}

discard_cell <- readRDS("~/SIGMR_single_cell/data/SMART/preprocess2/discard_cell.rds")
error_reads_list <- error_reads_list[-discard_cell]

error_reads_data <- as.data.frame(error_reads_list)
error_reads_data_sum <- apply(error_reads_data,1,sum)

frequency_all <- readRDS("~/SIGMR_single_cell/data/SMART/preprocess2/frequency_all.rds")


frequency_all_sum <- lapply(frequency_all,function(x) rowSums(x))

frequency_all_median <- apply(as.data.frame(frequency_all_sum),1,median)
frequency_all_sum <- apply(as.data.frame(frequency_all_sum),1,sum)


error_rate <- error_reads_data_sum/(error_reads_data_sum+frequency_all_sum)



#length(union(which(error_rate>0.001),which(frequency_all_average<=1)))
#length(which(frequency_all_average<=10))

rm(list=ls()[! ls() %in% c("error_rate","frequency_all_median","methylation_rate")])
gc()
union_set <- union(union(which(error_rate>=0.01),which(is.na(error_rate))),which(frequency_all_median<=10))


methylation_rate_processed <- lapply(methylation_rate, function(df) df[-union_set])

# see the histogram of the number of the sites methylation rate >=0.5
methylation_rate_high_num_processed <- as.numeric(lapply(methylation_rate_processed, function(x) length(which(x>=0.3))))

num_controlcell <- length(which(grepl("YTHmutcell",names(methylation_rate))))
num_testcell <- length(which(grepl("YTHcell",names(methylation_rate))))

data_methylation_rate_high_num_processed <-  data.frame(cbind(methylation_rate_high_num_processed,c(rep("control",num_controlcell),rep("test",num_testcell))))

colnames(data_methylation_rate_high_num_processed) <- c("num","group")
data_methylation_rate_high_num_processed$num <- as.numeric(data_methylation_rate_high_num_processed$num)

ggplot() +
  geom_histogram(data=data_methylation_rate_high_num_processed[1:num_controlcell,],alpha=0.8, aes(x=num,y=after_stat(density)),bins=100, fill="#69b3a2")+
    geom_label(aes(x=300, y=0.0002, label="control"), color="#69b3a2") +
  # Bottom
geom_histogram(data=data_methylation_rate_high_num_processed[-(1:num_controlcell),], aes(x = num, y = -after_stat(density)),bins=100, fill= "#404080")+
    geom_label(aes(x=300, y=-0.0001, label="test"), color="#404080")




setwd("/home/share/haozhe/SIGMR_single_cell/data/SMART/preprocess2")
saveRDS(union_set,"discard_sites.rds")

discard_cell <- readRDS("~/SIGMR_single_cell/data/SMART/preprocess2/discard_cell.rds")
methylation_rate_processed <- methylation_rate_processed[-discard_cell]

setwd("/home/share/haozhe/SIGMR_single_cell/data/SMART/preprocess2")
saveRDS(methylation_rate_processed,"methylation_rate_processed.rds")


```












#### discard them get final pre-processed data
```{r,eval=FALSE}

rm(list=ls()[! ls() %in% c("union_set")])
frequency_all <- readRDS("~/SIGMR_single_cell/data/SMART/preprocess2/frequency_all.rds")
frequency_all_processed <- lapply(frequency_all, function(df) df[-union_set,])

discard_cell <- readRDS("~/SIGMR_single_cell/data/SMART/preprocess2/discard_cell.rds")
frequency_all_processed <- frequency_all_processed[-discard_cell]

setwd("/home/share/haozhe/SIGMR_single_cell/data/SMART/preprocess2")
saveRDS(frequency_all_processed,"frequency_all_processed.rds")
```





Following data extraction, 510,554 candidate m6A sites were retained for further analysis.

The files associated with this study include:

*SNP File (SNP_all.rds):

Data Details: Stored in RDS format, containing a GRanges object.
Content: Each entry in the GRanges object represents a specific genomic region corresponding to an m6A modification site detected through scDART-seq.
seqnames: Factor Rle object containing chromosome or genomic sequence names.
ranges: IRanges object containing genomic intervals (start and end positions).
strand: Factor Rle object indicating the strand (directionality) of the genomic region.
mcols: DataFrame object containing optional metadata columns, such as quality scores, coverage depth, and mutation data.
seqinfo: Seqinfo object providing information about the genomic sequences present in the GRanges object.
Purpose: Provides detailed genomic information about identified m6A modification sites, facilitating further analysis of their distribution, characteristics, and genomic context in HEK293T cells.


*Frequency File (frequency_all.rds):

Data Details: Also stored in RDS format, consisting of a list.
Content: Each item in the list represents a single-cell, containing counts of methylated and unmethylated reads for corresponding m6A modification sites detected in scDART-seq data.
Purpose: Offers quantitative data on the abundance of methylated and unmethylated sequences at each m6A site across individual cells, enabling investigation of m6A modification patterns at a single-cell level.


*Expression TPM File (expression_TPM.rds):

Data Details: Stored as an RDS file, comprising a list.
Content: Each item in the list represents a single-cell, with corresponding TPM values for gene expression.
Purpose: Provides information on gene expression levels across individual cells, facilitating examination of potential correlations between m6A modification patterns and gene expression profiles in scDART-seq data from HEK293T cells.

*Gene Information File (gene_informations.rds):

Data Details: Stored as an RDS file, comprising a data frame.
Content: Includes information such as Gene ID, Gene Name, Reference, Strand, Start position, End position, and Coverage.
Purpose: Offers additional details about gene expression data, aiding in the interpretation and analysis of gene expression profiles in conjunction with m6A modification patterns.




All three files can access on: https://doi.org/10.6084/m9.figshare.25815862

with DOI: 10.6084/m9.figshare.25815862





# compare the performance for different model

## random chose the control group


## 2

### preprocess
```{r,eval=FALSE}
frequency_all_processed <- readRDS("~/data/SMART/preprocess2/frequency_all_processed.rds")
# get the meth reads and unmeth reads
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")

# get number of control cell and test cell
num_testcell <- length(which(grepl("YTHcell",names(frequency_all_processed))))
num_controlcell <- length(which(grepl("YTHmutcell",names(frequency_all_processed))))


rm(frequency_all_processed)

set.seed(1)
# we random choose 10 control data as the fake test data

test_fake <- sample(1:num_controlcell,10)
control_index <- setdiff(1:num_controlcell,test_fake)

# also random choose 10 real test data, combine them
test_index <- c( sample((num_controlcell+1):(num_controlcell+num_testcell),10), test_fake)


# select random cells
num_cell_control <- 2
control_index <- sample(control_index,num_cell_control)

path <-  paste0("~/data/SMART/results_ks/res",num_cell_control)
setwd(path )
saveRDS(control_index,"control_index.rds")

setwd(path)
saveRDS(test_index,"test_index.rds")

meth_control <- meth[,control_index]
meth_test <- meth[,test_index]

unmeth_control <- unmeth[,control_index]
unmeth_test <- unmeth[,test_index]

rm(meth,unmeth)
```

### put into different model

#### scDART
```{r,eval=FALSE}


start_time <- Sys.time()

# put into the scDART model
single_test_scDART <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  # combine as the total reads
  data1 <- meth_control+unmeth_control
  data2 <- meth_test+unmeth_test
  
  # calculate the mean p and the variance of the control data
  
  p_control <- rowSums(meth_control)/rowSums(data1)
  
  # find the site satisfies the three conditions
  if (length(dim(meth_test)[2])==0){
     # first condition
      index_1 <- which(data2>=20)
      
      # second condition
      p_test <- meth_test/data2
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1] 
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test>=2)
      index <- index_4 [index_4 %in% index_3]
  }else{
    index <- list()
    for (i in 1 : dim(meth_test)[2]){
      
      # first condition
      index_1 <- which(data2[,i]>=20)
      
      # second condition
      p_test <- meth_test[,i]/data2[,i]
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1]
      
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test[,i]>=2)
      index[[i]] <- index_4 [index_4 %in% index_3]
    }
    
    
    
  }
  
  
  return(index)
  
}

res_scDART <-  single_test_scDART(meth_control,meth_test,unmeth_control,unmeth_test)

res_scDART_ <- list()
num_sites <- dim(meth_control)[1]

for (i in 1:length(res_scDART)){
  
  
  # res_scDART
  
  res_scDART_[[i]] <- rep(0,num_sites)
  res_scDART_[[i]][res_scDART[[i]]] <- 1


    
  
}




end_time <- Sys.time()
run_time_scDART <- end_time-start_time

setwd(path )
saveRDS(res_scDART_,"res_scDART_.rds")

rm(res_scDART,res_scDART_)
```


#### DESeq2
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()

library(DESeq2)
DESeq2_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  meth_control <- data.frame(meth_control)
  meth_test <- data.frame(meth_test)
  unmeth_control <- data.frame(unmeth_control)
  unmeth_test <- data.frame(unmeth_test)

  
  
  
  sizeFactor <- function(data) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  # first log
  log_data <- log(data)
  log_data[is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}

  num <- dim(unmeth_control)[2]
  res <- list()
  for (i in 1:dim(meth_test)[2]){
    counts <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
    design = data.frame(
                       Trt = rep(c(rep("control", num),"test"),2), 
                    Meth = c(rep("meth",(num+1)),rep("unmeth", 
                            (num+1))))
    model = ~  Meth + Trt + Meth:Trt
    # 
    dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = design,
                                design = model)
    sf = sizeFactor(cbind((meth_control[,1:num]+unmeth_control[,1:num]),
                       (meth_test[,i]+unmeth_test[,i])))
    
    sizeFactors(dds) = rep(sf,2)
    dds <- DESeq(dds, test = "Wald")
    
    res[[i]] = DESeq2::results(dds, name = "Methunmeth.Trttest",cooksCutoff=FALSE) 
  }
    

 
  
  return(res)
}


res_DESeq2 <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)


end_time <- Sys.time()
run_time_DESeq2 <- end_time-start_time

setwd(path )
saveRDS(res_DESeq2,"res_DESeq2.rds")

rm(res_DESeq2)
```

#### edgeR
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()
library("edgeR")
edgeR_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  num <- dim(meth_control)[2]
  res <- list()
  
  for (i in 1: dim(meth_test)[2]){
       # create deglist
    countdata <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
   
    dgelist <- DGEList(counts = countdata)
    
    
    countdata_all <- cbind((meth_control+unmeth_control),(meth_test[,i]+unmeth_test[,i]))
   
    dgelist_all <- DGEList(counts = countdata_all)
    
    
    # normalize
    dgelist_norm_all <- calcNormFactors(dgelist_all, method = 'TMM')
    
    dgelist_norm <- calcNormFactors(dgelist, method = 'TMM')
    
    dgelist_norm@.Data[[2]] <- rbind(dgelist_norm_all@.Data[[2]],dgelist_norm_all@.Data[[2]])
    
    rownames(dgelist_norm@.Data[[2]]) <- paste0("Sample",1:(2*(num+1)))
    
    
    group1 <- factor(rep(c(rep("control",num),"test"),2))
    group2 <- factor(c(rep("meth",(num+1)),rep("unmeth",(num+1))))
    
    
    
    design <- model.matrix(~group1+group2+group1:group2)
    colnames(design) <- c("control", "test", "unmeth", "unmeth_and_test")
    
    dge <- estimateDisp(dgelist_norm, design, robust = TRUE)
    
    fit <- glmFit(dge, design, robust = TRUE)
    

    lrt <- glmLRT(fit,coef=4)@.Data
    
    res[[i]] <- lrt[[14]]
  }
  
  return(res)
 
}


res_edgeR <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)


end_time <- Sys.time()
run_time_edgeR <- end_time-start_time


setwd(path)
saveRDS(res_edgeR,"res_edgeR.rds")
rm(res_edgeR)

```


#### SigRM

##### SigRM_locfit
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_locfit <-  SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test,method_dispersion = "locfit")
end_time <- Sys.time()
run_time_SigRM_locfit <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_locfit,"res_SigRM_locfit.rds")
rm(res_SigRM_locfit)

```

##### SigRM_unbiased 
```{r,eval=FALSE}
start_time <- Sys.time()
res_SigRM_unbiased <-  SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test)
end_time <- Sys.time()
run_time_SigRM_unbiased <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_unbiased,"res_SigRM_unbiased.rds")
rm(res_SigRM_unbiased)

```

##### SigRM_mle
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_mle <-  SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test,method_dispersion = "mle")
end_time <- Sys.time()
run_time_SigRM_mle <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_mle,"res_SigRM_mle.rds")
rm(res_SigRM_mle)

```


### runtime 
```{r,eval=FALSE}
setwd(path)
runtime <- data.frame(run_time_scDART, run_time_DESeq2, run_time_edgeR, run_time_SigRM_locfit, run_time_SigRM_unbiased, run_time_SigRM_mle)
save(runtime,file = "runtime.RData")
```



## 4

### preprocess
```{r,eval=FALSE}
frequency_all_processed <- readRDS("~/data/SMART/preprocess2/frequency_all_processed.rds")
# get the meth reads and unmeth reads
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")

# get number of control cell and test cell
num_testcell <- length(which(grepl("YTHcell",names(frequency_all_processed))))
num_controlcell <- length(which(grepl("YTHmutcell",names(frequency_all_processed))))


rm(frequency_all_processed)

set.seed(1)
# we random choose 10 control data as the fake test data

test_fake <- sample(1:num_controlcell,10)
control_index <- setdiff(1:num_controlcell,test_fake)

# also random choose 10 real test data, combine them
test_index <- c( sample((num_controlcell+1):(num_controlcell+num_testcell),10), test_fake)


# select random cells
num_cell_control <- 4
control_index <- sample(control_index,num_cell_control)

path <-  paste0("~/data/SMART/results_ks/res",num_cell_control)
setwd(path )
saveRDS(control_index,"control_index.rds")

setwd(path)
saveRDS(test_index,"test_index.rds")

meth_control <- meth[,control_index]
meth_test <- meth[,test_index]

unmeth_control <- unmeth[,control_index]
unmeth_test <- unmeth[,test_index]

rm(meth,unmeth)
```

### put into different model

#### scDART
```{r,eval=FALSE}


start_time <- Sys.time()

# put into the scDART model
single_test_scDART <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  # combine as the total reads
  data1 <- meth_control+unmeth_control
  data2 <- meth_test+unmeth_test
  
  # calculate the mean p and the variance of the control data
  
  p_control <- rowSums(meth_control)/rowSums(data1)
  
  # find the site satisfies the three conditions
  if (length(dim(meth_test)[2])==0){
     # first condition
      index_1 <- which(data2>=20)
      
      # second condition
      p_test <- meth_test/data2
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1] 
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test>=2)
      index <- index_4 [index_4 %in% index_3]
  }else{
    index <- list()
    for (i in 1 : dim(meth_test)[2]){
      
      # first condition
      index_1 <- which(data2[,i]>=20)
      
      # second condition
      p_test <- meth_test[,i]/data2[,i]
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1]
      
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test[,i]>=2)
      index[[i]] <- index_4 [index_4 %in% index_3]
    }
    
    
    
  }
  
  
  return(index)
  
}

res_scDART <-  single_test_scDART(meth_control,meth_test,unmeth_control,unmeth_test)

res_scDART_ <- list()
num_sites <- dim(meth_control)[1]

for (i in 1:length(res_scDART)){
  
  
  # res_scDART
  
  res_scDART_[[i]] <- rep(0,num_sites)
  res_scDART_[[i]][res_scDART[[i]]] <- 1


    
  
}




end_time <- Sys.time()
run_time_scDART <- end_time-start_time

setwd(path )
saveRDS(res_scDART_,"res_scDART_.rds")

rm(res_scDART,res_scDART_)
```


#### DESeq2
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()

library(DESeq2)
DESeq2_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  meth_control <- data.frame(meth_control)
  meth_test <- data.frame(meth_test)
  unmeth_control <- data.frame(unmeth_control)
  unmeth_test <- data.frame(unmeth_test)

  
  
  
  sizeFactor <- function(data) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  # first log
  log_data <- log(data)
  log_data[is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}

  num <- dim(unmeth_control)[2]
  res <- list()
  for (i in 1:dim(meth_test)[2]){
    counts <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
    design = data.frame(
                       Trt = rep(c(rep("control", num),"test"),2), 
                    Meth = c(rep("meth",(num+1)),rep("unmeth", 
                            (num+1))))
    model = ~  Meth + Trt + Meth:Trt
    # 
    dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = design,
                                design = model)
    sf = sizeFactor(cbind((meth_control[,1:num]+unmeth_control[,1:num]),
                       (meth_test[,i]+unmeth_test[,i])))
    
    sizeFactors(dds) = rep(sf,2)
    dds <- DESeq(dds, test = "Wald")
    
    res[[i]] = DESeq2::results(dds, name = "Methunmeth.Trttest",cooksCutoff=FALSE) 
  }
    

 
  
  return(res)
}


res_DESeq2 <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)


end_time <- Sys.time()
run_time_DESeq2 <- end_time-start_time

setwd(path )
saveRDS(res_DESeq2,"res_DESeq2.rds")

rm(res_DESeq2)
```

#### edgeR
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()
library("edgeR")
edgeR_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  num <- dim(meth_control)[2]
  res <- list()
  
  for (i in 1: dim(meth_test)[2]){
       # create deglist
    countdata <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
   
    dgelist <- DGEList(counts = countdata)
    
    
    countdata_all <- cbind((meth_control+unmeth_control),(meth_test[,i]+unmeth_test[,i]))
   
    dgelist_all <- DGEList(counts = countdata_all)
    
    
    # normalize
    dgelist_norm_all <- calcNormFactors(dgelist_all, method = 'TMM')
    
    dgelist_norm <- calcNormFactors(dgelist, method = 'TMM')
    
    dgelist_norm@.Data[[2]] <- rbind(dgelist_norm_all@.Data[[2]],dgelist_norm_all@.Data[[2]])
    
    rownames(dgelist_norm@.Data[[2]]) <- paste0("Sample",1:(2*(num+1)))
    
    
    group1 <- factor(rep(c(rep("control",num),"test"),2))
    group2 <- factor(c(rep("meth",(num+1)),rep("unmeth",(num+1))))
    
    
    
    design <- model.matrix(~group1+group2+group1:group2)
    colnames(design) <- c("control", "test", "unmeth", "unmeth_and_test")
    
    dge <- estimateDisp(dgelist_norm, design, robust = TRUE)
    
    fit <- glmFit(dge, design, robust = TRUE)
    

    lrt <- glmLRT(fit,coef=4)@.Data
    
    res[[i]] <- lrt[[14]]
  }
  
  return(res)
 
}


res_edgeR <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)


end_time <- Sys.time()
run_time_edgeR <- end_time-start_time


setwd(path)
saveRDS(res_edgeR,"res_edgeR.rds")
rm(res_edgeR)

```


#### SigRM

##### SigRM_locfit
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_locfit <-  SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test,method_dispersion = "locfit")
end_time <- Sys.time()
run_time_SigRM_locfit <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_locfit,"res_SigRM_locfit.rds")
rm(res_SigRM_locfit)

```

##### SigRM_unbiased 
```{r,eval=FALSE}
start_time <- Sys.time()
res_SigRM_unbiased <-  SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test)
end_time <- Sys.time()
run_time_SigRM_unbiased <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_unbiased,"res_SigRM_unbiased.rds")
rm(res_SigRM_unbiased)

```

##### SigRM_mle
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_mle <-  SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test,method_dispersion = "mle")
end_time <- Sys.time()
run_time_SigRM_mle <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_mle,"res_SigRM_mle.rds")
rm(res_SigRM_mle)

```


### runtime 
```{r,eval=FALSE}
setwd(path)
runtime <- data.frame(run_time_scDART, run_time_DESeq2, run_time_edgeR, run_time_SigRM_locfit, run_time_SigRM_unbiased, run_time_SigRM_mle)
save(runtime,file = "runtime.RData")
```


## 8

### preprocess
```{r,eval=FALSE}
frequency_all_processed <- readRDS("~/data/SMART/preprocess2/frequency_all_processed.rds")
# get the meth reads and unmeth reads
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")

# get number of control cell and test cell
num_testcell <- length(which(grepl("YTHcell",names(frequency_all_processed))))
num_controlcell <- length(which(grepl("YTHmutcell",names(frequency_all_processed))))


rm(frequency_all_processed)

set.seed(1)
# we random choose 10 control data as the fake test data

test_fake <- sample(1:num_controlcell,10)
control_index <- setdiff(1:num_controlcell,test_fake)

# also random choose 10 real test data, combine them
test_index <- c( sample((num_controlcell+1):(num_controlcell+num_testcell),10), test_fake)


# select random cells
num_cell_control <- 8
control_index <- sample(control_index,num_cell_control)

path <-  paste0("~/data/SMART/results_ks/res",num_cell_control)
setwd(path )
saveRDS(control_index,"control_index.rds")

setwd(path)
saveRDS(test_index,"test_index.rds")

meth_control <- meth[,control_index]
meth_test <- meth[,test_index]

unmeth_control <- unmeth[,control_index]
unmeth_test <- unmeth[,test_index]

rm(meth,unmeth)
```

### put into different model

#### scDART
```{r,eval=FALSE}


start_time <- Sys.time()

# put into the scDART model
single_test_scDART <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  # combine as the total reads
  data1 <- meth_control+unmeth_control
  data2 <- meth_test+unmeth_test
  
  # calculate the mean p and the variance of the control data
  
  p_control <- rowSums(meth_control)/rowSums(data1)
  
  # find the site satisfies the three conditions
  if (length(dim(meth_test)[2])==0){
     # first condition
      index_1 <- which(data2>=20)
      
      # second condition
      p_test <- meth_test/data2
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1] 
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test>=2)
      index <- index_4 [index_4 %in% index_3]
  }else{
    index <- list()
    for (i in 1 : dim(meth_test)[2]){
      
      # first condition
      index_1 <- which(data2[,i]>=20)
      
      # second condition
      p_test <- meth_test[,i]/data2[,i]
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1]
      
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test[,i]>=2)
      index[[i]] <- index_4 [index_4 %in% index_3]
    }
    
    
    
  }
  
  
  return(index)
  
}

res_scDART <-  single_test_scDART(meth_control,meth_test,unmeth_control,unmeth_test)

res_scDART_ <- list()
num_sites <- dim(meth_control)[1]

for (i in 1:length(res_scDART)){
  
  
  # res_scDART
  
  res_scDART_[[i]] <- rep(0,num_sites)
  res_scDART_[[i]][res_scDART[[i]]] <- 1


    
  
}




end_time <- Sys.time()
run_time_scDART <- end_time-start_time

setwd(path )
saveRDS(res_scDART_,"res_scDART_.rds")

rm(res_scDART,res_scDART_)
```


#### DESeq2
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()

library(DESeq2)
DESeq2_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  meth_control <- data.frame(meth_control)
  meth_test <- data.frame(meth_test)
  unmeth_control <- data.frame(unmeth_control)
  unmeth_test <- data.frame(unmeth_test)

  
  
  
  sizeFactor <- function(data) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  # first log
  log_data <- log(data)
  log_data[is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}

  num <- dim(unmeth_control)[2]
  res <- list()
  for (i in 1:dim(meth_test)[2]){
    counts <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
    design = data.frame(
                       Trt = rep(c(rep("control", num),"test"),2), 
                    Meth = c(rep("meth",(num+1)),rep("unmeth", 
                            (num+1))))
    model = ~  Meth + Trt + Meth:Trt
    # 
    dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = design,
                                design = model)
    sf = sizeFactor(cbind((meth_control[,1:num]+unmeth_control[,1:num]),
                       (meth_test[,i]+unmeth_test[,i])))
    
    sizeFactors(dds) = rep(sf,2)
    dds <- DESeq(dds, test = "Wald")
    
    res[[i]] = DESeq2::results(dds, name = "Methunmeth.Trttest",cooksCutoff=FALSE) 
  }
    

 
  
  return(res)
}


res_DESeq2 <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)


end_time <- Sys.time()
run_time_DESeq2 <- end_time-start_time

setwd(path )
saveRDS(res_DESeq2,"res_DESeq2.rds")

rm(res_DESeq2)
```

#### edgeR
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()
library("edgeR")
edgeR_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  num <- dim(meth_control)[2]
  res <- list()
  
  for (i in 1: dim(meth_test)[2]){
       # create deglist
    countdata <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
   
    dgelist <- DGEList(counts = countdata)
    
    
    countdata_all <- cbind((meth_control+unmeth_control),(meth_test[,i]+unmeth_test[,i]))
   
    dgelist_all <- DGEList(counts = countdata_all)
    
    
    # normalize
    dgelist_norm_all <- calcNormFactors(dgelist_all, method = 'TMM')
    
    dgelist_norm <- calcNormFactors(dgelist, method = 'TMM')
    
    dgelist_norm@.Data[[2]] <- rbind(dgelist_norm_all@.Data[[2]],dgelist_norm_all@.Data[[2]])
    
    rownames(dgelist_norm@.Data[[2]]) <- paste0("Sample",1:(2*(num+1)))
    
    
    group1 <- factor(rep(c(rep("control",num),"test"),2))
    group2 <- factor(c(rep("meth",(num+1)),rep("unmeth",(num+1))))
    
    
    
    design <- model.matrix(~group1+group2+group1:group2)
    colnames(design) <- c("control", "test", "unmeth", "unmeth_and_test")
    
    dge <- estimateDisp(dgelist_norm, design, robust = TRUE)
    
    fit <- glmFit(dge, design, robust = TRUE)
    

    lrt <- glmLRT(fit,coef=4)@.Data
    
    res[[i]] <- lrt[[14]]
  }
  
  return(res)
 
}


res_edgeR <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)


end_time <- Sys.time()
run_time_edgeR <- end_time-start_time


setwd(path)
saveRDS(res_edgeR,"res_edgeR.rds")
rm(res_edgeR)

```


#### SigRM

##### SigRM_locfit
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_locfit <-  SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test,method_dispersion = "locfit")
end_time <- Sys.time()
run_time_SigRM_locfit <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_locfit,"res_SigRM_locfit.rds")
rm(res_SigRM_locfit)

```

##### SigRM_unbiased 
```{r,eval=FALSE}
start_time <- Sys.time()
res_SigRM_unbiased <-  SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test)
end_time <- Sys.time()
run_time_SigRM_unbiased <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_unbiased,"res_SigRM_unbiased.rds")
rm(res_SigRM_unbiased)

```

##### SigRM_mle
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_mle <-  SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test,method_dispersion = "mle")
end_time <- Sys.time()
run_time_SigRM_mle <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_mle,"res_SigRM_mle.rds")
rm(res_SigRM_mle)

```


### runtime 
```{r,eval=FALSE}
setwd(path)
runtime <- data.frame(run_time_scDART, run_time_DESeq2, run_time_edgeR, run_time_SigRM_locfit, run_time_SigRM_unbiased, run_time_SigRM_mle)
save(runtime,file = "runtime.RData")
```


## 16

### preprocess
```{r,eval=FALSE}
frequency_all_processed <- readRDS("~/data/SMART/preprocess2/frequency_all_processed.rds")
# get the meth reads and unmeth reads
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")

# get number of control cell and test cell
num_testcell <- length(which(grepl("YTHcell",names(frequency_all_processed))))
num_controlcell <- length(which(grepl("YTHmutcell",names(frequency_all_processed))))


rm(frequency_all_processed)

set.seed(1)
# we random choose 10 control data as the fake test data

test_fake <- sample(1:num_controlcell,10)
control_index <- setdiff(1:num_controlcell,test_fake)

# also random choose 10 real test data, combine them
test_index <- c( sample((num_controlcell+1):(num_controlcell+num_testcell),10), test_fake)


# select random cells
num_cell_control <- 16
control_index <- sample(control_index,num_cell_control)

path <-  paste0("~/data/SMART/results_ks/res",num_cell_control)
setwd(path )
saveRDS(control_index,"control_index.rds")

setwd(path)
saveRDS(test_index,"test_index.rds")

meth_control <- meth[,control_index]
meth_test <- meth[,test_index]

unmeth_control <- unmeth[,control_index]
unmeth_test <- unmeth[,test_index]

rm(meth,unmeth)
```

### put into different model

#### scDART
```{r,eval=FALSE}


start_time <- Sys.time()

# put into the scDART model
single_test_scDART <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  # combine as the total reads
  data1 <- meth_control+unmeth_control
  data2 <- meth_test+unmeth_test
  
  # calculate the mean p and the variance of the control data
  
  p_control <- rowSums(meth_control)/rowSums(data1)
  
  # find the site satisfies the three conditions
  if (length(dim(meth_test)[2])==0){
     # first condition
      index_1 <- which(data2>=20)
      
      # second condition
      p_test <- meth_test/data2
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1] 
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test>=2)
      index <- index_4 [index_4 %in% index_3]
  }else{
    index <- list()
    for (i in 1 : dim(meth_test)[2]){
      
      # first condition
      index_1 <- which(data2[,i]>=20)
      
      # second condition
      p_test <- meth_test[,i]/data2[,i]
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1]
      
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test[,i]>=2)
      index[[i]] <- index_4 [index_4 %in% index_3]
    }
    
    
    
  }
  
  
  return(index)
  
}

res_scDART <-  single_test_scDART(meth_control,meth_test,unmeth_control,unmeth_test)

res_scDART_ <- list()
num_sites <- dim(meth_control)[1]

for (i in 1:length(res_scDART)){
  
  
  # res_scDART
  
  res_scDART_[[i]] <- rep(0,num_sites)
  res_scDART_[[i]][res_scDART[[i]]] <- 1


    
  
}




end_time <- Sys.time()
run_time_scDART <- end_time-start_time

setwd(path )
saveRDS(res_scDART_,"res_scDART_.rds")

rm(res_scDART,res_scDART_)
```


#### DESeq2
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()

library(DESeq2)
DESeq2_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  meth_control <- data.frame(meth_control)
  meth_test <- data.frame(meth_test)
  unmeth_control <- data.frame(unmeth_control)
  unmeth_test <- data.frame(unmeth_test)

  
  
  
  sizeFactor <- function(data) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  # first log
  log_data <- log(data)
  log_data[is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}

  num <- dim(unmeth_control)[2]
  res <- list()
  for (i in 1:dim(meth_test)[2]){
    counts <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
    design = data.frame(
                       Trt = rep(c(rep("control", num),"test"),2), 
                    Meth = c(rep("meth",(num+1)),rep("unmeth", 
                            (num+1))))
    model = ~  Meth + Trt + Meth:Trt
    # 
    dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = design,
                                design = model)
    sf = sizeFactor(cbind((meth_control[,1:num]+unmeth_control[,1:num]),
                       (meth_test[,i]+unmeth_test[,i])))
    
    sizeFactors(dds) = rep(sf,2)
    dds <- DESeq(dds, test = "Wald")
    
    res[[i]] = DESeq2::results(dds, name = "Methunmeth.Trttest",cooksCutoff=FALSE) 
  }
    

 
  
  return(res)
}


res_DESeq2 <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)


end_time <- Sys.time()
run_time_DESeq2 <- end_time-start_time

setwd(path )
saveRDS(res_DESeq2,"res_DESeq2.rds")

rm(res_DESeq2)
```

#### edgeR
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()
library("edgeR")
edgeR_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  num <- dim(meth_control)[2]
  res <- list()
  
  for (i in 1: dim(meth_test)[2]){
       # create deglist
    countdata <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
   
    dgelist <- DGEList(counts = countdata)
    
    
    countdata_all <- cbind((meth_control+unmeth_control),(meth_test[,i]+unmeth_test[,i]))
   
    dgelist_all <- DGEList(counts = countdata_all)
    
    
    # normalize
    dgelist_norm_all <- calcNormFactors(dgelist_all, method = 'TMM')
    
    dgelist_norm <- calcNormFactors(dgelist, method = 'TMM')
    
    dgelist_norm@.Data[[2]] <- rbind(dgelist_norm_all@.Data[[2]],dgelist_norm_all@.Data[[2]])
    
    rownames(dgelist_norm@.Data[[2]]) <- paste0("Sample",1:(2*(num+1)))
    
    
    group1 <- factor(rep(c(rep("control",num),"test"),2))
    group2 <- factor(c(rep("meth",(num+1)),rep("unmeth",(num+1))))
    
    
    
    design <- model.matrix(~group1+group2+group1:group2)
    colnames(design) <- c("control", "test", "unmeth", "unmeth_and_test")
    
    dge <- estimateDisp(dgelist_norm, design, robust = TRUE)
    
    fit <- glmFit(dge, design, robust = TRUE)
    

    lrt <- glmLRT(fit,coef=4)@.Data
    
    res[[i]] <- lrt[[14]]
  }
  
  return(res)
 
}


res_edgeR <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)


end_time <- Sys.time()
run_time_edgeR <- end_time-start_time


setwd(path)
saveRDS(res_edgeR,"res_edgeR.rds")
rm(res_edgeR)

```


#### SigRM

##### SigRM_locfit
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_locfit <-  SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test,method_dispersion = "locfit")
end_time <- Sys.time()
run_time_SigRM_locfit <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_locfit,"res_SigRM_locfit.rds")
rm(res_SigRM_locfit)

```

##### SigRM_unbiased 
```{r,eval=FALSE}
start_time <- Sys.time()
res_SigRM_unbiased <-  SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test)
end_time <- Sys.time()
run_time_SigRM_unbiased <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_unbiased,"res_SigRM_unbiased.rds")
rm(res_SigRM_unbiased)

```

##### SigRM_mle
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_mle <-  SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test,method_dispersion = "mle")
end_time <- Sys.time()
run_time_SigRM_mle <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_mle,"res_SigRM_mle.rds")
rm(res_SigRM_mle)

```


### runtime 
```{r,eval=FALSE}
setwd(path)
runtime <- data.frame(run_time_scDART, run_time_DESeq2, run_time_edgeR, run_time_SigRM_locfit, run_time_SigRM_unbiased, run_time_SigRM_mle)
save(runtime,file = "runtime.RData")
```

## 2

### preprocess
```{r,eval=FALSE}
frequency_all_processed <- readRDS("D:/data/SMART/preprocess2/frequency_all_processed.rds")
# get the meth reads and unmeth reads
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")

# get number of control cell and test cell
num_testcell <- length(which(grepl("YTHcell",names(frequency_all_processed))))
num_controlcell <- length(which(grepl("YTHmutcell",names(frequency_all_processed))))


rm(frequency_all_processed)

set.seed(1)
# we random choose 10 control data as the fake test data

test_fake <- sample(1:num_controlcell,10)
control_index <- setdiff(1:num_controlcell,test_fake)

# also random choose 10 real test data, combine them
test_index <- c( sample((num_controlcell+1):(num_controlcell+num_testcell),10), test_fake)


# select random cells
num_cell_control <- 2
control_index <- sample(control_index,num_cell_control)

path <-  paste0("D:/data/SMART/results_ks/res",num_cell_control)
setwd(path )
saveRDS(control_index,"control_index.rds")

setwd(path)
saveRDS(test_index,"test_index.rds")

meth_control <- meth[,control_index]
meth_test <- meth[,test_index]

unmeth_control <- unmeth[,control_index]
unmeth_test <- unmeth[,test_index]

rm(meth,unmeth)
```

### put into different model

#### TRESS
```{r,eval=FALSE}
library(TRESS)
sizeFactor <- function(data, narm=FALSE) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  
  # first log
  log_data <- log(data)
  log_data [is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data, na.rm=narm)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}
WaldTest_2 <- function(Coef, Cov, Contrast, nullModel ="standN"){
  # simple Wald test for: Contrast^T%*%beta = 0
  #### nullModel: which null model to use to calculate p-value
  Contrast = as.matrix(Contrast, ncol = 1)
  se = rep(NA, nrow(Coef))
  #for (i in 1:nrow(Coef)) {
  for (i in seq_len(nrow(Coef))) {
    if(!all(is.na(Cov[[i]]))){
      tmp = Cov[[i]]
      se[i] = sqrt( t(Contrast)%*%Cov[[i]]%*% Contrast)
    }
  }
  lor = t(Contrast)%*%t(Coef)
  stat = (t(Contrast)%*%t(Coef))/se
  tmp.infer = DMRInfer(stat, nullModel = nullModel)
  return(res = data.frame(logOR = as.numeric(lor),
                          lorSE = as.numeric(se),
                          stat = round(as.numeric(stat), 3),
                          pvalue = as.numeric(tmp.infer$pvalue),
                          padj = as.numeric(tmp.infer$padj)))
}

TRESS_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  data = list()
  data[[1]] <- meth_control
  data[[2]] <- meth_test
  data[[3]] <- unmeth_control
  data[[4]] <- unmeth_test
  res=list()
  
  
  
    replicate = dim(data[[1]])[2]
    counts =
cbind(data[[3]],data[[1]],data[[4]],data[[2]])[, c (c(rbind(1:replicate,(replicate+1):(2*replicate))),c((2*replicate+1):(2*replicate+2)))]
    
    sf <- sizeFactor(cbind(data[[3]],data[[1]],data[[4]],data[[2]]), narm = FALSE) 
  if (anyNA(sf)){
    sf <- sizeFactor(cbind(data[[3]],data[[1]],data[[4]],data[[2]]), narm = TRUE) 
  }
  names(sf) <- NULL
    TRESS.DMR = CallDMRs.paramEsti(counts =counts,
           sf = sf,
           variable = data.frame(predictor =           c(rep("cor",replicate),rep("tr",1))),model = ~1+predictor,FALSE)
    
    
    a = WaldTest_2(TRESS.DMR$Coef,TRESS.DMR$Cov,  c(0, 1),      nullModel = "standN")
    res =a
 
    res
}

start_time <- Sys.time()
res_TRESS <-  list()
for (i in c(1,3:20)){
  meth_control <- meth_control
  meth_test_ <- data.frame(meth_test[,i])
  unmeth_control <- unmeth_control
  unmeth_test_ <- data.frame(unmeth_test[,i])
  
  res_TRESS[[i]] <- TRESS_test(meth_control,meth_test_,unmeth_control,unmeth_test_)
  
}

# the second sample 371049 has error in TRESS so we without running the 371049 in the second sample
i = 2
meth_control <- meth_control[-371049,]
meth_test_ <- data.frame(meth_test[-371049,i])
unmeth_control <- unmeth_control[-371049,]
unmeth_test_ <- data.frame(unmeth_test[-371049,i])

data_frame <- TRESS_test(meth_control,meth_test_,unmeth_control,unmeth_test_)
data_frame <- rbind(rbind(data_frame[1:371048,],rep(NA,5)),data_frame[371049:510553,])
res_TRESS[[2]] <- data_frame





end_time <- Sys.time()
run_time_TRESS <- end_time-start_time



setwd(path)
saveRDS(res_TRESS,"res_TRESS.rds")
saveRDS(run_time_TRESS,"run_time_TRESS.rds")
rm(res_TRESS)
```


## 4

### preprocess
```{r,eval=FALSE}
frequency_all_processed <- readRDS("D:/data/SMART/preprocess2/frequency_all_processed.rds")
# get the meth reads and unmeth reads
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")

# get number of control cell and test cell
num_testcell <- length(which(grepl("YTHcell",names(frequency_all_processed))))
num_controlcell <- length(which(grepl("YTHmutcell",names(frequency_all_processed))))


rm(frequency_all_processed)

set.seed(1)
# we random choose 10 control data as the fake test data

test_fake <- sample(1:num_controlcell,10)
control_index <- setdiff(1:num_controlcell,test_fake)

# also random choose 10 real test data, combine them
test_index <- c( sample((num_controlcell+1):(num_controlcell+num_testcell),10), test_fake)


# select random cells
num_cell_control <- 4
control_index <- sample(control_index,num_cell_control)

path <-  paste0("D:/data/SMART/results_ks/res",num_cell_control)
setwd(path )
saveRDS(control_index,"control_index.rds")

setwd(path)
saveRDS(test_index,"test_index.rds")

meth_control <- meth[,control_index]
meth_test <- meth[,test_index]

unmeth_control <- unmeth[,control_index]
unmeth_test <- unmeth[,test_index]

rm(meth,unmeth)
```

### put into different model

#### TRESS
```{r,eval=FALSE}
library(TRESS)
sizeFactor <- function(data, narm=FALSE) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  
  # first log
  log_data <- log(data)
  log_data [is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data, na.rm=narm)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}
WaldTest_2 <- function(Coef, Cov, Contrast, nullModel ="standN"){
  # simple Wald test for: Contrast^T%*%beta = 0
  #### nullModel: which null model to use to calculate p-value
  Contrast = as.matrix(Contrast, ncol = 1)
  se = rep(NA, nrow(Coef))
  #for (i in 1:nrow(Coef)) {
  for (i in seq_len(nrow(Coef))) {
    if(!all(is.na(Cov[[i]]))){
      tmp = Cov[[i]]
      se[i] = sqrt( t(Contrast)%*%Cov[[i]]%*% Contrast)
    }
  }
  lor = t(Contrast)%*%t(Coef)
  stat = (t(Contrast)%*%t(Coef))/se
  tmp.infer = DMRInfer(stat, nullModel = nullModel)
  return(res = data.frame(logOR = as.numeric(lor),
                          lorSE = as.numeric(se),
                          stat = round(as.numeric(stat), 3),
                          pvalue = as.numeric(tmp.infer$pvalue),
                          padj = as.numeric(tmp.infer$padj)))
}

TRESS_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  data = list()
  data[[1]] <- meth_control
  data[[2]] <- meth_test
  data[[3]] <- unmeth_control
  data[[4]] <- unmeth_test
  res=list()
  
  
  
    replicate = dim(data[[1]])[2]
    counts =
cbind(data[[3]],data[[1]],data[[4]],data[[2]])[, c (c(rbind(1:replicate,(replicate+1):(2*replicate))),c((2*replicate+1):(2*replicate+2)))]
    
    sf <- sizeFactor(cbind(data[[3]],data[[1]],data[[4]],data[[2]]), narm = FALSE) 
  if (anyNA(sf)){
    sf <- sizeFactor(cbind(data[[3]],data[[1]],data[[4]],data[[2]]), narm = TRUE) 
  }
  names(sf) <- NULL
    TRESS.DMR = CallDMRs.paramEsti(counts =counts,
           sf = sf,
           variable = data.frame(predictor =           c(rep("cor",replicate),rep("tr",1))),model = ~1+predictor,FALSE)
    
    
    a = WaldTest_2(TRESS.DMR$Coef,TRESS.DMR$Cov,  c(0, 1),      nullModel = "standN")
    res =a
 
    res
}

start_time <- Sys.time()
res_TRESS <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth_control
  meth_test_ <- meth_test[,i]
  unmeth_control <- unmeth_control
  unmeth_test_ <- unmeth_test[,i]
  
  res_TRESS[[i]] <- TRESS_test(meth_control,meth_test_,unmeth_control,unmeth_test_)
  
}



end_time <- Sys.time()
run_time_TRESS <- end_time-start_time



setwd(path)
saveRDS(res_TRESS,"res_TRESS.rds")
saveRDS(run_time_TRESS,"run_time_TRESS.rds")
rm(res_TRESS)
```

## choose the cell with the similar gene expression pattern as the control groups
## 2

### preprocess
```{r,eval=FALSE}
frequency_all_processed <- readRDS("~/data/SMART/preprocess2/frequency_all_processed.rds")
# get the meth reads and unmeth reads
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")

# get number of control cell and test cell
num_testcell <- length(which(grepl("YTHcell",names(frequency_all_processed))))
num_controlcell <- length(which(grepl("YTHmutcell",names(frequency_all_processed))))


rm(frequency_all_processed)

set.seed(1)
# we random choose 10 control data as the fake test data

test_fake <- sample(1:num_controlcell,10)

test_real <- sample((num_controlcell+1):(num_controlcell+num_testcell),10)
# also random choose 10 real test data, combine them
test_index <- c(test_real, test_fake)


expression_logTPM <- readRDS("~/data/SMART/preprocess2/expression_TPM.rds")
expression_logTPM <- log2(expression_logTPM+1)
index <- c(setdiff(1:num_controlcell,test_fake),(num_controlcell+1):(num_controlcell+num_testcell),test_fake)
expression_logTPM <- expression_logTPM[,index]






similarity_data <- cor(expression_logTPM, method = "pearson")

library(doBy)
  similarity_index <- apply(similarity_data,2, function(x){which.maxn(x,
                                      (num_testcell+num_controlcell))})
  similarity_index <- as.matrix(similarity_index[,(1+num_controlcell-10):(num_testcell+num_controlcell)])
  list_create <- as.list(1: (num_testcell+10))
  
  list_function <- function(x){
    similarity_index_ <- similarity_index[which(similarity_index[,
                                  x]%in%c(1:(num_controlcell-10))),x]
    return(similarity_index_)
  }
  
  similarity_index <- as.data.frame(lapply(list_create,list_function))

        
test_index_ <- c(test_index[1:10]-391,992:1001)

similarity_index <- similarity_index [test_index_]

num_cell_control <- 2

control_index <- lapply(similarity_index,function(df) df[1:num_cell_control])




path <-  paste0("~/data/SMART/results_ks/res",num_cell_control,"sim")
setwd(path )
saveRDS(control_index,"control_index.rds")

setwd(path)
saveRDS(test_index,"test_index.rds")

```

### put into different model

#### scDART
```{r,eval=FALSE}


start_time <- Sys.time()

# put into the scDART model
single_test_scDART <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  # combine as the total reads
  data1 <- meth_control+unmeth_control
  data2 <- meth_test+unmeth_test
  
  # calculate the mean p and the variance of the control data
  
  p_control <- rowSums(meth_control)/rowSums(data1)
  
  # find the site satisfies the three conditions
  if (length(dim(meth_test)[2])==0){
     # first condition
      index_1 <- which(data2>=20)
      
      # second condition
      p_test <- meth_test/data2
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1] 
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test>=2)
      index <- index_4 [index_4 %in% index_3]
  }else{
    index <- list()
    for (i in 1 : dim(meth_test)[2]){
      
      # first condition
      index_1 <- which(data2[,i]>=20)
      
      # second condition
      p_test <- meth_test[,i]/data2[,i]
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1]
      
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test[,i]>=2)
      index[[i]] <- index_4 [index_4 %in% index_3]
    }
    
    
    
  }
  
  
  return(index)
  
}

res_scDART <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_scDART[[i]] <- single_test_scDART(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



res_scDART_ <- list()
num_sites <- dim(meth_control)[1]

for (i in 1:length(res_scDART)){
  
  
  # res_scDART
  
  res_scDART_[[i]] <- rep(0,num_sites)
  res_scDART_[[i]][res_scDART[[i]]] <- 1


    
  
}




end_time <- Sys.time()
run_time_scDART <- end_time-start_time

setwd(path )
saveRDS(res_scDART_,"res_scDART_.rds")

rm(res_scDART,res_scDART_)
```


#### DESeq2
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()

library(DESeq2)
DESeq2_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  meth_control <- data.frame(meth_control)
  meth_test <- data.frame(meth_test)
  unmeth_control <- data.frame(unmeth_control)
  unmeth_test <- data.frame(unmeth_test)

  num <- dim(meth_control)[2]
  
  sizeFactor <- function(data) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  # first log
  log_data <- log(data)
  log_data[is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}

  
  res <- list()
  for (i in 1:dim(meth_test)[2]){
    counts <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
    design = data.frame(
                       Trt = rep(c(rep("control", num),"test"),2), 
                    Meth = c(rep("meth",(num+1)),rep("unmeth", 
                            (num+1))))
    model = ~ Meth + Trt + Meth:Trt
    # 
    dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = design,
                                design = model)
    sf = sizeFactor(cbind((meth_control+unmeth_control),
                       (meth_test[,i]+unmeth_test[,i])))
    
    sizeFactors(dds) = rep(sf,2)
    dds <- DESeq(dds, test = "Wald")
    
    res[[i]] = DESeq2::results(dds, name = "Methunmeth.Trttest",cooksCutoff=FALSE) 
  }
    

 
  
  return(res)
}

res_DESeq2 <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_DESeq2[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



end_time <- Sys.time()
run_time_DESeq2 <- end_time-start_time

setwd(path )
saveRDS(res_DESeq2,"res_DESeq2.rds")

rm(res_DESeq2)
```

#### edgeR
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()
library("edgeR")
edgeR_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  num <- dim(meth_control)[2]
  
  res <- list()
  
  for (i in 1: dim(meth_test)[2]){
       # create deglist
    countdata <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
   
    dgelist <- DGEList(counts = countdata)
    
    
    countdata_all <- cbind((meth_control+unmeth_control),(meth_test[,i]+unmeth_test[,i]))
   
    dgelist_all <- DGEList(counts = countdata_all)
    
    
    # normalize
    dgelist_norm_all <- calcNormFactors(dgelist_all, method = 'TMM')
    
    dgelist_norm <- calcNormFactors(dgelist, method = 'TMM')
    
    dgelist_norm@.Data[[2]] <- rbind(dgelist_norm_all@.Data[[2]],dgelist_norm_all@.Data[[2]])
    
    rownames(dgelist_norm@.Data[[2]]) <- paste0("Sample",1:(2*(num+1)))
    
    
    group1 <- factor(rep(c(rep("control",num),"test"),2))
    group2 <- factor(c(rep("meth",(num+1)),rep("unmeth",(num+1))))
    
    
    
    design <- model.matrix(~0+group1+group2+group1:group2)
    colnames(design) <- c("control", "test", "unmeth", "unmeth_and_test")
    
    dge <- estimateDisp(dgelist_norm, design, robust = TRUE)
    
    fit <- glmFit(dge, design, robust = TRUE)
    

    lrt <- glmLRT(fit,coef=4)@.Data
    
    res[[i]] <- lrt[[14]]
  }
  
  return(res)
 
}


res_edgeR <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,test_index[i]])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,test_index[i]])
  
  res_edgeR[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



end_time <- Sys.time()
run_time_edgeR <- end_time-start_time


setwd(path)
saveRDS(res_edgeR,"res_edgeR.rds")
rm(res_edgeR)

```


#### SigRM

##### SigRM_locfit
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_locfit <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_SigRM_locfit[[i]] <- as.data.frame(SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test,
                                                   method_dispersion = "locfit")) 
  
}

end_time <- Sys.time()
run_time_SigRM_locfit <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_locfit,"res_SigRM_locfit.rds")
rm(res_SigRM_locfit)

```

##### SigRM_unbiased
```{r,eval=FALSE}
start_time <- Sys.time()
res_SigRM_unbiased <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_SigRM_unbiased[[i]] <- as.data.frame(SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test)) 
  
}
end_time <- Sys.time()
run_time_SigRM_unbiased <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_unbiased,"res_SigRM_unbiased.rds")
rm(res_SigRM_unbiased)

```

##### SigRM_mle
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_mle <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_SigRM_mle[[i]] <- as.data.frame(SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test, method_dispersion = "mle")) 
  
}
end_time <- Sys.time()
run_time_SigRM_mle <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_mle,"res_SigRM_mle.rds")
rm(res_SigRM_mle)

```

### runtime 
```{r,eval=FALSE}
setwd(path)
runtime <- data.frame(run_time_scDART, run_time_DESeq2, run_time_edgeR, run_time_SigRM_locfit, run_time_SigRM_unbiased, run_time_SigRM_mle)
save(runtime,file = "runtime.RData")
```


## 4

### preprocess
```{r,eval=FALSE}
frequency_all_processed <- readRDS("~/data/SMART/preprocess2/frequency_all_processed.rds")
# get the meth reads and unmeth reads
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")

# get number of control cell and test cell
num_testcell <- length(which(grepl("YTHcell",names(frequency_all_processed))))
num_controlcell <- length(which(grepl("YTHmutcell",names(frequency_all_processed))))


rm(frequency_all_processed)

set.seed(1)
# we random choose 10 control data as the fake test data

test_fake <- sample(1:num_controlcell,10)

test_real <- sample((num_controlcell+1):(num_controlcell+num_testcell),10)
# also random choose 10 real test data, combine them
test_index <- c(test_real, test_fake)


expression_logTPM <- readRDS("~/data/SMART/preprocess2/expression_TPM.rds")
expression_logTPM <- log2(expression_logTPM+1)
index <- c(setdiff(1:num_controlcell,test_fake),(num_controlcell+1):(num_controlcell+num_testcell),test_fake)
expression_logTPM <- expression_logTPM[,index]






similarity_data <- cor(expression_logTPM, method = "pearson")

library(doBy)
  similarity_index <- apply(similarity_data,2, function(x){which.maxn(x,
                                      (num_testcell+num_controlcell))})
  similarity_index <- as.matrix(similarity_index[,(1+num_controlcell-10):(num_testcell+num_controlcell)])
  list_create <- as.list(1: (num_testcell+10))
  
  list_function <- function(x){
    similarity_index_ <- similarity_index[which(similarity_index[,
                                  x]%in%c(1:(num_controlcell-10))),x]
    return(similarity_index_)
  }
  
  similarity_index <- as.data.frame(lapply(list_create,list_function))

        
test_index_ <- c(test_index[1:10]-391,992:1001)

similarity_index <- similarity_index [test_index_]

num_cell_control <- 4

control_index <- lapply(similarity_index,function(df) df[1:num_cell_control])




path <-  paste0("~/data/SMART/results_ks/res",num_cell_control,"sim")
setwd(path )
saveRDS(control_index,"control_index.rds")

setwd(path)
saveRDS(test_index,"test_index.rds")

```

### put into different model

#### scDART
```{r,eval=FALSE}


start_time <- Sys.time()

# put into the scDART model
single_test_scDART <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  # combine as the total reads
  data1 <- meth_control+unmeth_control
  data2 <- meth_test+unmeth_test
  
  # calculate the mean p and the variance of the control data
  
  p_control <- rowSums(meth_control)/rowSums(data1)
  
  # find the site satisfies the three conditions
  if (length(dim(meth_test)[2])==0){
     # first condition
      index_1 <- which(data2>=20)
      
      # second condition
      p_test <- meth_test/data2
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1] 
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test>=2)
      index <- index_4 [index_4 %in% index_3]
  }else{
    index <- list()
    for (i in 1 : dim(meth_test)[2]){
      
      # first condition
      index_1 <- which(data2[,i]>=20)
      
      # second condition
      p_test <- meth_test[,i]/data2[,i]
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1]
      
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test[,i]>=2)
      index[[i]] <- index_4 [index_4 %in% index_3]
    }
    
    
    
  }
  
  
  return(index)
  
}

res_scDART <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_scDART[[i]] <- single_test_scDART(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



res_scDART_ <- list()
num_sites <- dim(meth_control)[1]

for (i in 1:length(res_scDART)){
  
  
  # res_scDART
  
  res_scDART_[[i]] <- rep(0,num_sites)
  res_scDART_[[i]][res_scDART[[i]]] <- 1


    
  
}




end_time <- Sys.time()
run_time_scDART <- end_time-start_time

setwd(path )
saveRDS(res_scDART_,"res_scDART_.rds")

rm(res_scDART,res_scDART_)
```


#### DESeq2
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()

library(DESeq2)
DESeq2_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  meth_control <- data.frame(meth_control)
  meth_test <- data.frame(meth_test)
  unmeth_control <- data.frame(unmeth_control)
  unmeth_test <- data.frame(unmeth_test)

  num <- dim(meth_control)[2]
  
  sizeFactor <- function(data) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  # first log
  log_data <- log(data)
  log_data[is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}

  
  res <- list()
  for (i in 1:dim(meth_test)[2]){
    counts <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
    design = data.frame(
                       Trt = rep(c(rep("control", num),"test"),2), 
                    Meth = c(rep("meth",(num+1)),rep("unmeth", 
                            (num+1))))
    model = ~ Meth + Trt + Meth:Trt
    # 
    dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = design,
                                design = model)
    sf = sizeFactor(cbind((meth_control+unmeth_control),
                       (meth_test[,i]+unmeth_test[,i])))
    
    sizeFactors(dds) = rep(sf,2)
    dds <- DESeq(dds, test = "Wald")
    
    res[[i]] = DESeq2::results(dds, name = "Methunmeth.Trttest",cooksCutoff=FALSE) 
  }
    

 
  
  return(res)
}

res_DESeq2 <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_DESeq2[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



end_time <- Sys.time()
run_time_DESeq2 <- end_time-start_time

setwd(path )
saveRDS(res_DESeq2,"res_DESeq2.rds")

rm(res_DESeq2)
```

#### edgeR
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()
library("edgeR")
edgeR_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  num <- dim(meth_control)[2]
  
  res <- list()
  
  for (i in 1: dim(meth_test)[2]){
       # create deglist
    countdata <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
   
    dgelist <- DGEList(counts = countdata)
    
    
    countdata_all <- cbind((meth_control+unmeth_control),(meth_test[,i]+unmeth_test[,i]))
   
    dgelist_all <- DGEList(counts = countdata_all)
    
    
    # normalize
    dgelist_norm_all <- calcNormFactors(dgelist_all, method = 'TMM')
    
    dgelist_norm <- calcNormFactors(dgelist, method = 'TMM')
    
    dgelist_norm@.Data[[2]] <- rbind(dgelist_norm_all@.Data[[2]],dgelist_norm_all@.Data[[2]])
    
    rownames(dgelist_norm@.Data[[2]]) <- paste0("Sample",1:(2*(num+1)))
    
    
    group1 <- factor(rep(c(rep("control",num),"test"),2))
    group2 <- factor(c(rep("meth",(num+1)),rep("unmeth",(num+1))))
    
    
    
    design <- model.matrix(~0+group1+group2+group1:group2)
    colnames(design) <- c("control", "test", "unmeth", "unmeth_and_test")
    
    dge <- estimateDisp(dgelist_norm, design, robust = TRUE)
    
    fit <- glmFit(dge, design, robust = TRUE)
    

    lrt <- glmLRT(fit,coef=4)@.Data
    
    res[[i]] <- lrt[[14]]
  }
  
  return(res)
 
}


res_edgeR <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,test_index[i]])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,test_index[i]])
  
  res_edgeR[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



end_time <- Sys.time()
run_time_edgeR <- end_time-start_time


setwd(path)
saveRDS(res_edgeR,"res_edgeR.rds")
rm(res_edgeR)

```


#### SigRM

##### SigRM_locfit
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_locfit <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_SigRM_locfit[[i]] <- as.data.frame(SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test,
                                                   method_dispersion = "locfit")) 
  
}

end_time <- Sys.time()
run_time_SigRM_locfit <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_locfit,"res_SigRM_locfit.rds")
rm(res_SigRM_locfit)

```

##### SigRM_unbiased
```{r,eval=FALSE}
start_time <- Sys.time()
res_SigRM_unbiased <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_SigRM_unbiased[[i]] <- as.data.frame(SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test)) 
  
}
end_time <- Sys.time()
run_time_SigRM_unbiased <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_unbiased,"res_SigRM_unbiased.rds")
rm(res_SigRM_unbiased)

```

##### SigRM_mle
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_mle <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_SigRM_mle[[i]] <- as.data.frame(SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test, method_dispersion = "mle")) 
  
}
end_time <- Sys.time()
run_time_SigRM_mle <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_mle,"res_SigRM_mle.rds")
rm(res_SigRM_mle)

```

### runtime 
```{r,eval=FALSE}
setwd(path)
runtime <- data.frame(run_time_scDART, run_time_DESeq2, run_time_edgeR, run_time_SigRM_locfit, run_time_SigRM_unbiased, run_time_SigRM_mle)
save(runtime,file = "runtime.RData")
```


## 8

### preprocess
```{r,eval=FALSE}
frequency_all_processed <- readRDS("~/data/SMART/preprocess2/frequency_all_processed.rds")
# get the meth reads and unmeth reads
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")

# get number of control cell and test cell
num_testcell <- length(which(grepl("YTHcell",names(frequency_all_processed))))
num_controlcell <- length(which(grepl("YTHmutcell",names(frequency_all_processed))))


rm(frequency_all_processed)

set.seed(1)
# we random choose 10 control data as the fake test data

test_fake <- sample(1:num_controlcell,10)

test_real <- sample((num_controlcell+1):(num_controlcell+num_testcell),10)
# also random choose 10 real test data, combine them
test_index <- c(test_real, test_fake)


expression_logTPM <- readRDS("~/data/SMART/preprocess2/expression_TPM.rds")
expression_logTPM <- log2(expression_logTPM+1)
index <- c(setdiff(1:num_controlcell,test_fake),(num_controlcell+1):(num_controlcell+num_testcell),test_fake)
expression_logTPM <- expression_logTPM[,index]






similarity_data <- cor(expression_logTPM, method = "pearson")

library(doBy)
  similarity_index <- apply(similarity_data,2, function(x){which.maxn(x,
                                      (num_testcell+num_controlcell))})
  similarity_index <- as.matrix(similarity_index[,(1+num_controlcell-10):(num_testcell+num_controlcell)])
  list_create <- as.list(1: (num_testcell+10))
  
  list_function <- function(x){
    similarity_index_ <- similarity_index[which(similarity_index[,
                                  x]%in%c(1:(num_controlcell-10))),x]
    return(similarity_index_)
  }
  
  similarity_index <- as.data.frame(lapply(list_create,list_function))

        
test_index_ <- c(test_index[1:10]-391,992:1001)

similarity_index <- similarity_index [test_index_]

num_cell_control <- 8

control_index <- lapply(similarity_index,function(df) df[1:num_cell_control])




path <-  paste0("~/data/SMART/results_ks/res",num_cell_control,"sim")
setwd(path )
saveRDS(control_index,"control_index.rds")

setwd(path)
saveRDS(test_index,"test_index.rds")

```

### put into different model

#### scDART
```{r,eval=FALSE}


start_time <- Sys.time()

# put into the scDART model
single_test_scDART <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  # combine as the total reads
  data1 <- meth_control+unmeth_control
  data2 <- meth_test+unmeth_test
  
  # calculate the mean p and the variance of the control data
  
  p_control <- rowSums(meth_control)/rowSums(data1)
  
  # find the site satisfies the three conditions
  if (length(dim(meth_test)[2])==0){
     # first condition
      index_1 <- which(data2>=20)
      
      # second condition
      p_test <- meth_test/data2
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1] 
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test>=2)
      index <- index_4 [index_4 %in% index_3]
  }else{
    index <- list()
    for (i in 1 : dim(meth_test)[2]){
      
      # first condition
      index_1 <- which(data2[,i]>=20)
      
      # second condition
      p_test <- meth_test[,i]/data2[,i]
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1]
      
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test[,i]>=2)
      index[[i]] <- index_4 [index_4 %in% index_3]
    }
    
    
    
  }
  
  
  return(index)
  
}

res_scDART <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_scDART[[i]] <- single_test_scDART(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



res_scDART_ <- list()
num_sites <- dim(meth_control)[1]

for (i in 1:length(res_scDART)){
  
  
  # res_scDART
  
  res_scDART_[[i]] <- rep(0,num_sites)
  res_scDART_[[i]][res_scDART[[i]]] <- 1


    
  
}




end_time <- Sys.time()
run_time_scDART <- end_time-start_time

setwd(path )
saveRDS(res_scDART_,"res_scDART_.rds")

rm(res_scDART,res_scDART_)
```


#### DESeq2
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()

library(DESeq2)
DESeq2_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  meth_control <- data.frame(meth_control)
  meth_test <- data.frame(meth_test)
  unmeth_control <- data.frame(unmeth_control)
  unmeth_test <- data.frame(unmeth_test)

  num <- dim(meth_control)[2]
  
  sizeFactor <- function(data) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  # first log
  log_data <- log(data)
  log_data[is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}

  
  res <- list()
  for (i in 1:dim(meth_test)[2]){
    counts <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
    design = data.frame(
                       Trt = rep(c(rep("control", num),"test"),2), 
                    Meth = c(rep("meth",(num+1)),rep("unmeth", 
                            (num+1))))
    model = ~ Meth + Trt + Meth:Trt
    # 
    dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = design,
                                design = model)
    sf = sizeFactor(cbind((meth_control+unmeth_control),
                       (meth_test[,i]+unmeth_test[,i])))
    
    sizeFactors(dds) = rep(sf,2)
    dds <- DESeq(dds, test = "Wald")
    
    res[[i]] = DESeq2::results(dds, name = "Methunmeth.Trttest",cooksCutoff=FALSE) 
  }
    

 
  
  return(res)
}

res_DESeq2 <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_DESeq2[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



end_time <- Sys.time()
run_time_DESeq2 <- end_time-start_time

setwd(path )
saveRDS(res_DESeq2,"res_DESeq2.rds")

rm(res_DESeq2)
```

#### edgeR
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()
library("edgeR")
edgeR_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  num <- dim(meth_control)[2]
  
  res <- list()
  
  for (i in 1: dim(meth_test)[2]){
       # create deglist
    countdata <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
   
    dgelist <- DGEList(counts = countdata)
    
    
    countdata_all <- cbind((meth_control+unmeth_control),(meth_test[,i]+unmeth_test[,i]))
   
    dgelist_all <- DGEList(counts = countdata_all)
    
    
    # normalize
    dgelist_norm_all <- calcNormFactors(dgelist_all, method = 'TMM')
    
    dgelist_norm <- calcNormFactors(dgelist, method = 'TMM')
    
    dgelist_norm@.Data[[2]] <- rbind(dgelist_norm_all@.Data[[2]],dgelist_norm_all@.Data[[2]])
    
    rownames(dgelist_norm@.Data[[2]]) <- paste0("Sample",1:(2*(num+1)))
    
    
    group1 <- factor(rep(c(rep("control",num),"test"),2))
    group2 <- factor(c(rep("meth",(num+1)),rep("unmeth",(num+1))))
    
    
    
    design <- model.matrix(~0+group1+group2+group1:group2)
    colnames(design) <- c("control", "test", "unmeth", "unmeth_and_test")
    
    dge <- estimateDisp(dgelist_norm, design, robust = TRUE)
    
    fit <- glmFit(dge, design, robust = TRUE)
    

    lrt <- glmLRT(fit,coef=4)@.Data
    
    res[[i]] <- lrt[[14]]
  }
  
  return(res)
 
}


res_edgeR <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,test_index[i]])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,test_index[i]])
  
  res_edgeR[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



end_time <- Sys.time()
run_time_edgeR <- end_time-start_time


setwd(path)
saveRDS(res_edgeR,"res_edgeR.rds")
rm(res_edgeR)

```


#### SigRM

##### SigRM_locfit
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_locfit <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_SigRM_locfit[[i]] <- as.data.frame(SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test,
                                                   method_dispersion = "locfit")) 
  
}

end_time <- Sys.time()
run_time_SigRM_locfit <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_locfit,"res_SigRM_locfit.rds")
rm(res_SigRM_locfit)

```

##### SigRM_unbiased
```{r,eval=FALSE}
start_time <- Sys.time()
res_SigRM_unbiased <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_SigRM_unbiased[[i]] <- as.data.frame(SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test)) 
  
}
end_time <- Sys.time()
run_time_SigRM_unbiased <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_unbiased,"res_SigRM_unbiased.rds")
rm(res_SigRM_unbiased)

```

##### SigRM_mle
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_mle <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_SigRM_mle[[i]] <- as.data.frame(SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test, method_dispersion = "mle")) 
  
}
end_time <- Sys.time()
run_time_SigRM_mle <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_mle,"res_SigRM_mle.rds")
rm(res_SigRM_mle)

```

### runtime 
```{r,eval=FALSE}
setwd(path)
runtime <- data.frame(run_time_scDART, run_time_DESeq2, run_time_edgeR, run_time_SigRM_locfit, run_time_SigRM_unbiased, run_time_SigRM_mle)
save(runtime,file = "runtime.RData")
```


## 16

### preprocess
```{r,eval=FALSE}
frequency_all_processed <- readRDS("~/data/SMART/preprocess2/frequency_all_processed.rds")
# get the meth reads and unmeth reads
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")

# get number of control cell and test cell
num_testcell <- length(which(grepl("YTHcell",names(frequency_all_processed))))
num_controlcell <- length(which(grepl("YTHmutcell",names(frequency_all_processed))))


rm(frequency_all_processed)

set.seed(1)
# we random choose 10 control data as the fake test data

test_fake <- sample(1:num_controlcell,10)

test_real <- sample((num_controlcell+1):(num_controlcell+num_testcell),10)
# also random choose 10 real test data, combine them
test_index <- c(test_real, test_fake)


expression_logTPM <- readRDS("~/data/SMART/preprocess2/expression_TPM.rds")
expression_logTPM <- log2(expression_logTPM+1)
index <- c(setdiff(1:num_controlcell,test_fake),(num_controlcell+1):(num_controlcell+num_testcell),test_fake)
expression_logTPM <- expression_logTPM[,index]






similarity_data <- cor(expression_logTPM, method = "pearson")

library(doBy)
  similarity_index <- apply(similarity_data,2, function(x){which.maxn(x,
                                      (num_testcell+num_controlcell))})
  similarity_index <- as.matrix(similarity_index[,(1+num_controlcell-10):(num_testcell+num_controlcell)])
  list_create <- as.list(1: (num_testcell+10))
  
  list_function <- function(x){
    similarity_index_ <- similarity_index[which(similarity_index[,
                                  x]%in%c(1:(num_controlcell-10))),x]
    return(similarity_index_)
  }
  
  similarity_index <- as.data.frame(lapply(list_create,list_function))

        
test_index_ <- c(test_index[1:10]-391,992:1001)

similarity_index <- similarity_index [test_index_]

num_cell_control <- 16

control_index <- lapply(similarity_index,function(df) df[1:num_cell_control])




path <-  paste0("~/data/SMART/results_ks/res",num_cell_control,"sim")
setwd(path )
saveRDS(control_index,"control_index.rds")

setwd(path)
saveRDS(test_index,"test_index.rds")

```

### put into different model

#### scDART
```{r,eval=FALSE}


start_time <- Sys.time()

# put into the scDART model
single_test_scDART <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  # combine as the total reads
  data1 <- meth_control+unmeth_control
  data2 <- meth_test+unmeth_test
  
  # calculate the mean p and the variance of the control data
  
  p_control <- rowSums(meth_control)/rowSums(data1)
  
  # find the site satisfies the three conditions
  if (length(dim(meth_test)[2])==0){
     # first condition
      index_1 <- which(data2>=20)
      
      # second condition
      p_test <- meth_test/data2
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1] 
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test>=2)
      index <- index_4 [index_4 %in% index_3]
  }else{
    index <- list()
    for (i in 1 : dim(meth_test)[2]){
      
      # first condition
      index_1 <- which(data2[,i]>=20)
      
      # second condition
      p_test <- meth_test[,i]/data2[,i]
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1]
      
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test[,i]>=2)
      index[[i]] <- index_4 [index_4 %in% index_3]
    }
    
    
    
  }
  
  
  return(index)
  
}

res_scDART <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_scDART[[i]] <- single_test_scDART(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



res_scDART_ <- list()
num_sites <- dim(meth_control)[1]

for (i in 1:length(res_scDART)){
  
  
  # res_scDART
  
  res_scDART_[[i]] <- rep(0,num_sites)
  res_scDART_[[i]][res_scDART[[i]]] <- 1


    
  
}




end_time <- Sys.time()
run_time_scDART <- end_time-start_time

setwd(path )
saveRDS(res_scDART_,"res_scDART_.rds")

rm(res_scDART,res_scDART_)
```


#### DESeq2
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()

library(DESeq2)
DESeq2_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  meth_control <- data.frame(meth_control)
  meth_test <- data.frame(meth_test)
  unmeth_control <- data.frame(unmeth_control)
  unmeth_test <- data.frame(unmeth_test)

  num <- dim(meth_control)[2]
  
  sizeFactor <- function(data) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  # first log
  log_data <- log(data)
  log_data[is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}

  
  res <- list()
  for (i in 1:dim(meth_test)[2]){
    counts <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
    design = data.frame(
                       Trt = rep(c(rep("control", num),"test"),2), 
                    Meth = c(rep("meth",(num+1)),rep("unmeth", 
                            (num+1))))
    model = ~ Meth + Trt + Meth:Trt
    # 
    dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = design,
                                design = model)
    sf = sizeFactor(cbind((meth_control+unmeth_control),
                       (meth_test[,i]+unmeth_test[,i])))
    
    sizeFactors(dds) = rep(sf,2)
    dds <- DESeq(dds, test = "Wald")
    
    res[[i]] = DESeq2::results(dds, name = "Methunmeth.Trttest",cooksCutoff=FALSE) 
  }
    

 
  
  return(res)
}

res_DESeq2 <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_DESeq2[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



end_time <- Sys.time()
run_time_DESeq2 <- end_time-start_time

setwd(path )
saveRDS(res_DESeq2,"res_DESeq2.rds")

rm(res_DESeq2)
```

#### edgeR
```{r,eval=FALSE}
set.seed(1)
start_time <- Sys.time()
library("edgeR")
edgeR_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  num <- dim(meth_control)[2]
  
  res <- list()
  
  for (i in 1: dim(meth_test)[2]){
       # create deglist
    countdata <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
   
    dgelist <- DGEList(counts = countdata)
    
    
    countdata_all <- cbind((meth_control+unmeth_control),(meth_test[,i]+unmeth_test[,i]))
   
    dgelist_all <- DGEList(counts = countdata_all)
    
    
    # normalize
    dgelist_norm_all <- calcNormFactors(dgelist_all, method = 'TMM')
    
    dgelist_norm <- calcNormFactors(dgelist, method = 'TMM')
    
    dgelist_norm@.Data[[2]] <- rbind(dgelist_norm_all@.Data[[2]],dgelist_norm_all@.Data[[2]])
    
    rownames(dgelist_norm@.Data[[2]]) <- paste0("Sample",1:(2*(num+1)))
    
    
    group1 <- factor(rep(c(rep("control",num),"test"),2))
    group2 <- factor(c(rep("meth",(num+1)),rep("unmeth",(num+1))))
    
    
    
    design <- model.matrix(~0+group1+group2+group1:group2)
    colnames(design) <- c("control", "test", "unmeth", "unmeth_and_test")
    
    dge <- estimateDisp(dgelist_norm, design, robust = TRUE)
    
    fit <- glmFit(dge, design, robust = TRUE)
    

    lrt <- glmLRT(fit,coef=4)@.Data
    
    res[[i]] <- lrt[[14]]
  }
  
  return(res)
 
}


res_edgeR <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,test_index[i]])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,test_index[i]])
  
  res_edgeR[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



end_time <- Sys.time()
run_time_edgeR <- end_time-start_time


setwd(path)
saveRDS(res_edgeR,"res_edgeR.rds")
rm(res_edgeR)

```


#### SigRM

##### SigRM_locfit
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_locfit <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_SigRM_locfit[[i]] <- as.data.frame(SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test,
                                                   method_dispersion = "locfit")) 
  
}

end_time <- Sys.time()
run_time_SigRM_locfit <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_locfit,"res_SigRM_locfit.rds")
rm(res_SigRM_locfit)

```

##### SigRM_unbiased
```{r,eval=FALSE}
start_time <- Sys.time()
res_SigRM_unbiased <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_SigRM_unbiased[[i]] <- as.data.frame(SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test)) 
  
}
end_time <- Sys.time()
run_time_SigRM_unbiased <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_unbiased,"res_SigRM_unbiased.rds")
rm(res_SigRM_unbiased)

```

##### SigRM_mle
```{r,eval=FALSE}
start_time <- Sys.time()
# put into the SigRM model 
library(SigRM)
res_SigRM_mle <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- meth[,test_index[i]]
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- unmeth[,test_index[i]]
  
  res_SigRM_mle[[i]] <- as.data.frame(SigRMtest(meth_control,meth_test,unmeth_control,unmeth_test, method_dispersion = "mle")) 
  
}
end_time <- Sys.time()
run_time_SigRM_mle <- end_time-start_time

setwd(path)
saveRDS(res_SigRM_mle,"res_SigRM_mle.rds")
rm(res_SigRM_mle)

```

### runtime 
```{r,eval=FALSE}
setwd(path)
runtime <- data.frame(run_time_scDART, run_time_DESeq2, run_time_edgeR, run_time_SigRM_locfit, run_time_SigRM_unbiased, run_time_SigRM_mle)
save(runtime,file = "runtime.RData")
```

## 2

### preprocess
```{r,eval=FALSE}
frequency_all_processed <- readRDS("~/data/SMART/preprocess2/frequency_all_processed.rds")
# get the meth reads and unmeth reads
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")

# get number of control cell and test cell
num_testcell <- length(which(grepl("YTHcell",names(frequency_all_processed))))
num_controlcell <- length(which(grepl("YTHmutcell",names(frequency_all_processed))))


rm(frequency_all_processed)

set.seed(1)
# we random choose 10 control data as the fake test data

test_fake <- sample(1:num_controlcell,10)

test_real <- sample((num_controlcell+1):(num_controlcell+num_testcell),10)
# also random choose 10 real test data, combine them
test_index <- c(test_real, test_fake)


expression_logTPM <- readRDS("~/data/SMART/preprocess2/expression_TPM.rds")
expression_logTPM <- log2(expression_logTPM+1)
index <- c(setdiff(1:num_controlcell,test_fake),(num_controlcell+1):(num_controlcell+num_testcell),test_fake)
expression_logTPM <- expression_logTPM[,index]






similarity_data <- cor(expression_logTPM, method = "pearson")

library(doBy)
  similarity_index <- apply(similarity_data,2, function(x){which.maxn(x,
                                      (num_testcell+num_controlcell))})
  similarity_index <- as.matrix(similarity_index[,(1+num_controlcell-10):(num_testcell+num_controlcell)])
  list_create <- as.list(1: (num_testcell+10))
  
  list_function <- function(x){
    similarity_index_ <- similarity_index[which(similarity_index[,
                                  x]%in%c(1:(num_controlcell-10))),x]
    return(similarity_index_)
  }
  
  similarity_index <- as.data.frame(lapply(list_create,list_function))

        
test_index_ <- c(test_index[1:10]-391,992:1001)

similarity_index <- similarity_index [test_index_]

num_cell_control <- 2

control_index <- lapply(similarity_index,function(df) df[1:num_cell_control])




path <-  paste0("~/data/SMART/results_ks/res",num_cell_control,"sim")
setwd(path )
saveRDS(control_index,"control_index.rds")

setwd(path)
saveRDS(test_index,"test_index.rds")

```

### put into different model

#### TRESS
```{r,eval=FALSE}
library(TRESS)
sizeFactor <- function(data, narm=FALSE) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  
  # first log
  log_data <- log(data)
  log_data [is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data, na.rm=narm)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}
WaldTest_2 <- function(Coef, Cov, Contrast, nullModel ="standN"){
  # simple Wald test for: Contrast^T%*%beta = 0
  #### nullModel: which null model to use to calculate p-value
  Contrast = as.matrix(Contrast, ncol = 1)
  se = rep(NA, nrow(Coef))
  #for (i in 1:nrow(Coef)) {
  for (i in seq_len(nrow(Coef))) {
    if(!all(is.na(Cov[[i]]))){
      tmp = Cov[[i]]
      se[i] = sqrt( t(Contrast)%*%Cov[[i]]%*% Contrast)
    }
  }
  lor = t(Contrast)%*%t(Coef)
  stat = (t(Contrast)%*%t(Coef))/se
  tmp.infer = DMRInfer(stat, nullModel = nullModel)
  return(res = data.frame(logOR = as.numeric(lor),
                          lorSE = as.numeric(se),
                          stat = round(as.numeric(stat), 3),
                          pvalue = as.numeric(tmp.infer$pvalue),
                          padj = as.numeric(tmp.infer$padj)))
}

TRESS_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  data = list()
  data[[1]] <- meth_control
  data[[2]] <- meth_test
  data[[3]] <- unmeth_control
  data[[4]] <- unmeth_test
  res=list()
  
  
  
    replicate = dim(data[[1]])[2]
    counts =
cbind(data[[3]],data[[1]],data[[4]],data[[2]])[, c (c(rbind(1:replicate,(replicate+1):(2*replicate))),c((2*replicate+1):(2*replicate+2)))]
    
    sf <- sizeFactor(cbind(data[[3]],data[[1]],data[[4]],data[[2]]), narm = FALSE) 
  if (anyNA(sf)){
    sf <- sizeFactor(cbind(data[[3]],data[[1]],data[[4]],data[[2]]), narm = TRUE) 
  }
  names(sf) <- NULL
    TRESS.DMR = CallDMRs.paramEsti(counts =counts,
           sf = sf,
           variable = data.frame(predictor =           c(rep("cor",replicate),rep("tr",1))),model = ~1+predictor,FALSE)
    
    
    a = WaldTest_2(TRESS.DMR$Coef,TRESS.DMR$Cov,  c(0, 1),      nullModel = "standN")
    res =a
 
    res
}

start_time <- Sys.time()
res_TRESS <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,test_index[i]])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,test_index[i]])
  
  res_TRESS[[i]] <- TRESS_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



end_time <- Sys.time()
run_time_TRESS <- end_time-start_time



setwd(path)
saveRDS(res_TRESS,"res_TRESS.rds")
saveRDS(run_time_TRESS,"run_time_TRESS.rds")
rm(res_TRESS)
```



## 4

### preprocess
```{r,eval=FALSE}
frequency_all_processed <- readRDS("~/data/SMART/preprocess2/frequency_all_processed.rds")
# get the meth reads and unmeth reads
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")

# get number of control cell and test cell
num_testcell <- length(which(grepl("YTHcell",names(frequency_all_processed))))
num_controlcell <- length(which(grepl("YTHmutcell",names(frequency_all_processed))))


rm(frequency_all_processed)

set.seed(1)
# we random choose 10 control data as the fake test data

test_fake <- sample(1:num_controlcell,10)

test_real <- sample((num_controlcell+1):(num_controlcell+num_testcell),10)
# also random choose 10 real test data, combine them
test_index <- c(test_real, test_fake)


expression_logTPM <- readRDS("~/data/SMART/preprocess2/expression_TPM.rds")
expression_logTPM <- log2(expression_logTPM+1)
index <- c(setdiff(1:num_controlcell,test_fake),(num_controlcell+1):(num_controlcell+num_testcell),test_fake)
expression_logTPM <- expression_logTPM[,index]






similarity_data <- cor(expression_logTPM, method = "pearson")

library(doBy)
  similarity_index <- apply(similarity_data,2, function(x){which.maxn(x,
                                      (num_testcell+num_controlcell))})
  similarity_index <- as.matrix(similarity_index[,(1+num_controlcell-10):(num_testcell+num_controlcell)])
  list_create <- as.list(1: (num_testcell+10))
  
  list_function <- function(x){
    similarity_index_ <- similarity_index[which(similarity_index[,
                                  x]%in%c(1:(num_controlcell-10))),x]
    return(similarity_index_)
  }
  
  similarity_index <- as.data.frame(lapply(list_create,list_function))

        
test_index_ <- c(test_index[1:10]-391,992:1001)

similarity_index <- similarity_index [test_index_]

num_cell_control <- 4

control_index <- lapply(similarity_index,function(df) df[1:num_cell_control])




path <-  paste0("~/data/SMART/results_ks/res",num_cell_control,"sim")
setwd(path )
saveRDS(control_index,"control_index.rds")

setwd(path)
saveRDS(test_index,"test_index.rds")

```

### put into different model

#### TRESS
```{r,eval=FALSE}
library(TRESS)
sizeFactor <- function(data, narm=FALSE) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  
  # first log
  log_data <- log(data)
  log_data [is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data, na.rm=narm)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}
WaldTest_2 <- function(Coef, Cov, Contrast, nullModel ="standN"){
  # simple Wald test for: Contrast^T%*%beta = 0
  #### nullModel: which null model to use to calculate p-value
  Contrast = as.matrix(Contrast, ncol = 1)
  se = rep(NA, nrow(Coef))
  #for (i in 1:nrow(Coef)) {
  for (i in seq_len(nrow(Coef))) {
    if(!all(is.na(Cov[[i]]))){
      tmp = Cov[[i]]
      se[i] = sqrt( t(Contrast)%*%Cov[[i]]%*% Contrast)
    }
  }
  lor = t(Contrast)%*%t(Coef)
  stat = (t(Contrast)%*%t(Coef))/se
  tmp.infer = DMRInfer(stat, nullModel = nullModel)
  return(res = data.frame(logOR = as.numeric(lor),
                          lorSE = as.numeric(se),
                          stat = round(as.numeric(stat), 3),
                          pvalue = as.numeric(tmp.infer$pvalue),
                          padj = as.numeric(tmp.infer$padj)))
}

TRESS_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  data = list()
  data[[1]] <- meth_control
  data[[2]] <- meth_test
  data[[3]] <- unmeth_control
  data[[4]] <- unmeth_test
  res=list()
  
  
  
    replicate = dim(data[[1]])[2]
    counts =
cbind(data[[3]],data[[1]],data[[4]],data[[2]])[, c (c(rbind(1:replicate,(replicate+1):(2*replicate))),c((2*replicate+1):(2*replicate+2)))]
    
    sf <- sizeFactor(cbind(data[[3]],data[[1]],data[[4]],data[[2]]), narm = FALSE) 
  if (anyNA(sf)){
    sf <- sizeFactor(cbind(data[[3]],data[[1]],data[[4]],data[[2]]), narm = TRUE) 
  }
  names(sf) <- NULL
    TRESS.DMR = CallDMRs.paramEsti(counts =counts,
           sf = sf,
           variable = data.frame(predictor =           c(rep("cor",replicate),rep("tr",1))),model = ~1+predictor,FALSE)
    
    
    a = WaldTest_2(TRESS.DMR$Coef,TRESS.DMR$Cov,  c(0, 1),      nullModel = "standN")
    res =a
 
    res
}

start_time <- Sys.time()
res_TRESS <-  list()
for (i in 1:length(test_index)){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,test_index[i]])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,test_index[i]])
  
  res_TRESS[[i]] <- TRESS_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}



end_time <- Sys.time()
run_time_TRESS <- end_time-start_time



setwd(path)
saveRDS(res_TRESS,"res_TRESS.rds")
saveRDS(run_time_TRESS,"run_time_TRESS.rds")
rm(res_TRESS)
```

### ks test 
```{r,eval=FALSE}
rm(list=ls())
num_cell_control <- 16 #can change to 2,4,8,16

path <-  paste0("D:/data/SMART/results_ks/res",num_cell_control)

setwd(path)

res_DESeq2 <- readRDS(paste0(path,"/res_DESeq2.rds"))
res_scDART_ <- readRDS(paste0(path,"/res_scDART_.rds"))
res_SigRM_locfit <- readRDS(paste0(path,"/res_SigRM_locfit.rds"))
res_SigRM_mle <- readRDS(paste0(path,"/res_SigRM_mle.rds"))
res_SigRM_unbiased <- readRDS(paste0(path,"/res_SigRM_unbiased.rds"))
res_edgeR <- readRDS(paste0(path,"/res_edgeR.rds"))



# only see the sites need has the results
# if the site without any read in the test or control 
# we do not need to give a result

# the following is the index of site need have results, which are the results SigRM report

index_nona <- list()
set.seed(1)
for (i in 1: length(res_DESeq2)){
  index_nona[[i]] <- which(!is.na(res_SigRM_mle[[5]][,i]))
}





res_SigRM_unbiased_ks <- c()
res_SigRM_locfit_ks <- c()
res_SigRM_mle_ks <- c()
res_scDART_ks <- c()
res_DESeq2_ks <- c()

ks_function <- function(x,y){
  
  ks_test <- ks.test(x,y)
  l1 <- length(ks_test$data$x)
  l2 <- length(ks_test$data$y)
  D <- ks_test$statistic
  ks_staitics <- D/sqrt(1/l1+1/l2)
  
  return(ks_staitics)
  
}


ks_list_function <- function (x,index_x,y,index_y){
  x <- Map(function(x, index_x) x[index_x], x, index_x)
  y <- Map(function(y, index_y) y[index_y], y, index_y)

  l_y <- length(y)
  ks_list <- list()
  for (i in 1:l_y){
    a <- lapply(x,function(x) ks_function(x,y[[i]]))
    ks_list <- c(ks_list,a)
  }
  return(ks_list)
}


   
  res_SigRM_locfit_ks <- ks_list_function(res_SigRM_locfit[[5]][1:10],index_nona[1:10],res_SigRM_locfit[[5]][11:20],index_nona[11:20])
  
    res_SigRM_mle_ks <- ks_list_function(res_SigRM_mle[[5]][1:10],index_nona[1:10],res_SigRM_mle[[5]][11:20],index_nona[11:20])
    
   res_SigRM_unbiased_ks <- ks_list_function(res_SigRM_unbiased[[5]][1:10],index_nona[1:10],res_SigRM_unbiased[[5]][11:20],index_nona[11:20])
     
  res_scDART_ks <- ks_list_function(res_scDART_[1:10],index_nona[1:  10],res_scDART_[11:20],index_nona[
    11:20])
  

  
  res_DESeq2_pvalue <- lapply(res_DESeq2,function(df) df[,5])

  
  res_DESeq2_ks <- ks_list_function(res_DESeq2_pvalue[1:10],index_nona[1:  10],res_DESeq2_pvalue[11:20],index_nona[11:20])
  

  
    res_edgeR_pvalue <- lapply(res_edgeR,function(df) df[,4])

  
  res_edgeR_ks <- ks_list_function(res_edgeR_pvalue[1:10],index_nona[1:  10],res_edgeR_pvalue[11:20],index_nona[11:20])
    
  

kstest_data_origin <- matrix(c(as.numeric(res_SigRM_locfit_ks),as.numeric(res_SigRM_mle_ks),as.numeric(res_SigRM_unbiased_ks),as.numeric(res_scDART_ks), as.numeric(res_DESeq2_ks),as.numeric(res_edgeR_ks),
                     rep("SigRM_locfit",100),rep("SigRM_mle",100),rep("SigRM_unbiased",100),rep("scDART",100),rep("DESeq2",100),
                     rep("edgeR",100)),ncol=2)
kstest_data_origin  <- data.frame(kstest_data_origin)
colnames(kstest_data_origin ) <- c("ks","method")
kstest_data_origin$ks <- as.numeric(kstest_data_origin$ks)



setwd(path)
saveRDS(kstest_data_origin,"kstest_data_origin.rds")

```






### ks test sim
```{r,eval=FALSE}
rm(list=ls())
num_cell_control <- 16 #can change to 2,4,8,16

path <-  paste0("D:/data/SMART/results_ks/res",num_cell_control,"sim")

res_DESeq2 <- readRDS(paste0(path,"/res_DESeq2.rds"))
res_scDART_ <- readRDS(paste0(path,"/res_scDART_.rds"))
res_SigRM_locfit <- readRDS(paste0(path,"/res_SigRM_locfit.rds"))
res_SigRM_mle <- readRDS(paste0(path,"/res_SigRM_mle.rds"))
res_SigRM_unbiased <- readRDS(paste0(path,"/res_SigRM_unbiased.rds"))
res_edgeR <- readRDS(paste0(path,"/res_edgeR.rds"))




# only see the sites need has the results
# if the site without any read in the test or control 
# we do not need to give a result

# the following is the index of site need have results, which are the results SigRM report
index_nona <- list()
set.seed(1)
for (i in 1: length(res_DESeq2)){
  index_nona[[i]] <- which(!is.na(res_SigRM_mle[[i]][[5]]))
}


res_SigRM_unbiased_ks <- c()
res_SigRM_locfit_ks <- c()
res_SigRM_mle_ks <- c()
res_scDART_ks <- c()
res_DESeq2_ks <- c()

ks_function <- function(x,y){
  
  ks_test <- ks.test(x,y)
  l1 <- length(ks_test$data$x)
  l2 <- length(ks_test$data$y)
  D <- ks_test$statistic
  ks_staitics <- D/sqrt(1/l1+1/l2)
  
  return(ks_staitics)
  
}


ks_list_function <- function (x,index_x,y,index_y){
  x <- Map(function(x, index_x) x[index_x], x, index_x)
  y <- Map(function(y, index_y) y[index_y], y, index_y)

  l_y <- length(y)
  ks_list <- list()
  for (i in 1:l_y){
    a <- lapply(x,function(x,index_x) ks_function(x,y[[i]]))
    ks_list <- c(ks_list,a)
  }
  return(ks_list)
}




    
    res_SigRM_locfit_pvalue <- lapply(res_SigRM_locfit,function(df) df[,5])
    
    res_SigRM_locfit_ks <- ks_list_function(res_SigRM_locfit_pvalue[1:10],index_nona[1:10],res_SigRM_locfit_pvalue[11:20],index_nona[11:20])
    
    res_SigRM_mle_pvalue <- lapply(res_SigRM_mle,function(df) df[,5])
  
    res_SigRM_mle_ks <- ks_list_function(res_SigRM_mle_pvalue[1:10],index_nona[1:10],res_SigRM_mle_pvalue[11:20],index_nona[11:20])
    
    res_SigRM_unbiased_pvalue <- lapply(res_SigRM_unbiased,function(df) df[,5])
    
    res_SigRM_unbiased_ks <- ks_list_function(res_SigRM_unbiased_pvalue[1:10],index_nona[1:10],res_SigRM_unbiased_pvalue[11:20],index_nona[11:20])
     
    res_scDART_ks <- ks_list_function(res_scDART_[1:10],index_nona[1:  10],res_scDART_[11:20],index_nona[
    11:20])
  
    res_DESeq2_pvalue <- lapply(res_DESeq2,function(df) df[[1]][,5])
 
  
    res_DESeq2_ks <- ks_list_function(res_DESeq2_pvalue[1:10],index_nona[1:  10],res_DESeq2_pvalue[11:20],index_nona[11:20])
  
    res_edgeR_pvalue <- lapply(res_edgeR,function(df) df[[1]][,4])
 
  
    res_edgeR_ks <- ks_list_function(res_edgeR_pvalue[1:10],index_nona[1:  10],res_edgeR_pvalue[11:20],index_nona[11:20])
    
    
   


kstest_data_origin <- matrix(c(as.numeric(res_SigRM_locfit_ks),as.numeric(res_SigRM_mle_ks),as.numeric(res_SigRM_unbiased_ks),as.numeric(res_scDART_ks), as.numeric(res_DESeq2_ks),as.numeric(res_edgeR_ks),
                     rep("SigRM_locfit",100),rep("SigRM_mle",100),rep("SigRM_unbiased",100),rep("scDART",100),rep("DESeq2",100),
                     rep("edgeR",100)),ncol=2)
kstest_data_origin  <- data.frame(kstest_data_origin)
colnames(kstest_data_origin ) <- c("ks","method")
kstest_data_origin$ks <- as.numeric(kstest_data_origin$ks)


setwd(path)
saveRDS(kstest_data_origin,"kstest_data_origin.rds")

```




### ks test 
```{r,eval=FALSE}
rm(list=ls())
num_cell_control <- 2 #can change to 2,4

path <-  paste0("D:/data/SMART/results_ks/res",num_cell_control)

setwd(path)

res_SigRM_mle <- readRDS(paste0(path,"/res_SigRM_mle.rds"))
res_TRESS <- readRDS(paste0(path,"/res_TRESS.rds"))


# only see the sites need has the results
# if the site without any read in the test or control 
# we do not need to give a result



index_nona <- list()
set.seed(1)
for (i in 1: 20){
  index_nona[[i]] <- which(!is.na(res_SigRM_mle[[5]][,i]))
}




res_TRESS_ks <- c()

ks_function <- function(x,y){
  
  ks_test <- ks.test(x,y)
  l1 <- length(ks_test$data$x)
  l2 <- length(ks_test$data$y)
  D <- ks_test$statistic
  ks_staitics <- D/sqrt(1/l1+1/l2)
  
  return(ks_staitics)
  
}


ks_list_function <- function (x,index_x,y,index_y){
  x <- Map(function(x, index_x) x[index_x], x, index_x)
  y <- Map(function(y, index_y) y[index_y], y, index_y)

  l_y <- length(y)
  ks_list <- list()
  for (i in 1:l_y){
    a <- lapply(x,function(x) ks_function(x,y[[i]]))
    ks_list <- c(ks_list,a)
  }
  return(ks_list)
}


    
  res_TRESS_ks_pvalue <- lapply(res_TRESS,function(df) df[["pvalue"]])

  
  res_TRESS_ks <- ks_list_function(res_TRESS_ks_pvalue[1:10],index_nona[1:  10],res_TRESS_ks_pvalue[11:20],index_nona[11:20])



setwd(path)
saveRDS(res_TRESS_ks,"res_TRESS_ks.rds")

```






### ks test sim
```{r,eval=FALSE}
rm(list=ls())
num_cell_control <- 2 #can change to 2,4

path <-  paste0("D:/data/SMART/results_ks/res",num_cell_control,"sim")

res_SigRM_mle <- readRDS(paste0(path,"/res_SigRM_mle.rds"))
res_TRESS <- readRDS(paste0(path,"/res_TRESS.rds"))



# only see the sites need has the results
# if the site without any read in the test or control 
# we do not need to give a result

# the following is the index of site need have results, which are the results SigRM report

index_nona <- list()
set.seed(1)
for (i in 1: 20){
  index_nona[[i]] <- which(!is.na(res_SigRM_mle[[i]][[5]]))
}


res_TRESS_ks <- c()

ks_function <- function(x,y){
  
  ks_test <- ks.test(x,y)
  l1 <- length(ks_test$data$x)
  l2 <- length(ks_test$data$y)
  D <- ks_test$statistic
  ks_staitics <- D/sqrt(1/l1+1/l2)
  
  return(ks_staitics)
  
}


ks_list_function <- function (x,index_x,y,index_y){
  x <- Map(function(x, index_x) x[index_x], x, index_x)
  y <- Map(function(y, index_y) y[index_y], y, index_y)

  l_y <- length(y)
  ks_list <- list()
  for (i in 1:l_y){
    a <- lapply(x,function(x,index_x) ks_function(x,y[[i]]))
    ks_list <- c(ks_list,a)
  }
  return(ks_list)
}




   
res_TRESS_ks_pvalue <- lapply(res_TRESS,function(df) df[["pvalue"]])

  
  res_TRESS_ks <- ks_list_function(res_TRESS_ks_pvalue[1:10],index_nona[1:  10],res_TRESS_ks_pvalue[11:20],index_nona[11:20])


setwd(path)
saveRDS(res_TRESS_ks,"res_TRESS_ks.rds")
```





## plot the comparison of the different methods

### figure 2 real data 
```{r,eval=FALSE}
kstest_data_random_2 <- readRDS("D:/data/SMART/results_ks/res2/kstest_data_origin.rds")
kstest_data_random_4 <- readRDS("D:/data/SMART/results_ks/res4/kstest_data_origin.rds")
kstest_data_random_8 <- readRDS("D:/data/SMART/results_ks/res8/kstest_data_origin.rds")
kstest_data_random_16 <- readRDS("D:/data/SMART/results_ks/res8/kstest_data_origin.rds")
kstest_data_sim_2 <- readRDS("D:/data/SMART/results_ks/res2sim/kstest_data_origin.rds")
kstest_data_sim_4 <- readRDS("D:/data/SMART/results_ks/res4sim/kstest_data_origin.rds")
kstest_data_sim_8 <- readRDS("D:/data/SMART/results_ks/res8sim/kstest_data_origin.rds")
kstest_data_sim_16 <- readRDS("D:/data/SMART/results_ks/res16sim/kstest_data_origin.rds")

kstest_data_combine <- rbind(kstest_data_sim_2[1:100,],kstest_data_random_2[301:600,],kstest_data_sim_4[1:100,],kstest_data_random_4[301:600,],kstest_data_sim_8[1:100,],kstest_data_random_8[301:600,],kstest_data_sim_16[1:100,],kstest_data_random_16[301:600,])

kstest_data_combine$num <- c(rep("sample number = 2",400),rep("sample number = 4",400),rep("sample number = 8",400),rep("sample number = 16",400))
kstest_data_combine$num <-as.factor(kstest_data_combine$num)

kstest_data_combine$method[kstest_data_combine$method=="SigRM_locfit"] <- "SigRM"
kstest_data_combine$method[kstest_data_combine$method=="scDART"] <- "Bullseye"

kstest_data_combine$ks <- as.numeric(kstest_data_combine$ks)

kstest_data_combine$num <- factor(kstest_data_combine$num,levels = c("sample number = 2","sample number = 4","sample number = 8","sample number = 16"))

library(ggh4x)

kstest_data_combine$method <- factor(kstest_data_combine$method,levels =c("Bullseye","edgeR","DESeq2","SigRM"))

plot_real_ran_sim <- ggplot() +  geom_boxplot(data=kstest_data_combine,aes(y=log2(ks+1),x=method,color = method),outlier.shape = NA)+
  facet_wrap( ~ num, scales = "free_y",nrow = 1) +
scale_y_continuous(expand = c(0, 0))+ facetted_pos_scales(
  y = list(num == "sample number = 2" ~ scale_y_continuous(limits=c(2.8,7.8)),
           num == "sample number = 4" ~ scale_y_continuous(limits=c(2.8,7.8)),
           num == "sample number = 8" ~ scale_y_continuous(limits=c(2.8,7.8)),
           num == "sample number = 16" ~ scale_y_continuous(limits=c(2.8,7.8))))+guides(x = guide_axis(angle = -45))+ylab("Performance")+xlab("Method")
#

plot_real_ran_sim

```





### figure 3 compare random and sim in SigRM 
```{r,eval=FALSE}



kstest_data_random_2 <- readRDS("D:/data/SMART/results_ks/res2/kstest_data_origin.rds")
kstest_data_random_4 <- readRDS("D:/data/SMART/results_ks/res4/kstest_data_origin.rds")
kstest_data_random_8 <- readRDS("D:/data/SMART/results_ks/res8/kstest_data_origin.rds")
kstest_data_random_16 <- readRDS("D:/data/SMART/results_ks/res8/kstest_data_origin.rds")

kstest_data_random <- rbind(kstest_data_random_2,kstest_data_random_4,kstest_data_random_8,kstest_data_random_16)

kstest_data_random$num <- c(rep(2,600),rep(4,600),rep(8,600),rep(16,600))
kstest_data_random$num <-as.factor(kstest_data_random$num)


kstest_data_sim_2 <- readRDS("D:/data/SMART/results_ks/res2sim/kstest_data_origin.rds")
kstest_data_sim_4 <- readRDS("D:/data/SMART/results_ks/res4sim/kstest_data_origin.rds")
kstest_data_sim_8 <- readRDS("D:/data/SMART/results_ks/res8sim/kstest_data_origin.rds")
kstest_data_sim_16 <- readRDS("D:/data/SMART/results_ks/res16sim/kstest_data_origin.rds")


kstest_data_sim <- rbind(kstest_data_sim_2,kstest_data_sim_4,kstest_data_sim_8,kstest_data_sim_16)

kstest_data_sim$num <- c(rep(2,600),rep(4,600),rep(8,600),rep(16,600))
kstest_data_sim$num <-as.factor(kstest_data_sim$num)


kstest_data <- rbind(kstest_data_random,kstest_data_sim)

kstest_data $sample <- c(rep("random",2400),rep("similar",2400))
kstest_data $sample <- as.factor(kstest_data $sample)
library(dplyr)



kstest_data_SigRM_locfit <- kstest_data%>% filter(method == "SigRM_locfit")

kstest_data_SigRM_unbiased <- kstest_data%>% filter(method == "SigRM_unbiased")


kstest_data_SigRM_mle <- kstest_data%>% filter(method == "SigRM_mle")

kstest_data_DESeq2 <- kstest_data%>% filter(method == "DESeq2")

kstest_data_edgeR <- kstest_data%>% filter(method == "edgeR")

kstest_data_scDART <- kstest_data%>% filter(method == "scDART")
kstest_data_scDART$method <- "Bullseye"


kstest_data_SigRM <- as.data.frame(rbind(kstest_data_SigRM_locfit,
                 kstest_data_SigRM_unbiased,kstest_data_SigRM_mle))

kstest_data_others <- as.data.frame(rbind(kstest_data_DESeq2,
                 kstest_data_edgeR,kstest_data_scDART))

library(ggplot2)

kstest_data_SigRM$method[kstest_data_SigRM$method=="SigRM_locfit"] ="SigRM_locfit (default)"
plot_compare_ran_sim_SigRM <- ggplot(kstest_data_SigRM, aes( y=log2(ks+1),color=sample,x=num)) +
  geom_boxplot(outlier.shape = NA)+facet_wrap(vars(method), scales = "free")+ylab("Performance")+xlab("Sample Number")

plot_compare_ran_sim_SigRM




```
### figure 4 correlation plot
```{r,eval=FALSE}


res_SigRM_locfit <- readRDS("D:/data/SMART/results_ks/res4sim/res_SigRM_locfit.rds")
res <- rbind(res_SigRM_locfit[[1]],res_SigRM_locfit[[2]],res_SigRM_locfit[[3]],res_SigRM_locfit[[4]],res_SigRM_locfit[[5]],
             res_SigRM_locfit[[6]],res_SigRM_locfit[[7]],res_SigRM_locfit[[8]],res_SigRM_locfit[[9]],res_SigRM_locfit[[10]])


my_data <- cbind(as.numeric(log2(res[[6]]+1)),res[[5]],res[[3]],res[[4]],res[[8]])
colnames(my_data) <- c("expression","p-value","RR","OR","TCR")
my_data <- as.data.frame(my_data)
my_data$beta <- as.numeric(res[[1]])
my_data <- my_data[which(!is.na(rowSums(my_data))),]
my_data$M <- log2((my_data$beta+0.001)/(1-my_data$beta+0.001))

set.seed(10)
my_data_ <-my_data [sample(1:dim(my_data)[1],1000),]




library("psych")

pairs.panels(my_data_, ellipses=TRUE,scale = FALSE,hist.col = "#00AFBB",
             stars=FALSE,method = "spearman",digits=3,factor=2,alpha=0.5)
```



## Appendix



### figure 1

```{r,eval=FALSE}

load("D:/data/SMART/results_ks/res2sim/runtime.RData")
runtime_2_sim <- runtime
load("D:/data/SMART/results_ks/res4sim/runtime.RData")
runtime_4_sim <- runtime
load("D:/data/SMART/results_ks/res8sim/runtime.RData")
runtime_8_sim <- runtime
load("D:/data/SMART/results_ks/res16sim/runtime.RData")
runtime_16_sim <- runtime

runtime_2_sim <- unlist(lapply(runtime_2_sim,function(x) as.numeric(x, units = "secs")))
runtime_4_sim <- unlist(lapply(runtime_4_sim,function(x) as.numeric(x, units = "secs")))
runtime_8_sim <- unlist(lapply(runtime_8_sim,function(x) as.numeric(x, units = "secs")))
runtime_16_sim <- unlist(lapply(runtime_16_sim,function(x) as.numeric(x, units = "secs")))

run_time_TRESS_2_sim <- readRDS("D:/data/SMART/results_ks/res2sim/run_time_TRESS.rds")
run_time_TRESS_4_sim <- readRDS("D:/data/SMART/results_ks/res4sim/run_time_TRESS.rds")


run_time_TRESS_2_sim <-  as.numeric(run_time_TRESS_2_sim, units = "secs")
run_time_TRESS_4_sim <-  as.numeric(run_time_TRESS_4_sim, units = "secs")
  
run_time <- data.frame(cbind(c(runtime_2_sim[c(1:4)],run_time_TRESS_2_sim,runtime_4_sim[c(1:4)],run_time_TRESS_4_sim,runtime_8_sim[c(1:4)],NA,runtime_16_sim[c(1:4)],NA),rep(c("Bullseye","DESeq2","edgeR","SigRM","TRESS"),4)))


run_time <- cbind(run_time, c(rep("sample number = 2",5),rep("sample number = 4",5),rep("sample number = 8",5),rep("sample number = 16",5)))
colnames(run_time) <- c("running time", "method","num")
run_time$num <- as.ordered(run_time$num)
rownames(run_time) <- NULL

library(ggh4x)

run_time$method <- factor(run_time$method,levels =c("TRESS","Bullseye","edgeR","DESeq2","SigRM"))
run_time$`running time` <- as.numeric(run_time$`running time`)
run_time$num <- factor(run_time$num,levels = c("sample number = 2","sample number = 4","sample number = 8","sample number = 16"))

#run_time$num <- as.numeric(run_time$num)

plot_app1 <- ggplot(data=run_time,aes(y=log(`running time`,2),x=method,color = method,fill = method)) +  geom_bar(stat = "identity")+
  facet_wrap( ~ num, nrow = 1, scales = "fixed")+guides(x = guide_axis(angle = -45))+ylab("log (Running time) ")

plot_app1
```



### figure 2
```{r,eval=FALSE}

kstest_data_random_2 <- readRDS("D:/data/SMART/results_ks/res2/kstest_data_origin.rds")
kstest_data_random_4 <- readRDS("D:/data/SMART/results_ks/res4/kstest_data_origin.rds")
kstest_data_random_8 <- readRDS("D:/data/SMART/results_ks/res8/kstest_data_origin.rds")
kstest_data_random_16 <- readRDS("D:/data/SMART/results_ks/res16/kstest_data_origin.rds")
kstest_data_sim_2 <- readRDS("D:/data/SMART/results_ks/res2sim/kstest_data_origin.rds")
kstest_data_sim_4 <- readRDS("D:/data/SMART/results_ks/res4sim/kstest_data_origin.rds")
kstest_data_sim_8 <- readRDS("D:/data/SMART/results_ks/res8sim/kstest_data_origin.rds")
kstest_data_sim_16 <- readRDS("D:/data/SMART/results_ks/res16sim/kstest_data_origin.rds")

# in origin data without TRESS, we need add TRESS

res_TRESS_ks_2 <- readRDS("D:/data/SMART/results_ks/res2/res_TRESS_ks.rds")
res_TRESS_ks_4 <- readRDS("D:/data/SMART/results_ks/res4/res_TRESS_ks.rds")
res_TRESS_ks_2_sim <- readRDS("D:/data/SMART/results_ks/res2sim/res_TRESS_ks.rds")
res_TRESS_ks_4_sim <- readRDS("D:/data/SMART/results_ks/res4sim/res_TRESS_ks.rds")
res_TRESS_ks_2 <- cbind(res_TRESS_ks_2,rep("TRESS",100))
res_TRESS_ks_4 <- cbind(res_TRESS_ks_4,rep("TRESS",100))
res_TRESS_ks_8 <- cbind(as.numeric(rep(0,100)),rep("TRESS",100))
res_TRESS_ks_16 <- cbind(as.numeric(rep(0,100)),rep("TRESS",100))
colnames(res_TRESS_ks_16) <- colnames(res_TRESS_ks_8) <-colnames(res_TRESS_ks_4) <-colnames(res_TRESS_ks_2) <-colnames(kstest_data_random_2) 

kstest_data_random_2 <- rbind(kstest_data_random_2,res_TRESS_ks_2 )
kstest_data_random_4 <- rbind(kstest_data_random_4,res_TRESS_ks_4 )
kstest_data_random_8 <- rbind(kstest_data_random_8,res_TRESS_ks_8 )
kstest_data_random_16 <- rbind(kstest_data_random_16,res_TRESS_ks_16 )

res_TRESS_ks_2_sim <- cbind(res_TRESS_ks_2_sim,rep("TRESS",100))
res_TRESS_ks_4_sim <- cbind(res_TRESS_ks_4_sim,rep("TRESS",100))
res_TRESS_ks_8_sim <- cbind(as.numeric(rep(NA,100)),rep("TRESS",100))
res_TRESS_ks_16_sim <- cbind(as.numeric(rep(NA,100)),rep("TRESS",100))
colnames(res_TRESS_ks_16_sim) <- colnames(res_TRESS_ks_8_sim) <-colnames(res_TRESS_ks_4_sim) <-colnames(res_TRESS_ks_2_sim) <-colnames(kstest_data_sim_2) 

kstest_data_sim_2 <- rbind(kstest_data_sim_2,res_TRESS_ks_2_sim )
kstest_data_sim_4 <- rbind(kstest_data_sim_4,res_TRESS_ks_4_sim )
kstest_data_sim_8 <- rbind(kstest_data_sim_8,res_TRESS_ks_8_sim )
kstest_data_sim_16 <- rbind(kstest_data_sim_16,res_TRESS_ks_16_sim )


kstest_data_combine <- rbind(kstest_data_random_2[1:100,],kstest_data_random_2[301:700,],kstest_data_random_4[1:100,],kstest_data_random_4[301:700,],kstest_data_random_8[1:100,],kstest_data_random_8[301:700,],kstest_data_random_16[1:100,],kstest_data_random_16[301:700,],kstest_data_sim_2[1:100,],kstest_data_sim_2[301:700,],kstest_data_sim_4[1:100,],kstest_data_sim_4[301:700,],kstest_data_sim_8[1:100,],kstest_data_sim_8[301:700,],kstest_data_sim_16[1:100,],kstest_data_sim_16[301:700,])

kstest_data_combine$num <- rep(c(rep(2,500),rep(4,500),rep(8,500),rep(16,500)),2)
kstest_data_combine$num <-as.factor(kstest_data_combine$num)

kstest_data_combine$control <- c(rep("Random",2000),rep("Similar",2000))
kstest_data_combine$control <-as.factor(kstest_data_combine$control)

kstest_data_combine$method[kstest_data_combine$method=="SigRM_locfit"] <- "SigRM"
kstest_data_combine$method[kstest_data_combine$method=="scDART"] <- "Bullseye"

kstest_data_combine$ks <- as.numeric(kstest_data_combine$ks)

library(ggh4x)

kstest_data_combine$method <- factor(kstest_data_combine$method,levels =c("TRESS","Bullseye","edgeR","DESeq2","SigRM"))

plot_app2 <- ggplot() +  geom_boxplot(data=kstest_data_combine,aes(y=log2(ks+1),x=num,color = method),outlier.shape = NA)+
  facet_wrap( ~ control, scales = "free_y") +
scale_y_continuous(expand = c(0, 0))+ facetted_pos_scales(
  y = list(control == "Random" ~ scale_y_continuous(limits=c(2.8,7.8)),
           control == "Similar" ~ scale_y_continuous(limits=c(2.8,7.8))))+ylab("Performance")+xlab("Sample Number")
#

plot_app2
```

# pesudo time inference
## SIgRM
## load_data
```{r,eval=FALSE}
library(SigRM)
expression_TPM <- readRDS("~/SIGMR_single_cell/data/SMART/preprocess2/expression_TPM.rds")
expression_log_TPM <- log(expression_TPM +1) 
expression_log_TPM_control <- expression_log_TPM[,1:391]

 frequency_all_processed <- readRDS("~/SIGMR_single_cell/data/SMART/preprocess2/frequency_all_processed.rds")
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")
meth_control <- meth[,1:391]

unmeth_control <- unmeth[,1:391]


```

## 0-200
```{r,eval=FALSE}
setwd("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim")
res_list_100 <- list()
start_time <- Sys.time()
for (i in 1: 100){
  unmeth_test <- unmeth[,391+i]
  meth_test <- meth[,391+i]
  expression_log_TPM_test <- expression_log_TPM[,391+i]
  expression <- cbind(expression_log_TPM_control,expression_log_TPM_test)
  res_list_100 [[i]] <- SigRM_similarity_test(meth_control,meth_test,unmeth_control,unmeth_test,expression = expression,num_control = 4,method_dispersion = "locfit"  )
}
end_time <- Sys.time()
end_time-start_time
saveRDS(res_list_100,"res_list_100.rds")



res_list_200 <- list()
start_time <- Sys.time()
for (i in 101: 200){
  unmeth_test <- unmeth[,391+i]
  meth_test <- meth[,391+i]
  expression_log_TPM_test <- expression_log_TPM[,391+i]
  expression <- cbind(expression_log_TPM_control,expression_log_TPM_test)
  res_list_200 [[i]] <- SigRM_similarity_test(meth_control,meth_test,unmeth_control,unmeth_test,expression = expression,num_control = 4,method_dispersion = "locfit"  )
}
end_time <- Sys.time()
end_time-start_time
saveRDS(res_list_200,"res_list_200.rds")
```


## 200-400

```{r,eval=FALSE}
setwd("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim")
res_list_300 <- list()
start_time <- Sys.time()
for (i in 201: 300){
  unmeth_test <- unmeth[,391+i]
  meth_test <- meth[,391+i]
  expression_log_TPM_test <- expression_log_TPM[,391+i]
  expression <- cbind(expression_log_TPM_control,expression_log_TPM_test)
  res_list_300 [[i]] <- SigRM_similarity_test(meth_control,meth_test,unmeth_control,unmeth_test,expression = expression,num_control = 4,method_dispersion = "locfit"  )
}
end_time <- Sys.time()
end_time-start_time
saveRDS(res_list_300,"res_list_300.rds")



res_list_400 <- list()
start_time <- Sys.time()
for (i in 301: 400){
  unmeth_test <- unmeth[,391+i]
  meth_test <- meth[,391+i]
  expression_log_TPM_test <- expression_log_TPM[,391+i]
  expression <- cbind(expression_log_TPM_control,expression_log_TPM_test)
  res_list_400 [[i]] <- SigRM_similarity_test(meth_control,meth_test,unmeth_control,unmeth_test,expression = expression,num_control = 4,method_dispersion = "locfit"  )
}
end_time <- Sys.time()
end_time-start_time
saveRDS(res_list_400,"res_list_400.rds")
```
## 400-600



```{r,eval=FALSE}
setwd("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim")
res_list_500 <- list()
start_time <- Sys.time()
for (i in 401: 500){
  unmeth_test <- unmeth[,391+i]
  meth_test <- meth[,391+i]
  expression_log_TPM_test <- expression_log_TPM[,391+i]
  expression <- cbind(expression_log_TPM_control,expression_log_TPM_test)
  res_list_500 [[i]] <- SigRM_similarity_test(meth_control,meth_test,unmeth_control,unmeth_test,expression = expression,num_control = 4,method_dispersion = "locfit"  )
}
end_time <- Sys.time()
end_time-start_time
saveRDS(res_list_500,"res_list_500.rds")



res_list_600 <- list()
start_time <- Sys.time()
for (i in 501: 600){
  unmeth_test <- unmeth[,391+i]
  meth_test <- meth[,391+i]
  expression_log_TPM_test <- expression_log_TPM[,391+i]
  expression <- cbind(expression_log_TPM_control,expression_log_TPM_test)
  res_list_600 [[i]] <- SigRM_similarity_test(meth_control,meth_test,unmeth_control,unmeth_test,expression = expression,num_control = 4,method_dispersion = "locfit"  )
}
end_time <- Sys.time()
end_time-start_time
saveRDS(res_list_600,"res_list_600.rds")
```


## 600-800

```{r,eval=FALSE}
setwd("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim")
res_list_700 <- list()
start_time <- Sys.time()
for (i in 601: 700){
  unmeth_test <- unmeth[,391+i]
  meth_test <- meth[,391+i]
  expression_log_TPM_test <- expression_log_TPM[,391+i]
  expression <- cbind(expression_log_TPM_control,expression_log_TPM_test)
  res_list_700 [[i]] <- SigRM_similarity_test(meth_control,meth_test,unmeth_control,unmeth_test,expression = expression,num_control = 4,method_dispersion = "locfit"  )
}
end_time <- Sys.time()
end_time-start_time
saveRDS(res_list_700,"res_list_700.rds")



res_list_800 <- list()
start_time <- Sys.time()
for (i in 701: 800){
  unmeth_test <- unmeth[,391+i]
  meth_test <- meth[,391+i]
  expression_log_TPM_test <- expression_log_TPM[,391+i]
  expression <- cbind(expression_log_TPM_control,expression_log_TPM_test)
  res_list_800 [[i]] <- SigRM_similarity_test(meth_control,meth_test,unmeth_control,unmeth_test,expression = expression,num_control = 4 )
}
end_time <- Sys.time()
end_time-start_time
saveRDS(res_list_800,"res_list_800.rds")
```


## 800-991

```{r,eval=FALSE}
setwd("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim")
res_list_900 <- list()
start_time <- Sys.time()
for (i in 801: 900){
  unmeth_test <- unmeth[,391+i]
  meth_test <- meth[,391+i]
  expression_log_TPM_test <- expression_log_TPM[,391+i]
  expression <- cbind(expression_log_TPM_control,expression_log_TPM_test)
  res_list_900 [[i]] <- SigRM_similarity_test(meth_control,meth_test,unmeth_control,unmeth_test,expression = expression,num_control = 4,method_dispersion = "locfit"  )
}
end_time <- Sys.time()
end_time-start_time
saveRDS(res_list_900,"res_list_900.rds")



res_list_991 <- list()
start_time <- Sys.time()
for (i in 901: 991){
  unmeth_test <- unmeth[,391+i]
  meth_test <- meth[,391+i]
  expression_log_TPM_test <- expression_log_TPM[,391+i]
  expression <- cbind(expression_log_TPM_control,expression_log_TPM_test)
  res_list_991 [[i]] <- SigRM_similarity_test(meth_control,meth_test,unmeth_control,unmeth_test,expression = expression,num_control = 4,method_dispersion = "locfit"  )
}
end_time <- Sys.time()
end_time-start_time
saveRDS(res_list_991,"res_list_991.rds")
```


## combine
```{r,eval=FALSE}
res_list_100 <- readRDS("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim/res_list_100.rds")
res_list_200 <- readRDS("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim/res_list_200.rds")
res_list_300 <- readRDS("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim/res_list_300.rds")
res_list_400 <- readRDS("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim/res_list_400.rds")
res_list_500 <- readRDS("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim/res_list_500.rds")
res_list_600 <- readRDS("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim/res_list_600.rds")
res_list_700 <- readRDS("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim/res_list_700.rds")
res_list_800 <- readRDS("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim/res_list_800.rds")
res_list_900 <- readRDS("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim/res_list_900.rds")
res_list_991 <- readRDS("~/SIGMR_single_cell/data/SMART/results_ks/res_all_4_sim/res_list_991.rds")



res_list <- list()
res_list[1:100] <- res_list_100
res_list[101:200] <- res_list_200[101:200]
res_list[201:300] <- res_list_300[201:300]
res_list[301:400] <- res_list_400[301:400]
res_list[401:500] <- res_list_500[401:500]
res_list[501:600] <- res_list_600[501:600]
res_list[601:700] <- res_list_700[601:700]
res_list[701:800] <- res_list_800[701:800]
res_list[801:900] <- res_list_900[801:900]
res_list[901:991] <- res_list_991[901:991]

saveRDS(res_list,"res_list.rds")
```

## edgeR
## load_data
```{r,eval=FALSE}
library("edgeR")
edgeR_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  num <- dim(meth_control)[2]
  
  res <- list()
  
  for (i in 1: dim(meth_test)[2]){
       # create deglist
    countdata <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
   
    dgelist <- DGEList(counts = countdata)
    
    
    countdata_all <- cbind((meth_control+unmeth_control),(meth_test[,i]+unmeth_test[,i]))
   
    dgelist_all <- DGEList(counts = countdata_all)
    
    
    # normalize
    dgelist_norm_all <- calcNormFactors(dgelist_all, method = 'TMM')
    
    dgelist_norm <- calcNormFactors(dgelist, method = 'TMM')
    
    dgelist_norm@.Data[[2]] <- rbind(dgelist_norm_all@.Data[[2]],dgelist_norm_all@.Data[[2]])
    
    rownames(dgelist_norm@.Data[[2]]) <- paste0("Sample",1:(2*(num+1)))
    
    
    group1 <- factor(rep(c(rep("control",num),"test"),2))
    group2 <- factor(c(rep("meth",(num+1)),rep("unmeth",(num+1))))
    
    
    
    design <- model.matrix(~0+group1+group2+group1:group2)
    colnames(design) <- c("control", "test", "unmeth", "unmeth_and_test")
    
    dge <- estimateDisp(dgelist_norm, design, robust = TRUE)
    
    fit <- glmFit(dge, design, robust = TRUE)
    

    lrt <- glmLRT(fit,coef=4)@.Data
    
    res[[i]] <- lrt[[14]]
  }
  
  return(res)
 
}

expression_TPM <- readRDS("~/data/SMART/preprocess2/expression_TPM.rds")
expression_log_TPM <- log(expression_TPM +1) 
expression_log_TPM_control <- expression_log_TPM[,1:391]

 frequency_all_processed <- readRDS("~/data/SMART/preprocess2/frequency_all_processed.rds")
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")
meth_control <- meth[,1:391]

unmeth_control <- unmeth[,1:391]



similarity_data <- cor(expression_log_TPM, method = "pearson")

library(doBy)
  similarity_index <- apply(similarity_data,2, function(x){which.maxn(x,
                                      (1382))})
  similarity_index <- as.matrix(similarity_index[,392:1382])
  list_create <- as.list(1: 991)
  
  list_function <- function(x){
    similarity_index_ <- similarity_index[which(similarity_index[,
                                  x]%in%c(1:391)),x]
    return(similarity_index_)
  }
  
  similarity_index <- lapply(list_create,list_function)

  
num_cell_control <- 4

control_index <- lapply(similarity_index,function(df) df[1:4])

```

## 0-200
```{r,eval=FALSE}
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
res_list_100 <- list()
start_time <- Sys.time()

for (i in 1:100){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_100[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
saveRDS(res_list_100,"res_list_100.rds")




res_list_200 <- list()
start_time <- Sys.time()

for (i in 101:200){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_200[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
saveRDS(res_list_200,"res_list_200.rds")


```


## 200-400

```{r,eval=FALSE}
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
res_list_300 <- list()
start_time <- Sys.time()

for (i in 201:300){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_300[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
saveRDS(res_list_300,"res_list_300.rds")




res_list_400 <- list()
start_time <- Sys.time()

for (i in 301:400){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_400[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
saveRDS(res_list_400,"res_list_400.rds")


```
## 400-600



```{r,eval=FALSE}
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
res_list_500 <- list()
start_time <- Sys.time()

for (i in 401:500){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_500[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
saveRDS(res_list_500,"res_list_500.rds")




res_list_600 <- list()
start_time <- Sys.time()

for (i in 501:600){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_600[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
saveRDS(res_list_600,"res_list_600.rds")


```


## 600-800

```{r,eval=FALSE}
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
res_list_700 <- list()
start_time <- Sys.time()

for (i in 601:700){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_700[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
saveRDS(res_list_700,"res_list_700.rds")




res_list_800 <- list()
start_time <- Sys.time()

for (i in 701:800){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_800[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
saveRDS(res_list_800,"res_list_800.rds")


```


## 800-991

```{r,eval=FALSE}
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
res_list_900 <- list()
start_time <- Sys.time()

for (i in 801:900){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_900[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
saveRDS(res_list_900,"res_list_900.rds")




res_list_991 <- list()
start_time <- Sys.time()

for (i in 901:991){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_991[[i]] <-  edgeR_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
saveRDS(res_list_991,"res_list_991.rds")
```


## combine
```{r,eval=FALSE}
res_list_100 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_edgeR/res_list_100.rds")
res_list_200 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_edgeR/res_list_200.rds")
res_list_300 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_edgeR/res_list_300.rds")
res_list_400 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_edgeR/res_list_400.rds")
res_list_500 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_edgeR/res_list_500.rds")
res_list_600 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_edgeR/res_list_600.rds")
res_list_700 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_edgeR/res_list_700.rds")
res_list_800 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_edgeR/res_list_800.rds")
res_list_900 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_edgeR/res_list_900.rds")
res_list_991 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_edgeR/res_list_991.rds")



res_list <- list()
res_list[1:100] <- res_list_100
res_list[101:200] <- res_list_200[101:200]
res_list[201:300] <- res_list_300[201:300]
res_list[301:400] <- res_list_400[301:400]
res_list[401:500] <- res_list_500[401:500]
res_list[501:600] <- res_list_600[501:600]
res_list[601:700] <- res_list_700[601:700]
res_list[701:800] <- res_list_800[701:800]
res_list[801:900] <- res_list_900[801:900]
res_list[901:991] <- res_list_991[901:991]
setwd("~/data/SMART/results_ks/res_all_4_sim_edgeR")
saveRDS(res_list,"res_list.rds")
```

## DESeq2
## load_data
```{r,eval=FALSE}

library(DESeq2)
DESeq2_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  meth_control <- data.frame(meth_control)
  meth_test <- data.frame(meth_test)
  unmeth_control <- data.frame(unmeth_control)
  unmeth_test <- data.frame(unmeth_test)

  num <- dim(meth_control)[2]
  
  sizeFactor <- function(data) {
  # make the elements no smaller than 0
  data <- as.matrix(data)
  
  # first log
  log_data <- log(data)
  log_data[is.infinite(log_data)] <- NA
  log_mean <- rowMeans(log_data)
  log_s <- log_data-log_mean
  
  # then exp
  s_size <- exp(apply(log_s,2,function(x)median(x,na.rm=TRUE)))
  return(s_size)
}

  
  res <- list()
  for (i in 1:dim(meth_test)[2]){
    counts <- cbind(meth_control,meth_test[,i],unmeth_control,unmeth_test[,i])
    design = data.frame(
                       Trt = rep(c(rep("control", num),"test"),2), 
                    Meth = c(rep("meth",(num+1)),rep("unmeth", 
                            (num+1))))
    model = ~ Meth + Trt + Meth:Trt
    # 
    dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = design,
                                design = model)
    sf = sizeFactor(cbind((meth_control+unmeth_control),
                       (meth_test[,i]+unmeth_test[,i])))
    
    sizeFactors(dds) = rep(sf,2)
    dds <- DESeq(dds, test = "Wald")
    
    res[[i]] = DESeq2::results(dds, name = "Methunmeth.Trttest") 
  }
    

 
  
  return(res)
}


expression_TPM <- readRDS("~/data/SMART/preprocess2/expression_TPM.rds")
expression_log_TPM <- log(expression_TPM +1) 
expression_log_TPM_control <- expression_log_TPM[,1:391]

 frequency_all_processed <- readRDS("~/data/SMART/preprocess2/frequency_all_processed.rds")
meth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,1]))
unmeth <- as.data.frame( lapply(frequency_all_processed,function(df) df[,2]))
colnames(meth) <- paste("meth_",names(frequency_all_processed),sep="")
colnames(unmeth) <- paste("unmeth_",names(frequency_all_processed),sep="")
meth_control <- meth[,1:391]

unmeth_control <- unmeth[,1:391]




similarity_data <- cor(expression_log_TPM, method = "pearson")

library(doBy)
  similarity_index <- apply(similarity_data,2, function(x){which.maxn(x,
                                      (1382))})
  similarity_index <- as.matrix(similarity_index[,392:1382])
  list_create <- as.list(1: 991)
  
  list_function <- function(x){
    similarity_index_ <- similarity_index[which(similarity_index[,
                                  x]%in%c(1:391)),x]
    return(similarity_index_)
  }
  
  similarity_index <- lapply(list_create,list_function)

  
num_cell_control <- 4

control_index <- lapply(similarity_index,function(df) df[1:4])


```

## 0-200
```{r,eval=FALSE}
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
res_list_100 <- list()
start_time <- Sys.time()

for (i in 1:100){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_100[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
saveRDS(res_list_100,"res_list_100.rds")




res_list_200 <- list()
start_time <- Sys.time()

for (i in 101:200){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_200[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
saveRDS(res_list_200,"res_list_200.rds")


```


## 200-400

```{r,eval=FALSE}
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
res_list_300 <- list()
start_time <- Sys.time()

for (i in 201:300){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_300[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
saveRDS(res_list_300,"res_list_300.rds")




res_list_400 <- list()
start_time <- Sys.time()

for (i in 301:400){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_400[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
saveRDS(res_list_400,"res_list_400.rds")


```
## 400-600



```{r,eval=FALSE}
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
res_list_500 <- list()
start_time <- Sys.time()

for (i in 401:500){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_500[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
saveRDS(res_list_500,"res_list_500.rds")




res_list_600 <- list()
start_time <- Sys.time()

for (i in 501:600){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_600[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
saveRDS(res_list_600,"res_list_600.rds")


```


## 600-800

```{r,eval=FALSE}
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
res_list_700 <- list()
start_time <- Sys.time()

for (i in 601:700){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_700[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
saveRDS(res_list_700,"res_list_700.rds")




res_list_800 <- list()
start_time <- Sys.time()

for (i in 701:800){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_800[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
saveRDS(res_list_800,"res_list_800.rds")


```


## 800-991

```{r,eval=FALSE}
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
res_list_900 <- list()
start_time <- Sys.time()

for (i in 801:900){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_900[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
saveRDS(res_list_900,"res_list_900.rds")




res_list_991 <- list()
start_time <- Sys.time()

for (i in 901:991){
  meth_control <- meth[,control_index[[i]]]
  meth_test <- data.frame(meth[,391+i])
  unmeth_control <- unmeth[,control_index[[i]]]
  unmeth_test <- data.frame(unmeth[,391+i])
  
  res_list_991[[i]] <-  DESeq2_test(meth_control,meth_test,unmeth_control,unmeth_test)
  
}

end_time <- Sys.time()
end_time-start_time
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
saveRDS(res_list_991,"res_list_991.rds")
```


## combine
```{r,eval=FALSE}
res_list_100 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_DESeq2/res_list_100.rds")
res_list_200 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_DESeq2/res_list_200.rds")
res_list_300 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_DESeq2/res_list_300.rds")
res_list_400 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_DESeq2/res_list_400.rds")
res_list_500 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_DESeq2/res_list_500.rds")
res_list_600 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_DESeq2/res_list_600.rds")
res_list_700 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_DESeq2/res_list_700.rds")
res_list_800 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_DESeq2/res_list_800.rds")
res_list_900 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_DESeq2/res_list_900.rds")
res_list_991 <- readRDS("~/data/SMART/results_ks/res_all_4_sim_DESeq2/res_list_991.rds")



res_list <- list()
res_list[1:100] <- res_list_100
res_list[101:200] <- res_list_200[101:200]
res_list[201:300] <- res_list_300[201:300]
res_list[301:400] <- res_list_400[301:400]
res_list[401:500] <- res_list_500[401:500]
res_list[501:600] <- res_list_600[501:600]
res_list[601:700] <- res_list_700[601:700]
res_list[701:800] <- res_list_800[701:800]
res_list[801:900] <- res_list_900[801:900]
res_list[901:991] <- res_list_991[901:991]
setwd("~/data/SMART/results_ks/res_all_4_sim_DESeq2")
saveRDS(res_list,"res_list.rds")
```



## visulaization
```{r,eval=FALSE}



res_SigRM_locfit <- readRDS("D:/data/SMART/results_ks/res4sim/res_SigRM_locfit.rds")
res <- rbind(res_SigRM_locfit[[1]],res_SigRM_locfit[[2]],res_SigRM_locfit[[3]],res_SigRM_locfit[[4]],res_SigRM_locfit[[5]],
             res_SigRM_locfit[[6]],res_SigRM_locfit[[7]],res_SigRM_locfit[[8]],res_SigRM_locfit[[9]],res_SigRM_locfit[[10]])


my_data <- cbind(as.numeric(log2(res[[6]]+1)),res[[5]],res[[3]],res[[4]],res[[8]])
colnames(my_data) <- c("expression","p-value","RR","OR","TCR")
my_data <- as.data.frame(my_data)
my_data$beta <- as.numeric(res[[1]])
my_data <- my_data[which(!is.na(rowSums(my_data))),]
my_data$M <- log2((my_data$beta+0.001)/(1-my_data$beta+0.001))

set.seed(10)
my_data_ <-my_data [sample(1:dim(my_data)[1],1000),]




library("psych")

pairs.panels(my_data_, ellipses=TRUE,scale = FALSE,hist.col = "#00AFBB",
             stars=FALSE,method = "spearman",digits=3,factor=2,alpha=0.5)
```


## validation
```{r,eval=FALSE}
library(latex2exp)
library(ggplot2)
original_data <- readRDS("D:/data/SMART/preprocess2/expression_TPM.rds")
original_data <- original_data [,-c(1:391)]

g1_s_genes <- c("CCNE1", "CCNE2","CDC25A","CDC6","E2F1","MCM2","MCM6","NPAT","PCNA","SLBP","SMARCD3","MAX", "ULK4")  # 选基因
s_genes <- c("BRCA1","RRM1","RRM2", "TYMS","MCM5", "DTL") 
g2_m_genes <- c("CCNA2","CCNF", "TOP2A","BIRC5","BUB1","BUB1B", "CCNB1", "CCNB2", "CDC20", "CDC25B","CDKN2D","CENPA","CKS2","PLK1","AURKA")  

## step2 : change to ensebl

library(org.Hs.eg.db)
library(clusterProfiler)
library(stringi)

g1_s_genes_ENSG <- bitr(g1_s_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")
s_genes_ENSG <- bitr(s_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")
g2_m_genes_ENSG <-bitr(g2_m_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")



#choose the selected genes from original data
g1_matrix <- original_data[g1_s_genes_ENSG$ENSEMBL,]
s_martrix <- original_data[s_genes_ENSG$ENSEMBL,]
g2_martrix <- original_data[g2_m_genes_ENSG$ENSEMBL,] 
#combine them together
g1ands <- rbind(g1_matrix,s_martrix)
select_TPM <- rbind(g1ands,g2_martrix)

select_TPM <- apply(select_TPM,2,as.numeric)


select_TPM <- select_TPM [c(7,9,10,16,18,19,22,26,34),]
# step 1: rescale
exp_log2 = log2(select_TPM + 1)


# step 2: calculate some rescaled phase likelihood based on a specific gene marker
for (s in 1:991) {
  exp_log2[, s] = (exp_log2[, s] - mean(exp_log2[, s] ))/sd(exp_log2[, s])
}
GeneRatio <- exp_log2

a <-cor(t(GeneRatio))
index_g1 <- c(1:3)#1:length(g1_s_genes_ENSG$ENSEMBL)
index_s <- c(4:6)#(1+length(g1_s_genes_ENSG$ENSEMBL)):(length(g1_s_genes_ENSG$ENSEMBL)+length(s_genes_ENSG$ENSEMBL))
index_g2 <- c(7:9)#(length(g1_s_genes_ENSG$ENSEMBL)+length(s_genes_ENSG$ENSEMBL)+1):(length(g1_s_genes_ENSG$ENSEMBL)+length(s_genes_ENSG$ENSEMBL)+length(g2_m_genes_ENSG$ENSEMBL))


s1 <- apply(a[index_g1,],2,sum)-1
s2 <- apply(a[index_s,],2,sum)-1
s3 <- apply(a[index_g2,],2,sum)-1
s <- rbind(s1,s2,s3)

sum_g1 <- (colSums(a[index_g1,])-1) /length(index_g1)
sum_s <- (colSums(a[index_s,]) -1) /length(index_s)
sum_g2 <- (colSums(a[index_g2,]) -1) /length(index_g2)

score_m <- rbind(sum_g1,sum_s,sum_g2)



# step 3: merge the likelihood from all marker genes of a phase
# For example, phase M1 with marker gene index marker1 = 1:3, M2 index marker2 = 4:6, M3 index marker3 = 7:9
GeneRatio <- t(apply(GeneRatio,1,function(x) (x-mean(x))/sd(x)))

M1_score <- colMeans(GeneRatio[index_g1,]) # "CDC25A" "NPAT" "MAX"
M2_score <- colMeans(GeneRatio[index_s,]) # "RRM1" "RRM2" "MCM5" "DTL"
M3_score <- colMeans(GeneRatio[index_g2,]) # "TOP2A" "BUB1" "BUB1B" "CCNB1"  "CCNB2"  "CDC20" "CENPA" "PLK1" "AURKA" 

M1_score <- (M1_score-mean(M1_score))/sd(M1_score)
M2_score <- (M2_score-mean(M2_score))/sd(M2_score)
M3_score <- (M3_score-mean(M3_score))/sd(M3_score)

score_matrix <- rbind(M1_score,M2_score,M3_score)
b <- cor(t(score_matrix))






# Step 4: calculate the posterior
plot(density(M1_score))
plot(density(M2_score))
plot(density(M3_score))
length(which(M1_score>=1&M2_score<=0.5&M3_score<=0.5))
length(which(M2_score>=1&M1_score<=0&M3_score<=0))
length(which(M3_score>=1.2&M1_score<=0&M2_score<=0))

score <- t(rbind(M1_score,M2_score,M3_score))


M1_pos <- M1_score  - 0.5*M2_score-0.5*M3_score
M2_pos <- M2_score  - 0.5*M1_score-0.5*M3_score
M3_pos <- M3_score  - 0.5*M1_score-0.5*M2_score


# Step 5: assign reliable labels to cells, with respect to a pre-defined cutoff

cutoff = 1.5

g1_labels <- names(M1_pos[M1_pos >= cutoff]) # those cells belong to M1
S_labels <- names(M2_pos[M2_pos >= cutoff])# those cells belong to M2
g2_labels <- names(M3_pos[M3_pos >= cutoff])# those cells belong to M3




#g1_labels <- names(M1_score)[which(M1_score>=1&M2_score<=0.5&M3_score<=0.5)]# those cells belong to M1
#S_labels <- names(M1_score)[which(M2_score>=1&M1_score<=0&M3_score<=0)]# those cells belong to M2
#g2_labels <- names(M1_score)[which(M3_score>=1.2&M1_score<=0&M2_score<=0)]# those cells belong to M3




# Step 6:dimension reduction
allpca <- prcomp(t(exp_log2),center = T,scale. = T)
pca_result <- allpca$x
pca_result <- as.data.frame(pca_result)

g1 <- pca_result[g1_labels,1:2]
g1 <- as.data.frame(g1)
s <- pca_result[S_labels,1:2]
s <- as.data.frame(s)
g2_m <- pca_result[g2_labels,1:2]
g2_m <- as.data.frame(g2_m)
other <- pca_result[-which(rownames(pca_result)%in%c(g1_labels,S_labels,g2_labels)),1:2]
other <- as.data.frame(other)

g1$phase <- "g1"
s$phase <- "s"
g2_m$phase <- "g2/m"
other$phase <- "other"
data_plot_2 <- rbind(other,g1,s,g2_m)

plot_markergene <- ggplot(data_plot_2,aes(x = PC1, y = PC2,color=phase))+
  #geom_point(data =pca_result,aes(x = PC1,y = PC2),size = 1)+ 
  geom_point()+scale_color_manual(values = c("other"="#C0C0C0","g2/m" = "#e20612","s" = "#00b0eb","g1" = "#ffd401") )+
  labs(title = "PCA (Cells)", x = "PC1", y = "PC 2")+theme_bw()+theme(plot.title = element_text(hjust = 0.5),panel.grid.major.x=element_blank(),panel.grid.minor.x=element_blank(),panel.grid.minor.y=element_blank(),panel.grid.major.y=element_blank())
#geom_text(data = s, aes(x = PC1, y = PC2, label = rownames(s)), vjust = -0.5, hjust = -0.5, size = 1)

plot_markergene

pca_result_expression <- pca_result






library(TSCAN)

input <- exp_log2

#input <- input[which(!is.na(rowSums(input))),]

#input <- t(apply(input,1,function(x) (x-mean(x))/sd(x)))
#input <- apply(input,2,function(x) (x-mean(x))/sd(x))
colnames(input) <- colnames(original_data)

lpsmclust <- exprmclust(input ,clusternum = 2:9, reduce = T)

## step6 : pca
pca_result <- lpsmclust$pcareduceres

data_plot_3 <- lpsmclust

```

## SigRM
### predict use labeled cells (SigRM)
```{r,eval=FALSE}

## step1 : choose gene
g1_s_genes <- c("MCM6", "PCNA", "SLBP")
s_genes <- c( "RRM2", "DTL","MCM5") 
g2_m_genes <- c("TOP2A",  "CCNB1", "AURKA" )  

## step2 : change to ensebl

library(org.Hs.eg.db)
library(clusterProfiler)
library(stringi)

g1_s_genes_ENSG <- bitr(g1_s_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")
s_genes_ENSG <- bitr(s_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")
g2_m_genes_ENSG <-bitr(g2_m_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")


## step3 : select the tpm data
select_gene <- c(g1_s_genes_ENSG$ENSEMBL,s_genes_ENSG$ENSEMBL,g2_m_genes_ENSG$ENSEMBL)

## step4 : select sites


# get the position of the gene:
expression <- readRDS("D:/data/SMART/origin/gene_expression_73_control.rds")

# get the index of the genes
index <- c()
for ( i in 1:length(select_gene)){
  index[i] <-which(expression[[1]]$Gene.ID%in%select_gene[i])
}

# get the position of the genes
position_start <- expression[[1]]$Start[index]
position_end <- expression[[1]]$End[index]


# get the sites information
SNP_all_51w <- readRDS("D:/data/SMART/preprocess2/SNP_all_51w.rds")

# the sites position
testSNP_all_position <- SNP_all_51w@ranges@start

# match the site to the gene
index_list <- list()
for ( i in 1:length(select_gene)){
  index_list[[i]] <-which(testSNP_all_position%in%(position_start[i]:position_end[i]))
  
  names(index_list[[i]]) <- paste0("gene_",i,"_",1:length(index_list[[i]]))
}

# all index of sites
index_site <- unlist(index_list)
site_length <- unlist(lapply(index_list, length))


chr_info <- rep(SNP_all_51w@seqnames@values,SNP_all_51w@seqnames@lengths)
chr_info_get <- chr_info[index_site]
chr_info_target <- factor(rep(c("chr2","chr20","chr4","chr2","chr1","chr22","chr17","chr5","chr20"),site_length),levels = levels(chr_info_get))

index_correct <- which(chr_info_get ==chr_info_target)
site_length_correct <- list()
start <- 1
end <- site_length[1]
for (i in 1:8){
  site_length_correct[[i]] <- length(which(index_correct%in%c(start:end)))
  start <- start+site_length[i]
  end <- end + site_length[i+1]
}
site_length_correct[[9]] <- length(which(index_correct%in%c(start:end)))
site_index_correct <- index_site [index_correct]

index_site <- site_index_correct
site_length <- unlist(site_length_correct)

# load the data
res_all <- readRDS("D:/data/SMART/results_ks/res_all_4_sim_SigRM/res_list.rds")


# get the TCR matrix 
res_all_TCR <- as.data.frame(lapply(res_all,function(df) df[[8]])) 
res_all_TCR_index_site <- res_all_TCR[index_site,]
#res_all_TCR_index_site [res_all_TCR_index_site<=0] <-0
# get the RR matrix 
res_all_RR <- as.data.frame(lapply(res_all,function(df) df[[3]])) 
res_all_RR_index_site <- res_all_RR[index_site,]

# get the OR matrix 
res_all_OR <- as.data.frame(lapply(res_all,function(df) df[[4]])) 
res_all_OR_index_site <- res_all_OR[index_site,]
res_all_OR_index_site [is.na(res_all_OR_index_site)] <-0
res_all_OR_index_site [res_all_OR_index_site<=0] <-0
#rm(res_all)



# select the OR >1
res_all_OR_index_site_mean <- apply(res_all_OR_index_site,1,function(x) mean(x,na.rm=TRUE))


cut_off=1

hight_num <- c()

hight_num[1] <- length(which(res_all_OR_index_site_mean[1:96][order(res_all_OR_index_site_mean[1:96],decreasing = TRUE)]>cut_off))
hight_num[2] <- length(which(res_all_OR_index_site_mean[97:110][order(res_all_OR_index_site_mean[97:110],decreasing = TRUE)]>cut_off))
hight_num[3] <- length(which(res_all_OR_index_site_mean[111:164][order(res_all_OR_index_site_mean[111:164],decreasing = TRUE)]>cut_off))
hight_num[4] <- length(which(res_all_OR_index_site_mean[165:251][order(res_all_OR_index_site_mean[165:251],decreasing = TRUE)]>cut_off))
hight_num[5] <- length(which(res_all_OR_index_site_mean[252:373][order(res_all_OR_index_site_mean[252:373],decreasing = TRUE)]>cut_off))
hight_num[6] <- length(which(res_all_OR_index_site_mean[374:415][order(res_all_OR_index_site_mean[374:415],decreasing = TRUE)]>cut_off))
hight_num[7] <- length(which(res_all_OR_index_site_mean[416:658][order(res_all_OR_index_site_mean[416:658],decreasing = TRUE)]>cut_off))
hight_num[8] <- length(which(res_all_OR_index_site_mean[659:698][order(res_all_OR_index_site_mean[659:698],decreasing = TRUE)]>cut_off))
hight_num[9] <- length(which(res_all_OR_index_site_mean[699:780][order(res_all_OR_index_site_mean[699:780],decreasing = TRUE)]>cut_off))




#round(hight_num*site_length[[1]])
order_50_list <- list()
order_50_list[[1]] <-order(res_all_OR_index_site_mean[1:site_length[[1]]],decreasing=TRUE)[1:hight_num[1]]
order_50_list[[1]] <- c(1:site_length[[1]])[order_50_list[[1]]]

for (i in 2:9){
  order_50_list[[i]] <-order(res_all_OR_index_site_mean[(sum(unlist(site_length[1:(i-1)]))+1):sum(unlist(site_length[1:i]))],decreasing=TRUE)[1:hight_num[i]]
  order_50_list[[i]] <- c((sum(unlist(site_length[1:(i-1)]))+1):sum(unlist(site_length[1:i]))) [order_50_list[[i]]]
}



# get the each cutoff>1

res_meta_OR_index_site_50_list <- list()
for (i in 1:9) {
  res_meta_OR_index_site_50_list[[i]] <- apply(res_all_OR_index_site [order_50_list[[i]],],2,function(x) mean(x,na.rm=TRUE))
}
res_meta_OR_index_site_50_list <- t(as.data.frame(res_meta_OR_index_site_50_list))
colnames(res_meta_OR_index_site_50_list) <- colnames(original_data)
rownames(res_meta_OR_index_site_50_list) <- NULL

for (i in 1:9){order_50_list[[i]] <- order_50_list[[i]][which(order_50_list[[i]]%in%index_correct)]}



library(TSCAN)

input <- res_meta_OR_index_site_50_list
input[is.na(input)] <- 0

#input <- input[which(!is.na(rowSums(input))),]

#input <- t(apply(input,1,function(x) (x-mean(x))/sd(x)))
#input <- apply(input,2,function(x) (x-mean(x))/sd(x))
colnames(input) <- colnames(original_data)

lpsmclust <- exprmclust(input ,clusternum = 4, reduce = T)

## step6 : pca
pca_result <- lpsmclust$pcareduceres
g1 <- pca_result[g1_labels,1:dim( lpsmclust$pcareduceres)[2]]
g1 <- as.data.frame(g1)
s <- pca_result[S_labels,1:dim( lpsmclust$pcareduceres)[2]]
s <- as.data.frame(s)
g2 <- pca_result[g2_labels,1:dim( lpsmclust$pcareduceres)[2]]
g2 <- as.data.frame(g2)
other <- pca_result[-which(rownames(pca_result)%in%c(g1_labels,S_labels,g2_labels)),1:dim( lpsmclust$pcareduceres)[2]]
other <- as.data.frame(other)



g1$phase <- "g1"
s$phase <- "s"
g2$phase <- "g2"
other$phase <- "other"
data_plot <- rbind(other,g1,s,g2)
data_plot <- data_plot[,c(1,2,4)]
data_plot[,2] <- -data_plot[,2]
data_plot[,1] <- -data_plot[,1]
data_plot_ <- data_plot
data_plot[,1] <- data_plot_[,2]
data_plot[,2] <- -data_plot_[,1]

data_plot <- data_plot[which(rownames(data_plot)%in%names(lpsmclust$clusterid)),]
data_plot <- data_plot[order(rownames(data_plot)),]
data_plot$cluster <- lpsmclust$clusterid[order(names(lpsmclust$clusterid))]
data_plot$cluster[which(data_plot$cluster=="1")] <- "5"
data_plot$cluster[which(data_plot$cluster=="2")] <- "6"
data_plot$cluster[which(data_plot$cluster=="3")] <- "7"
data_plot$cluster[which(data_plot$cluster=="5")] <- "3"
data_plot$cluster[which(data_plot$cluster=="6")] <- "1"
data_plot$cluster[which(data_plot$cluster=="7")] <- "2"

data_plot$phase[which(data_plot$phase=="g1")] == "G1"
data_plot$phase[which(data_plot$phase=="g2")] == "G2/M"
data_plot$phase[which(data_plot$phase=="s")] == "S"
data_plot$phase[which(data_plot$phase=="other")] == "Unclear"

data_plot$phase <- factor(data_plot$phase)
data_plot$cluster <- factor(data_plot$cluster)


center_1 <- apply(data_plot[which(data_plot$cluster=="1"),1:2],2,mean)
center_2 <- apply(data_plot[which(data_plot$cluster=="2"),1:2],2,mean)
center_3 <- apply(data_plot[which(data_plot$cluster=="3"),1:2],2,mean)
center_4 <- apply(data_plot[which(data_plot$cluster=="4"),1:2],2,mean)

center <- as.data.frame(rbind(center_2,center_3))
colnames(center) <- c("x","y")
center2 <- as.data.frame(rbind(center_3,center_4))
colnames(center2) <- c("x","y")
center3 <- as.data.frame(rbind(center_2,center_1))
colnames(center3) <- c("x","y")

plot_1 <-  ggplot() +
    geom_point(data = data_plot, aes(x = PC1, y = PC2, color = cluster),shape =20,size=1.7) +
    stat_ellipse(data = data_plot, level = 0.9, linetype = 1, size = 2, 
                 aes(x = PC1, y = PC2, fill = cluster), alpha = 0.3, segments = 50, geom = "polygon",type = "t")  +
    theme_bw() +
    geom_line(data = center, aes(x = x, y = y), size = 1) +
    geom_line(data = center2, aes(x = x, y = y), size = 1) +
    geom_line(data = center3, aes(x = x, y = y), size = 1) +
    annotate("text", label = "1", x = center_1[1], y = center_1[2], size = 7, colour = "black") +
    annotate("text", label = "2", x = center_2[1], y = center_2[2], size = 7, colour = "black") +
    annotate("text", label = "3", x = center_3[1], y = center_3[2], size = 7, colour = "black") +
    annotate("text", label = "4", x = center_4[1], y = center_4[2], size = 7, colour = "black") +
    theme(legend.key.width = unit(0.4, "cm"),legend.key.height = unit(0.5, "cm"),legend.position = c(0.911, 0.825),legend.background = element_rect(fill = "#f1f1f1")) +
    coord_cartesian(xlim = c(-5, +3), ylim = c(-4, 5))+scale_color_manual(values = c("1" = "#00C5CD","2" = "#63ba10","3" = "#CE69C9","4" = "#F8766D") )+scale_fill_manual(values = c("1" = "#00C5CD","2" = "#63ba10","3" = "#CE69C9","4" = "#F8766D") )+xlab(TeX("PC1 ($\\m^6A$)"))+ylab(TeX("PC2 ($\\m^6A$)"))


plot_2 <-  ggplot(data_plot, aes(x=PC1,y=PC2))+   geom_point( aes(color=phase),shape =20,size=1.7)+theme_bw()+ 
    theme(legend.key.width = unit(0.4, "cm"),legend.key.height = unit(0.5, "cm"),legend.position = c(0.9, 0.825),legend.background = element_rect(fill = "#f1f1f1")) +coord_cartesian(xlim = c(-5, +3), ylim = c(-4, 5))+scale_color_manual(values = c("other" = "#C0C0C0","s" = "#F8766D","g2" = "#00BA38","g1" = "#619CFF") )+xlab(TeX("PC1 ($\\m^6A$)"))+ylab(TeX("PC2 ($\\m^6A$)")) #

plot_1
plot_2

library(cowplot)
plot <- plot_grid(plot_1,plot_2, labels = c('a', 'b'),label_size = 20)
plot
```

### fisher test SigRM
```{r,eval=FALSE}


G1_exp <- rep(0,length(data_plot$phase))
G1_exp[which(data_plot$phase=="g1")] <- 1
s_exp <- rep(0,length(data_plot$phase))
s_exp[which(data_plot$phase=="s")] <- 1
G2_exp <- rep(0,length(data_plot$phase))
G2_exp[which(data_plot$phase=="g2")] <- 1
other_exp <- rep(0,length(data_plot$phase))
other_exp[which(data_plot$phase=="other")] <- 1

exp_list <- list()
exp_list[[1]] <- G1_exp
exp_list[[2]] <- s_exp
exp_list[[3]] <- G2_exp
exp_list[[4]] <- other_exp

cluster_1_sigrm <- rep(0,length(data_plot$cluster))
cluster_1_sigrm[which(data_plot$cluster=="1")] <- 1
cluster_2_sigrm <- rep(0,length(data_plot$cluster))
cluster_2_sigrm[which(data_plot$cluster=="2")] <- 1
cluster_3_sigrm <- rep(0,length(data_plot$cluster))
cluster_3_sigrm[which(data_plot$cluster=="3")] <- 1
cluster_4_sigrm <- rep(0,length(data_plot$cluster))
cluster_4_sigrm[which(data_plot$cluster=="4")] <- 1

sigrm_list <- list()
sigrm_list[[1]] <- cluster_1_sigrm
sigrm_list[[2]] <- cluster_2_sigrm
sigrm_list[[3]] <- cluster_3_sigrm
sigrm_list[[4]] <- cluster_4_sigrm

Sigrm_table_pvalue <- matrix(NA,4,4)
rownames(Sigrm_table_pvalue) <- c("G1","S","G2/M","unclear")
colnames(Sigrm_table_pvalue) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4")

Sigrm_table_OR <- matrix(NA,4,4)
rownames(Sigrm_table_pvalue) <- c("G1","S","G2/M","unclear")
colnames(Sigrm_table_pvalue) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4")

for (i in 1:4){
  for (j in 1: 4){
    a <- fisher.test(exp_list[[i]],sigrm_list[[j]],alternative = "greater")
    Sigrm_table_pvalue[i,j] <- a$p.value
    Sigrm_table_OR[i,j] <- a[["estimate"]][["odds ratio"]]
  }
}

Sigrm_table_pvalue
Sigrm_table_OR


```

## DESeq2
### predict use labeled cells (DESeq2)
```{r,eval=FALSE}

## step1 : choose gene
g1_s_genes <- c("MCM6", "PCNA", "SLBP")
s_genes <- c( "RRM2", "DTL","MCM5") 
g2_m_genes <- c("TOP2A",  "CCNB1", "AURKA" )  

## step2 : change to ensebl

library(org.Hs.eg.db)
library(clusterProfiler)
library(stringi)

g1_s_genes_ENSG <- bitr(g1_s_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")
s_genes_ENSG <- bitr(s_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")
g2_m_genes_ENSG <-bitr(g2_m_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")


## step3 : select the tpm data
select_gene <- c(g1_s_genes_ENSG$ENSEMBL,s_genes_ENSG$ENSEMBL,g2_m_genes_ENSG$ENSEMBL)

## step4 : select sites


# get the position of the gene:
expression <- readRDS("D:/data/SMART/origin/gene_expression_73_control.rds")

# get the index of the genes
index <- c()
for ( i in 1:length(select_gene)){
  index[i] <-which(expression[[1]]$Gene.ID%in%select_gene[i])
}

# get the position of the genes
position_start <- expression[[1]]$Start[index]
position_end <- expression[[1]]$End[index]


# get the sites information
SNP_all_51w <- readRDS("D:/data/SMART/preprocess2/SNP_all_51w.rds")

# the sites position
testSNP_all_position <- SNP_all_51w@ranges@start

# match the site to the gene
index_list <- list()
for ( i in 1:length(select_gene)){
  index_list[[i]] <-which(testSNP_all_position%in%(position_start[i]:position_end[i]))
  
  names(index_list[[i]]) <- paste0("gene_",i,"_",1:length(index_list[[i]]))
}

# all index of sites
index_site <- unlist(index_list)
site_length <- unlist(lapply(index_list, length))


chr_info <- rep(SNP_all_51w@seqnames@values,SNP_all_51w@seqnames@lengths)
chr_info_get <- chr_info[index_site]
chr_info_target <- factor(rep(c("chr2","chr20","chr4","chr2","chr1","chr22","chr17","chr5","chr20"),site_length),levels = levels(chr_info_get))

index_correct <- which(chr_info_get ==chr_info_target)
site_length_correct <- list()
start <- 1
end <- site_length[1]
for (i in 1:8){
  site_length_correct[[i]] <- length(which(index_correct%in%c(start:end)))
  start <- start+site_length[i]
  end <- end + site_length[i+1]
}
site_length_correct[[9]] <- length(which(index_correct%in%c(start:end)))
site_index_correct <- index_site [index_correct]

index_site <- site_index_correct
site_length <- unlist(site_length_correct)


res_all_DESeq2 <- readRDS("D:/data/SMART/results_ks/res_all_4_sim_DESeq2/res_list.rds")




# get the OR matrix 
res_all_OR <- as.data.frame(lapply(res_all_DESeq2,function(df) df[[1]]@listData[["log2FoldChange"]])) 
res_all_OR_index_site <- -res_all_OR[index_site,]
res_all_OR_index_site [is.na(res_all_OR_index_site)] <-0
res_all_OR_index_site [res_all_OR_index_site<=0] <-0
#rm(res_all)



res_all_OR_index_site_mean <- apply(res_all_OR_index_site,1,function(x) mean(x,na.rm=TRUE))


hight_num <- c(19,7,20,51,15,17,23,11,22)



#round(hight_num*site_length[[1]])
order_50_list <- list()
order_50_list[[1]] <-order(res_all_OR_index_site_mean[1:site_length[[1]]],decreasing=TRUE)[1:hight_num[1]]
order_50_list[[1]] <- c(1:site_length[[1]])[order_50_list[[1]]]

for (i in 2:9){
  order_50_list[[i]] <-order(res_all_OR_index_site_mean[(sum(unlist(site_length[1:(i-1)]))+1):sum(unlist(site_length[1:i]))],decreasing=TRUE)[1:hight_num[i]]
  order_50_list[[i]] <- c((sum(unlist(site_length[1:(i-1)]))+1):sum(unlist(site_length[1:i]))) [order_50_list[[i]]]
}




res_meta_OR_index_site_50_list <- list()
for (i in 1:9) {
  res_meta_OR_index_site_50_list[[i]] <- apply(res_all_OR_index_site [order_50_list[[i]],],2,function(x) mean(x,na.rm=TRUE))
}
res_meta_OR_index_site_50_list <- t(as.data.frame(res_meta_OR_index_site_50_list))
colnames(res_meta_OR_index_site_50_list) <- colnames(original_data)
rownames(res_meta_OR_index_site_50_list) <- NULL




library(TSCAN)

input <- res_meta_OR_index_site_50_list
input[is.na(input)] <- 0

#input <- input[which(!is.na(rowSums(input))),]

#input <- t(apply(input,1,function(x) (x-mean(x))/sd(x)))
#input <- apply(input,2,function(x) (x-mean(x))/sd(x))
colnames(input) <- colnames(original_data)

lpsmclust <- exprmclust(input ,clusternum = 4, reduce = T)

## step6 : pca
pca_result <- lpsmclust$pcareduceres
g1 <- pca_result[g1_labels,1:dim( lpsmclust$pcareduceres)[2]]
g1 <- as.data.frame(g1)
s <- pca_result[S_labels,1:dim( lpsmclust$pcareduceres)[2]]
s <- as.data.frame(s)
g2 <- pca_result[g2_labels,1:dim( lpsmclust$pcareduceres)[2]]
g2 <- as.data.frame(g2)
other <- pca_result[-which(rownames(pca_result)%in%c(g1_labels,S_labels,g2_labels)),1:dim( lpsmclust$pcareduceres)[2]]
other <- as.data.frame(other)



g1$phase <- "g1"
s$phase <- "s"
g2$phase <- "g2"
other$phase <- "other"
data_plot <- rbind(other,g1,s,g2)


data_plot <- data_plot[which(rownames(data_plot)%in%names(lpsmclust$clusterid)),]
data_plot <- data_plot[order(rownames(data_plot)),]
data_plot$cluster <- lpsmclust$clusterid[order(names(lpsmclust$clusterid))]
data_plot$phase <- factor(data_plot$phase)
data_plot$cluster <- factor(data_plot$cluster)


center_1 <- apply(data_plot[which(data_plot$cluster=="1"),1:2],2,mean)
center_2 <- apply(data_plot[which(data_plot$cluster=="2"),1:2],2,mean)
center_3 <- apply(data_plot[which(data_plot$cluster=="3"),1:2],2,mean)
center_4 <- apply(data_plot[which(data_plot$cluster=="4"),1:2],2,mean)

center <- as.data.frame(rbind(center_3,center_4))
colnames(center) <- c("x","y")
center2 <- as.data.frame(rbind(center_2,center_3))
colnames(center2) <- c("x","y")
center3 <- as.data.frame(rbind(center_2,center_1))
colnames(center3) <- c("x","y")

plot_1 <-  ggplot() +
    geom_point(data = data_plot, aes(x = PC1, y = PC2, color = cluster)) +
    stat_ellipse(data = data_plot, level = 0.9, linetype = 1, size = 2, 
                 aes(x = PC1, y = PC2, fill = cluster), alpha = 0.3, segments = 50, geom = "polygon",type = "t")  +
    theme_bw() +
    geom_line(data = center, aes(x = x, y = y), size = 1) +
    geom_line(data = center2, aes(x = x, y = y), size = 1) +
    geom_line(data = center3, aes(x = x, y = y), size = 1) +
    annotate("text", label = "1", x = center_1[1], y = center_1[2], size = 7, colour = "black") +
    annotate("text", label = "2", x = center_2[1], y = center_2[2], size = 7, colour = "black") +
    annotate("text", label = "3", x = center_3[1], y = center_3[2], size = 7, colour = "black") +
    annotate("text", label = "4", x = center_4[1], y = center_4[2], size = 7, colour = "black") +
    theme(legend.key.width = unit(0.5, "cm"))


plot_2 <-  ggplot(data_plot, aes(x=PC1,y=PC2))+   geom_point( aes(color=phase))+theme_bw()+ theme(legend.key.width = unit(0.5, "cm"))+scale_color_manual(values = c("other" = "#C0C0C0","s" = "#F8766D","g2" = "#00BA38","g1" = "#619CFF") )

plot_1
plot_2
```
### fisher test DESeq2
```{r,eval=FALSE}
G1_exp <- rep(0,length(data_plot$phase))
G1_exp[which(data_plot$phase=="g1")] <- 1
s_exp <- rep(0,length(data_plot$phase))
s_exp[which(data_plot$phase=="s")] <- 1
G2_exp <- rep(0,length(data_plot$phase))
G2_exp[which(data_plot$phase=="g2")] <- 1
other_exp <- rep(0,length(data_plot$phase))
other_exp[which(data_plot$phase=="other")] <- 1

exp_list <- list()
exp_list[[1]] <- G1_exp
exp_list[[2]] <- s_exp
exp_list[[3]] <- G2_exp
exp_list[[4]] <- other_exp

cluster_1_DESeq2 <- rep(0,length(data_plot$cluster))
cluster_1_DESeq2[which(data_plot$cluster=="1")] <- 1
cluster_2_DESeq2 <- rep(0,length(data_plot$cluster))
cluster_2_DESeq2[which(data_plot$cluster=="2")] <- 1
cluster_3_DESeq2 <- rep(0,length(data_plot$cluster))
cluster_3_DESeq2[which(data_plot$cluster=="3")] <- 1
cluster_4_DESeq2 <- rep(0,length(data_plot$cluster))
cluster_4_DESeq2[which(data_plot$cluster=="4")] <- 1

DESeq2_list <- list()
DESeq2_list[[1]] <- cluster_1_DESeq2
DESeq2_list[[2]] <- cluster_2_DESeq2
DESeq2_list[[3]] <- cluster_3_DESeq2
DESeq2_list[[4]] <- cluster_4_DESeq2

DESeq2_table_pvalue <- matrix(NA,4,4)
rownames(DESeq2_table_pvalue) <- c("G1","S","G2/M","unclear")
colnames(DESeq2_table_pvalue) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4")

DESeq2_table_OR <- matrix(NA,4,4)
rownames(DESeq2_table_pvalue) <- c("G1","S","G2/M","unclear")
colnames(DESeq2_table_pvalue) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4")

for (i in 1:4){
  for (j in 1: 4){
    a <- fisher.test(exp_list[[i]],DESeq2_list[[j]],alternative = "greater")
    DESeq2_table_pvalue[i,j] <- a$p.value
    DESeq2_table_OR[i,j] <- a[["estimate"]][["odds ratio"]]
  }
}

DESeq2_table_pvalue
DESeq2_table_OR
```

## edgeR
### predict use labeled cells (edgeR)
```{r,eval=FALSE}

## step1 : choose gene
g1_s_genes <- c("MCM6", "PCNA", "SLBP")
s_genes <- c( "RRM2", "DTL","MCM5") 
g2_m_genes <- c("TOP2A",  "CCNB1", "AURKA" )  

## step2 : change to ensebl

library(org.Hs.eg.db)
library(clusterProfiler)
library(stringi)

g1_s_genes_ENSG <- bitr(g1_s_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")
s_genes_ENSG <- bitr(s_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")
g2_m_genes_ENSG <-bitr(g2_m_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")


## step3 : select the tpm data
select_gene <- c(g1_s_genes_ENSG$ENSEMBL,s_genes_ENSG$ENSEMBL,g2_m_genes_ENSG$ENSEMBL)

## step4 : select sites


# get the position of the gene:
expression <- readRDS("D:/data/SMART/origin/gene_expression_73_control.rds")

# get the index of the genes
index <- c()
for ( i in 1:length(select_gene)){
  index[i] <-which(expression[[1]]$Gene.ID%in%select_gene[i])
}

# get the position of the genes
position_start <- expression[[1]]$Start[index]
position_end <- expression[[1]]$End[index]


# get the sites information
SNP_all_51w <- readRDS("D:/data/SMART/preprocess2/SNP_all_51w.rds")

# the sites position
testSNP_all_position <- SNP_all_51w@ranges@start

# match the site to the gene
index_list <- list()
for ( i in 1:length(select_gene)){
  index_list[[i]] <-which(testSNP_all_position%in%(position_start[i]:position_end[i]))
  
  names(index_list[[i]]) <- paste0("gene_",i,"_",1:length(index_list[[i]]))
}

# all index of sites
index_site <- unlist(index_list)
site_length <- unlist(lapply(index_list, length))


chr_info <- rep(SNP_all_51w@seqnames@values,SNP_all_51w@seqnames@lengths)
chr_info_get <- chr_info[index_site]
chr_info_target <- factor(rep(c("chr2","chr20","chr4","chr2","chr1","chr22","chr17","chr5","chr20"),site_length),levels = levels(chr_info_get))

index_correct <- which(chr_info_get ==chr_info_target)
site_length_correct <- list()
start <- 1
end <- site_length[1]
for (i in 1:8){
  site_length_correct[[i]] <- length(which(index_correct%in%c(start:end)))
  start <- start+site_length[i]
  end <- end + site_length[i+1]
}
site_length_correct[[9]] <- length(which(index_correct%in%c(start:end)))
site_index_correct <- index_site [index_correct]

index_site <- site_index_correct
site_length <- unlist(site_length_correct)


res_all_edgeR <- readRDS("D:/data/SMART/results_ks/res_all_4_sim_edgeR/res_list.rds")




# get the OR matrix 
res_all_OR <- as.data.frame(lapply(res_all_edgeR,function(df) df[[1]][["logFC"]])) 
res_all_OR_index_site <- -res_all_OR[index_site,]
res_all_OR_index_site [is.na(res_all_OR_index_site)] <-0
res_all_OR_index_site [res_all_OR_index_site<=0] <-0
#rm(res_all)




res_all_OR_index_site_mean <- apply(res_all_OR_index_site,1,function(x) mean(x,na.rm=TRUE))

hight_num <- c(19,7,20,51,15,17,23,11,22)





#round(hight_num*site_length[[1]])
order_50_list <- list()
order_50_list[[1]] <-order(res_all_OR_index_site_mean[1:site_length[[1]]],decreasing=TRUE)[1:hight_num[1]]
order_50_list[[1]] <- c(1:site_length[[1]])[order_50_list[[1]]]

for (i in 2:9){
  order_50_list[[i]] <-order(res_all_OR_index_site_mean[(sum(unlist(site_length[1:(i-1)]))+1):sum(unlist(site_length[1:i]))],decreasing=TRUE)[1:hight_num[i]]
  order_50_list[[i]] <- c((sum(unlist(site_length[1:(i-1)]))+1):sum(unlist(site_length[1:i]))) [order_50_list[[i]]]
}




res_meta_OR_index_site_50_list <- list()
for (i in 1:9) {
  res_meta_OR_index_site_50_list[[i]] <- apply(res_all_OR_index_site [order_50_list[[i]],],2,function(x) mean(x,na.rm=TRUE))
}
res_meta_OR_index_site_50_list <- t(as.data.frame(res_meta_OR_index_site_50_list))
colnames(res_meta_OR_index_site_50_list) <- colnames(original_data)
rownames(res_meta_OR_index_site_50_list) <- NULL



library(TSCAN)

input <- res_meta_OR_index_site_50_list
input[is.na(input)] <- 0

#input <- input[which(!is.na(rowSums(input))),]

#input <- t(apply(input,1,function(x) (x-mean(x))/sd(x)))
#input <- apply(input,2,function(x) (x-mean(x))/sd(x))
colnames(input) <- colnames(original_data)

lpsmclust <- exprmclust(input ,clusternum = 4, reduce = T)

## step6 : pca
pca_result <- lpsmclust$pcareduceres
g1 <- pca_result[g1_labels,1:dim( lpsmclust$pcareduceres)[2]]
g1 <- as.data.frame(g1)
s <- pca_result[S_labels,1:dim( lpsmclust$pcareduceres)[2]]
s <- as.data.frame(s)
g2 <- pca_result[g2_labels,1:dim( lpsmclust$pcareduceres)[2]]
g2 <- as.data.frame(g2)
other <- pca_result[-which(rownames(pca_result)%in%c(g1_labels,S_labels,g2_labels)),1:dim( lpsmclust$pcareduceres)[2]]
other <- as.data.frame(other)



g1$phase <- "g1"
s$phase <- "s"
g2$phase <- "g2"
other$phase <- "other"
data_plot <- rbind(other,g1,s,g2)


data_plot <- data_plot[which(rownames(data_plot)%in%names(lpsmclust$clusterid)),]
data_plot <- data_plot[order(rownames(data_plot)),]
data_plot$cluster <- lpsmclust$clusterid[order(names(lpsmclust$clusterid))]
data_plot$cluster[which(data_plot$cluster=="2")] <- "5"
data_plot$cluster[which(data_plot$cluster=="3")] <- "6"
data_plot$cluster[which(data_plot$cluster=="5")] <- "3"
data_plot$cluster[which(data_plot$cluster=="6")] <- "2"
data_plot$phase <- factor(data_plot$phase)
data_plot$cluster <- factor(data_plot$cluster)


center_1 <- apply(data_plot[which(data_plot$cluster=="1"),1:2],2,mean)
center_2 <- apply(data_plot[which(data_plot$cluster=="2"),1:2],2,mean)
center_3 <- apply(data_plot[which(data_plot$cluster=="3"),1:2],2,mean)
center_4 <- apply(data_plot[which(data_plot$cluster=="4"),1:2],2,mean)

center <- as.data.frame(rbind(center_3,center_4))
colnames(center) <- c("x","y")
center2 <- as.data.frame(rbind(center_2,center_3))
colnames(center2) <- c("x","y")
center3 <- as.data.frame(rbind(center_2,center_1))
colnames(center3) <- c("x","y")

plot_1 <-  ggplot() +
    geom_point(data = data_plot, aes(x = PC1, y = PC2, color = cluster)) +
    stat_ellipse(data = data_plot, level = 0.9, linetype = 1, size = 2, 
                 aes(x = PC1, y = PC2, fill = cluster), alpha = 0.3, segments = 50, geom = "polygon",type = "t")  +
    theme_bw() +
    geom_line(data = center, aes(x = x, y = y), size = 1) +
    geom_line(data = center2, aes(x = x, y = y), size = 1) +
    geom_line(data = center3, aes(x = x, y = y), size = 1) +
    annotate("text", label = "1", x = center_1[1], y = center_1[2], size = 7, colour = "black") +
    annotate("text", label = "2", x = center_2[1], y = center_2[2], size = 7, colour = "black") +
    annotate("text", label = "3", x = center_3[1], y = center_3[2], size = 7, colour = "black") +
    annotate("text", label = "4", x = center_4[1], y = center_4[2], size = 7, colour = "black") +
    theme(legend.key.width = unit(0.5, "cm"))


plot_2 <-  ggplot(data_plot, aes(x=PC1,y=PC2))+   geom_point( aes(color=phase))+theme_bw()+ theme(legend.key.width = unit(0.5, "cm"))+scale_color_manual(values = c("other" = "#C0C0C0","s" = "#F8766D","g2" = "#00BA38","g1" = "#619CFF") )

plot_1
plot_2
```
### fisher test edgeR
```{r,eval=FALSE}
G1_exp <- rep(0,length(data_plot$phase))
G1_exp[which(data_plot$phase=="g1")] <- 1
s_exp <- rep(0,length(data_plot$phase))
s_exp[which(data_plot$phase=="s")] <- 1
G2_exp <- rep(0,length(data_plot$phase))
G2_exp[which(data_plot$phase=="g2")] <- 1
other_exp <- rep(0,length(data_plot$phase))
other_exp[which(data_plot$phase=="other")] <- 1

exp_list <- list()
exp_list[[1]] <- G1_exp
exp_list[[2]] <- s_exp
exp_list[[3]] <- G2_exp
exp_list[[4]] <- other_exp

cluster_1_edgeR <- rep(0,length(data_plot$cluster))
cluster_1_edgeR[which(data_plot$cluster=="1")] <- 1
cluster_2_edgeR <- rep(0,length(data_plot$cluster))
cluster_2_edgeR[which(data_plot$cluster=="2")] <- 1
cluster_3_edgeR <- rep(0,length(data_plot$cluster))
cluster_3_edgeR[which(data_plot$cluster=="3")] <- 1
cluster_4_edgeR <- rep(0,length(data_plot$cluster))
cluster_4_edgeR[which(data_plot$cluster=="4")] <- 1

edgeR_list <- list()
edgeR_list[[1]] <- cluster_1_edgeR
edgeR_list[[2]] <- cluster_2_edgeR
edgeR_list[[3]] <- cluster_3_edgeR
edgeR_list[[4]] <- cluster_4_edgeR

edgeR_table_pvalue <- matrix(NA,4,4)
rownames(edgeR_table_pvalue) <- c("G1","S","G2/M","unclear")
colnames(edgeR_table_pvalue) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4")

edgeR_table_OR <- matrix(NA,4,4)
rownames(edgeR_table_pvalue) <- c("G1","S","G2/M","unclear")
colnames(edgeR_table_pvalue) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4")

for (i in 1:4){
  for (j in 1: 4){
    a <- fisher.test(exp_list[[i]],edgeR_list[[j]],alternative = "greater")
    edgeR_table_pvalue[i,j] <- a$p.value
    edgeR_table_OR[i,j] <- a[["estimate"]][["odds ratio"]]
  }
}

edgeR_table_pvalue
edgeR_table_OR
```

## Bullseye
### predict use labeled cells (Bullseye)
```{r,eval=FALSE}

## step1 : choose gene
g1_s_genes <- c("MCM6", "PCNA", "SLBP")
s_genes <- c( "RRM2", "DTL","MCM5") 
g2_m_genes <- c("TOP2A",  "CCNB1", "AURKA" )  

## step2 : change to ensebl

library(org.Hs.eg.db)
library(clusterProfiler)
library(stringi)

g1_s_genes_ENSG <- bitr(g1_s_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")
s_genes_ENSG <- bitr(s_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")
g2_m_genes_ENSG <-bitr(g2_m_genes, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")


## step3 : select the tpm data
select_gene <- c(g1_s_genes_ENSG$ENSEMBL,s_genes_ENSG$ENSEMBL,g2_m_genes_ENSG$ENSEMBL)

## step4 : select sites


# get the position of the gene:
expression <- readRDS("D:/data/SMART/origin/gene_expression_73_control.rds")

# get the index of the genes
index <- c()
for ( i in 1:length(select_gene)){
  index[i] <-which(expression[[1]]$Gene.ID%in%select_gene[i])
}

# get the position of the genes
position_start <- expression[[1]]$Start[index]
position_end <- expression[[1]]$End[index]


# get the sites information
SNP_all_51w <- readRDS("D:/data/SMART/preprocess2/SNP_all_51w.rds")

# the sites position
testSNP_all_position <- SNP_all_51w@ranges@start

# match the site to the gene
index_list <- list()
for ( i in 1:length(select_gene)){
  index_list[[i]] <-which(testSNP_all_position%in%(position_start[i]:position_end[i]))
  
  names(index_list[[i]]) <- paste0("gene_",i,"_",1:length(index_list[[i]]))
}

# all index of sites
index_site <- unlist(index_list)
site_length <- unlist(lapply(index_list, length))


chr_info <- rep(SNP_all_51w@seqnames@values,SNP_all_51w@seqnames@lengths)
chr_info_get <- chr_info[index_site]
chr_info_target <- factor(rep(c("chr2","chr20","chr4","chr2","chr1","chr22","chr17","chr5","chr20"),site_length),levels = levels(chr_info_get))

index_correct <- which(chr_info_get ==chr_info_target)
site_length_correct <- list()
start <- 1
end <- site_length[1]
for (i in 1:8){
  site_length_correct[[i]] <- length(which(index_correct%in%c(start:end)))
  start <- start+site_length[i]
  end <- end + site_length[i+1]
}
site_length_correct[[9]] <- length(which(index_correct%in%c(start:end)))
site_index_correct <- index_site [index_correct]

index_site <- site_index_correct
site_length <- unlist(site_length_correct)


frequency_all_processed <- readRDS("D:/data/SMART/preprocess2/frequency_all_processed.rds")
frequency_all_processed <- frequency_all_processed [392:1382]

frequency_all_processed_index_site_m <- as.data.frame(lapply(frequency_all_processed,function(df) df[["methylated_reads"]][index_site])) 
frequency_all_processed_index_site_unm <-   as.data.frame(lapply(frequency_all_processed,function(df) df[["unmethylated_reads"]][index_site])) 

# get the OR matrix 
res_all_OR_index_site <- log2((frequency_all_processed_index_site_m+1)/(frequency_all_processed_index_site_unm+1))
res_all_OR_index_site [is.na(res_all_OR_index_site)] <-0
#rm(res_all)



res_all_OR_index_site_mean <- apply(res_all_OR_index_site,1,function(x) mean(x,na.rm=TRUE))

hight_num <- c(19,7,20,51,15,17,23,11,22)



#round(hight_num*site_length[[1]])
order_50_list <- list()
order_50_list[[1]] <-order(res_all_OR_index_site_mean[1:site_length[[1]]],decreasing=TRUE)[1:hight_num[1]]
order_50_list[[1]] <- c(1:site_length[[1]])[order_50_list[[1]]]

for (i in 2:9){
  order_50_list[[i]] <-order(res_all_OR_index_site_mean[(sum(unlist(site_length[1:(i-1)]))+1):sum(unlist(site_length[1:i]))],decreasing=TRUE)[1:hight_num[i]]
  order_50_list[[i]] <- c((sum(unlist(site_length[1:(i-1)]))+1):sum(unlist(site_length[1:i]))) [order_50_list[[i]]]
}




res_meta_OR_index_site_50_list <- list()
for (i in 1:9) {
  res_meta_OR_index_site_50_list[[i]] <- apply(res_all_OR_index_site [order_50_list[[i]],],2,function(x) mean(x,na.rm=TRUE))
}
res_meta_OR_index_site_50_list <- t(as.data.frame(res_meta_OR_index_site_50_list))
colnames(res_meta_OR_index_site_50_list) <- colnames(original_data)
rownames(res_meta_OR_index_site_50_list) <- NULL




library(TSCAN)

input <- res_meta_OR_index_site_50_list
input[is.na(input)] <- 0

#input <- input[which(!is.na(rowSums(input))),]

#input <- t(apply(input,1,function(x) (x-mean(x))/sd(x)))
#input <- apply(input,2,function(x) (x-mean(x))/sd(x))
colnames(input) <- colnames(original_data)

lpsmclust <- exprmclust(input ,clusternum = 4, reduce = T)

## step6 : pca
pca_result <- lpsmclust$pcareduceres
g1 <- pca_result[g1_labels,1:dim( lpsmclust$pcareduceres)[2]]
g1 <- as.data.frame(g1)
s <- pca_result[S_labels,1:dim( lpsmclust$pcareduceres)[2]]
s <- as.data.frame(s)
g2 <- pca_result[g2_labels,1:dim( lpsmclust$pcareduceres)[2]]
g2 <- as.data.frame(g2)
other <- pca_result[-which(rownames(pca_result)%in%c(g1_labels,S_labels,g2_labels)),1:dim( lpsmclust$pcareduceres)[2]]
other <- as.data.frame(other)



g1$phase <- "g1"
s$phase <- "s"
g2$phase <- "g2"
other$phase <- "other"
data_plot <- rbind(other,g1,s,g2)


data_plot <- data_plot[which(rownames(data_plot)%in%names(lpsmclust$clusterid)),]
data_plot <- data_plot[order(rownames(data_plot)),]
data_plot$cluster <- lpsmclust$clusterid[order(names(lpsmclust$clusterid))]
data_plot$cluster[which(data_plot$cluster=="2")] <- "5"
data_plot$cluster[which(data_plot$cluster=="3")] <- "6"
data_plot$cluster[which(data_plot$cluster=="5")] <- "3"
data_plot$cluster[which(data_plot$cluster=="6")] <- "2"
data_plot$cluster[which(data_plot$cluster=="3")] <- "7"
data_plot$cluster[which(data_plot$cluster=="4")] <- "7"
data_plot$cluster[which(data_plot$cluster=="7")] <- "4"
data_plot$cluster[which(data_plot$cluster=="8")] <- "3"
data_plot$phase <- factor(data_plot$phase)
data_plot$cluster <- factor(data_plot$cluster)


center_1 <- apply(data_plot[which(data_plot$cluster=="1"),1:2],2,mean)
center_2 <- apply(data_plot[which(data_plot$cluster=="2"),1:2],2,mean)
center_3 <- apply(data_plot[which(data_plot$cluster=="3"),1:2],2,mean)
center_4 <- apply(data_plot[which(data_plot$cluster=="4"),1:2],2,mean)

center <- as.data.frame(rbind(center_3,center_4))
colnames(center) <- c("x","y")
center2 <- as.data.frame(rbind(center_2,center_3))
colnames(center2) <- c("x","y")
center3 <- as.data.frame(rbind(center_2,center_1))
colnames(center3) <- c("x","y")

plot_1 <-  ggplot() +
    geom_point(data = data_plot, aes(x = PC1, y = PC2, color = cluster)) +
    stat_ellipse(data = data_plot, level = 0.9, linetype = 1, size = 2, 
                 aes(x = PC1, y = PC2, fill = cluster), alpha = 0.3, segments = 50, geom = "polygon",type = "t")  +
    theme_bw() +
    geom_line(data = center, aes(x = x, y = y), size = 1) +
    geom_line(data = center2, aes(x = x, y = y), size = 1) +
    geom_line(data = center3, aes(x = x, y = y), size = 1) +
    annotate("text", label = "1", x = center_1[1], y = center_1[2], size = 7, colour = "black") +
    annotate("text", label = "2", x = center_2[1], y = center_2[2], size = 7, colour = "black") +
    annotate("text", label = "3", x = center_3[1], y = center_3[2], size = 7, colour = "black") +
    annotate("text", label = "4", x = center_4[1], y = center_4[2], size = 7, colour = "black") +
    theme(legend.key.width = unit(0.5, "cm"))


plot_2 <-  ggplot(data_plot, aes(x=PC1,y=PC2))+   geom_point( aes(color=phase))+theme_bw()+ theme(legend.key.width = unit(0.5, "cm"))+scale_color_manual(values = c("other" = "#C0C0C0","s" = "#F8766D","g2" = "#00BA38","g1" = "#619CFF") )

plot_1
plot_2
```
### fisher test Bullseye
```{r,eval=FALSE}
G1_exp <- rep(0,length(data_plot$phase))
G1_exp[which(data_plot$phase=="g1")] <- 1
s_exp <- rep(0,length(data_plot$phase))
s_exp[which(data_plot$phase=="s")] <- 1
G2_exp <- rep(0,length(data_plot$phase))
G2_exp[which(data_plot$phase=="g2")] <- 1
other_exp <- rep(0,length(data_plot$phase))
other_exp[which(data_plot$phase=="other")] <- 1

exp_list <- list()
exp_list[[1]] <- G1_exp
exp_list[[2]] <- s_exp
exp_list[[3]] <- G2_exp
exp_list[[4]] <- other_exp

cluster_1_Bullseye <- rep(0,length(data_plot$cluster))
cluster_1_Bullseye[which(data_plot$cluster=="1")] <- 1
cluster_2_Bullseye <- rep(0,length(data_plot$cluster))
cluster_2_Bullseye[which(data_plot$cluster=="2")] <- 1
cluster_3_Bullseye <- rep(0,length(data_plot$cluster))
cluster_3_Bullseye[which(data_plot$cluster=="3")] <- 1
cluster_4_Bullseye <- rep(0,length(data_plot$cluster))
cluster_4_Bullseye[which(data_plot$cluster=="4")] <- 1

Bullseye_list <- list()
Bullseye_list[[1]] <- cluster_1_Bullseye
Bullseye_list[[2]] <- cluster_2_Bullseye
Bullseye_list[[3]] <- cluster_3_Bullseye
Bullseye_list[[4]] <- cluster_4_Bullseye

Bullseye_table_pvalue <- matrix(NA,4,4)
rownames(Bullseye_table_pvalue) <- c("G1","S","G2/M","unclear")
colnames(Bullseye_table_pvalue) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4")

Bullseye_table_OR <- matrix(NA,4,4)
rownames(Bullseye_table_pvalue) <- c("G1","S","G2/M","unclear")
colnames(Bullseye_table_pvalue) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4")

for (i in 1:4){
  for (j in 1: 4){
    a <- fisher.test(exp_list[[i]],Bullseye_list[[j]],alternative = "greater")
    Bullseye_table_pvalue[i,j] <- a$p.value
    Bullseye_table_OR[i,j] <- a[["estimate"]][["odds ratio"]]
  }
}

Bullseye_table_pvalue
Bullseye_table_OR
```

## vis
```{r,eval=FALSE}
SigRM_table_pvalue <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/SigRM_table_pvalue.rds")
DESeq2_table_pvalue <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/DESeq2_table_pvalue.rds")
edgeR_table_pvalue <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/edgeR_table_pvalue.rds")
Bullseye_table_pvalue <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/Bullseye_table_pvalue.rds")



data_cluster_log_mean <- c()
 
data_cluster_log_mean[1] <- mean(log(apply(SigRM_table_pvalue,2, min)))
data_cluster_log_mean[2] <- mean(log(apply(DESeq2_table_pvalue,2, min)))
data_cluster_log_mean[3] <- mean(log(apply(edgeR_table_pvalue,2, min)))
data_cluster_log_mean[4] <- mean(log(apply(Bullseye_table_pvalue,2, min)))

data_phases_log_mean <- c()
 
data_phases_log_mean[1] <- mean(log(apply(SigRM_table_pvalue,1, min)))
data_phases_log_mean[2] <- mean(log(apply(DESeq2_table_pvalue,1, min)))
data_phases_log_mean[3] <- mean(log(apply(edgeR_table_pvalue,1, min)))
data_phases_log_mean[4] <- mean(log(apply(Bullseye_table_pvalue,1, min)))


data_plot <- as.data.frame(cbind(as.numeric(c(data_cluster_log_mean,data_phases_log_mean)),rep(c("SigRM","DESeq2","edgeR","Bullseye"),2),rep(c("Clusters","Phases"),c(4,4)) ))

colnames(data_plot) <- c("Value","Methods","Dim")
data_plot$Value <- -as.numeric(data_plot$Value)

library(ggplot2)
plot <- ggplot(data=data_plot)+geom_bar(aes(x=Methods,y=Value,fill=Methods),stat="identity")+facet_wrap(Dim~.)+ylab("Performance")
```

### iM6A
## read the data and data set
```{r,eval=FALSE}
index_site <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/index_site.rds")

site_length <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/site_length.rds")



SNP_all_51w <- readRDS("D:/data/SMART/preprocess2/SNP_all_51w.rds")
chr_info <- rep(SNP_all_51w@seqnames@values,SNP_all_51w@seqnames@lengths)
chr_info_get <- chr_info[index_site]
chr_info_target <- factor(rep(c("chr2","chr20","chr4","chr2","chr1","chr22","chr17","chr5","chr20"),site_length),levels = levels(chr_info_get))

index_correct <- which(chr_info_get ==chr_info_target)
site_length_correct <- list()
start <- 1
end <- site_length[1]
for (i in 1:8){
  site_length_correct[[i]] <- length(which(index_correct%in%c(start:end)))
  start <- start+site_length[i]
  end <- end + site_length[i+1]
}
site_length_correct[[9]] <- length(which(index_correct%in%c(start:end)))
site_index_correct <- index_site [index_correct]

library(readr)
iM6A_humanRRACH <- read_csv("D:/data/iM6A/iM6A_humanRRACH/iM6A_humanRRACH.csv")



```


## create data frame SigRM
```{r,eval=FALSE}
order_list <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/order_list_SigRM.rds")
index_site_185 <- site_index_correct[unlist(order_list)]
names(index_site_185) <- NULL
strand <- SNP_all_51w$transcript_strand[index_site_185]
start_c <- unlist(SNP_all_51w@ranges@start[index_site_185])
start <- c()
for (i in 1:185){
  if(strand[i]=="+"|is.na(strand[i])){
    start[i] <- start_c[i]-1
  } else{
    start[i] <- start_c[i]+1
  }
}
end <- start+1
gene_name <- rep(c("MCM6","PCNA","SLBP","RRM2","DTL","MCM5","TOP2A","CCNB1","AURKA"),unlist(lapply(order_list,length)))



data_frame_site <- data.frame(cbind(index_site_185,strand,start_c,start,end,gene_name))


index_gene_name_match <- which(iM6A_humanRRACH$name %in%unique(gene_name))

iM6A_humanRRACH_gene_match <- iM6A_humanRRACH [index_gene_name_match,]


# our is hg38, iM6A is hg19
# we need to change hg38 to hg19


# for MCM6, the length of MCM6 is not change compare hg38 and hg19
# hg38 (135839626-135876443 ) hg19 (136597196-136634013) we need add 757570
data_frame_site$start_hg19 <- NA
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="MCM6")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="MCM6")]) + 757570

# for PCNA, the length of PCNA is not change compare hg38 and hg19
# hg38 (5114953-5126622) hg19 (5095599-5107268) we need minus 19354
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="PCNA")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="PCNA")]) - 19354

# for SLBP, the length of SLBP is not change compare hg38 and hg19
# hg38 (1692731-1712319) hg19 (1694458-1714046) we need add 1727
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="SLBP")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="SLBP")]) + 1727

# for RRM2, the length of RRM2 is change compare hg38 (88271) and hg19 (88270)
# hg38 (10122739-10211010 ) the 10160429 is the new site position
# hg19 (10262866-10351136 ) we need add 140127/140126
RRM2_vector <- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="RRM2")])
RRM2_vector_new <- c()
for (i in 1:length(which(data_frame_site$gene_name=="RRM2"))){
  if(RRM2_vector[i] < 10160429){
    RRM2_vector_new[i] <- RRM2_vector[i] + 140127
  } else if (RRM2_vector[i] < 10160429){
    RRM2_vector_new[i] <- NA
  } else{
    RRM2_vector_new[i] <- RRM2_vector[i] + 140126
  }
}

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="RRM2")]<-RRM2_vector_new
  


# for DTL, the length of DTL is change compare hg38 (38202) and hg19 (69265)
# hg38 (212066811-212105013 ) 
# hg19 (212209090 (212240153)-212278355 ) we need add 173342

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="DTL")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="DTL")]) + 173342


# for MCM5, the length of MCM5 is change compare hg38 (54891) and hg19 (25291)
# hg38 (35400140-35455031 ) 
# hg19 (35796133-35821424 35851024) we need add 395993

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="MCM5")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="MCM5")]) + 395993


# for TOP2A, the length of TOP2A is change compare hg38 (26390) and hg19 (29371)
# hg38 (40391506-40417896) 
# hg19 (38544777(38547758)-38574148) we need MINUS 1843748

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="TOP2A")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="TOP2A")]) -1843748


# for CCNB1, the length of CCNB1 is not change compare hg38 and hg19
# hg38 (69167150-69178245 ) hg19 (68462977-68474072 ) we need minus 704173
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="CCNB1")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="CCNB1")]) - 704173

# for AURKA, the length of AURKA is not change compare hg38 and hg19
# hg38 (56369390-56392215 ) hg19 (54944446-54967271 ) we need minus 1424944
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="AURKA")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="AURKA")]) - 1424944






TABLE_SigRM <- iM6A_humanRRACH_gene_match[which(iM6A_humanRRACH_gene_match$End%in%data_frame_site$start_hg19),]
```





## create data frame DESeq2
```{r,eval=FALSE}
order_list <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/order_list_DESeq2.rds")
index_site_185 <- site_index_correct[unlist(order_list)]
names(index_site_185) <- NULL
strand <- SNP_all_51w$transcript_strand[index_site_185]
start_c <- unlist(SNP_all_51w@ranges@start[index_site_185])
start <- c()
for (i in 1:185){
  if(strand[i]=="+"|is.na(strand[i])){
    start[i] <- start_c[i]-1
  } else{
    start[i] <- start_c[i]+1
  }
}
end <- start+1
gene_name <- rep(c("MCM6","PCNA","SLBP","RRM2","DTL","MCM5","TOP2A","CCNB1","AURKA"),unlist(lapply(order_list,length)))



data_frame_site <- data.frame(cbind(index_site_185,strand,start_c,start,end,gene_name))


index_gene_name_match <- which(iM6A_humanRRACH$name %in%unique(gene_name))

iM6A_humanRRACH_gene_match <- iM6A_humanRRACH [index_gene_name_match,]


# our is hg38, iM6A is hg19
# we need to change hg38 to hg19


# for MCM6, the length of MCM6 is not change compare hg38 and hg19
# hg38 (135839626-135876443 ) hg19 (136597196-136634013) we need add 757570
data_frame_site$start_hg19 <- NA
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="MCM6")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="MCM6")]) + 757570

# for PCNA, the length of PCNA is not change compare hg38 and hg19
# hg38 (5114953-5126622) hg19 (5095599-5107268) we need minus 19354
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="PCNA")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="PCNA")]) - 19354

# for SLBP, the length of SLBP is not change compare hg38 and hg19
# hg38 (1692731-1712319) hg19 (1694458-1714046) we need add 1727
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="SLBP")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="SLBP")]) + 1727

# for RRM2, the length of RRM2 is change compare hg38 (88271) and hg19 (88270)
# hg38 (10122739-10211010 ) the 10160429 is the new site position
# hg19 (10262866-10351136 ) we need add 140127/140126
RRM2_vector <- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="RRM2")])
RRM2_vector_new <- c()
for (i in 1:length(which(data_frame_site$gene_name=="RRM2"))){
  if(RRM2_vector[i] < 10160429){
    RRM2_vector_new[i] <- RRM2_vector[i] + 140127
  } else if (RRM2_vector[i] < 10160429){
    RRM2_vector_new[i] <- NA
  } else{
    RRM2_vector_new[i] <- RRM2_vector[i] + 140126
  }
}

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="RRM2")]<-RRM2_vector_new
  


# for DTL, the length of DTL is change compare hg38 (38202) and hg19 (69265)
# hg38 (212066811-212105013 ) 
# hg19 (212209090 (212240153)-212278355 ) we need add 173342

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="DTL")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="DTL")]) + 173342


# for MCM5, the length of MCM5 is change compare hg38 (54891) and hg19 (25291)
# hg38 (35400140-35455031 ) 
# hg19 (35796133-35821424 35851024) we need add 395993

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="MCM5")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="MCM5")]) + 395993


# for TOP2A, the length of TOP2A is change compare hg38 (26390) and hg19 (29371)
# hg38 (40391506-40417896) 
# hg19 (38544777(38547758)-38574148) we need MINUS 1843748

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="TOP2A")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="TOP2A")]) -1843748


# for CCNB1, the length of CCNB1 is not change compare hg38 and hg19
# hg38 (69167150-69178245 ) hg19 (68462977-68474072 ) we need minus 704173
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="CCNB1")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="CCNB1")]) - 704173

# for AURKA, the length of AURKA is not change compare hg38 and hg19
# hg38 (56369390-56392215 ) hg19 (54944446-54967271 ) we need minus 1424944
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="AURKA")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="AURKA")]) - 1424944






TABLE_DESeq2 <- iM6A_humanRRACH_gene_match[which(iM6A_humanRRACH_gene_match$End%in%data_frame_site$start_hg19),]
```



## create data frame SigRM
```{r,eval=FALSE}
order_list <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/order_list_edgeR.rds")
index_site_185 <- site_index_correct[unlist(order_list)]
names(index_site_185) <- NULL
strand <- SNP_all_51w$transcript_strand[index_site_185]
start_c <- unlist(SNP_all_51w@ranges@start[index_site_185])
start <- c()
for (i in 1:185){
  if(strand[i]=="+"|is.na(strand[i])){
    start[i] <- start_c[i]-1
  } else{
    start[i] <- start_c[i]+1
  }
}
end <- start+1
gene_name <- rep(c("MCM6","PCNA","SLBP","RRM2","DTL","MCM5","TOP2A","CCNB1","AURKA"),unlist(lapply(order_list,length)))



data_frame_site <- data.frame(cbind(index_site_185,strand,start_c,start,end,gene_name))


index_gene_name_match <- which(iM6A_humanRRACH$name %in%unique(gene_name))

iM6A_humanRRACH_gene_match <- iM6A_humanRRACH [index_gene_name_match,]


# our is hg38, iM6A is hg19
# we need to change hg38 to hg19


# for MCM6, the length of MCM6 is not change compare hg38 and hg19
# hg38 (135839626-135876443 ) hg19 (136597196-136634013) we need add 757570
data_frame_site$start_hg19 <- NA
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="MCM6")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="MCM6")]) + 757570

# for PCNA, the length of PCNA is not change compare hg38 and hg19
# hg38 (5114953-5126622) hg19 (5095599-5107268) we need minus 19354
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="PCNA")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="PCNA")]) - 19354

# for SLBP, the length of SLBP is not change compare hg38 and hg19
# hg38 (1692731-1712319) hg19 (1694458-1714046) we need add 1727
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="SLBP")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="SLBP")]) + 1727

# for RRM2, the length of RRM2 is change compare hg38 (88271) and hg19 (88270)
# hg38 (10122739-10211010 ) the 10160429 is the new site position
# hg19 (10262866-10351136 ) we need add 140127/140126
RRM2_vector <- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="RRM2")])
RRM2_vector_new <- c()
for (i in 1:length(which(data_frame_site$gene_name=="RRM2"))){
  if(RRM2_vector[i] < 10160429){
    RRM2_vector_new[i] <- RRM2_vector[i] + 140127
  } else if (RRM2_vector[i] < 10160429){
    RRM2_vector_new[i] <- NA
  } else{
    RRM2_vector_new[i] <- RRM2_vector[i] + 140126
  }
}

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="RRM2")]<-RRM2_vector_new
  


# for DTL, the length of DTL is change compare hg38 (38202) and hg19 (69265)
# hg38 (212066811-212105013 ) 
# hg19 (212209090 (212240153)-212278355 ) we need add 173342

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="DTL")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="DTL")]) + 173342


# for MCM5, the length of MCM5 is change compare hg38 (54891) and hg19 (25291)
# hg38 (35400140-35455031 ) 
# hg19 (35796133-35821424 35851024) we need add 395993

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="MCM5")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="MCM5")]) + 395993


# for TOP2A, the length of TOP2A is change compare hg38 (26390) and hg19 (29371)
# hg38 (40391506-40417896) 
# hg19 (38544777(38547758)-38574148) we need MINUS 1843748

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="TOP2A")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="TOP2A")]) -1843748


# for CCNB1, the length of CCNB1 is not change compare hg38 and hg19
# hg38 (69167150-69178245 ) hg19 (68462977-68474072 ) we need minus 704173
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="CCNB1")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="CCNB1")]) - 704173

# for AURKA, the length of AURKA is not change compare hg38 and hg19
# hg38 (56369390-56392215 ) hg19 (54944446-54967271 ) we need minus 1424944
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="AURKA")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="AURKA")]) - 1424944






TABLE_edgeR <- iM6A_humanRRACH_gene_match[which(iM6A_humanRRACH_gene_match$End%in%data_frame_site$start_hg19),]
```



## create data frame Bullseye
```{r,eval=FALSE}
order_list <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/order_list_Bullseye.rds")
index_site_185 <- site_index_correct[unlist(order_list)]
names(index_site_185) <- NULL
strand <- SNP_all_51w$transcript_strand[index_site_185]
start_c <- unlist(SNP_all_51w@ranges@start[index_site_185])
start <- c()
for (i in 1:185){
  if(strand[i]=="+"|is.na(strand[i])){
    start[i] <- start_c[i]-1
  } else{
    start[i] <- start_c[i]+1
  }
}
end <- start+1
gene_name <- rep(c("MCM6","PCNA","SLBP","RRM2","DTL","MCM5","TOP2A","CCNB1","AURKA"),unlist(lapply(order_list,length)))



data_frame_site <- data.frame(cbind(index_site_185,strand,start_c,start,end,gene_name))


index_gene_name_match <- which(iM6A_humanRRACH$name %in%unique(gene_name))

iM6A_humanRRACH_gene_match <- iM6A_humanRRACH [index_gene_name_match,]


# our is hg38, iM6A is hg19
# we need to change hg38 to hg19


# for MCM6, the length of MCM6 is not change compare hg38 and hg19
# hg38 (135839626-135876443 ) hg19 (136597196-136634013) we need add 757570
data_frame_site$start_hg19 <- NA
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="MCM6")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="MCM6")]) + 757570

# for PCNA, the length of PCNA is not change compare hg38 and hg19
# hg38 (5114953-5126622) hg19 (5095599-5107268) we need minus 19354
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="PCNA")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="PCNA")]) - 19354

# for SLBP, the length of SLBP is not change compare hg38 and hg19
# hg38 (1692731-1712319) hg19 (1694458-1714046) we need add 1727
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="SLBP")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="SLBP")]) + 1727

# for RRM2, the length of RRM2 is change compare hg38 (88271) and hg19 (88270)
# hg38 (10122739-10211010 ) the 10160429 is the new site position
# hg19 (10262866-10351136 ) we need add 140127/140126
RRM2_vector <- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="RRM2")])
RRM2_vector_new <- c()
for (i in 1:length(which(data_frame_site$gene_name=="RRM2"))){
  if(RRM2_vector[i] < 10160429){
    RRM2_vector_new[i] <- RRM2_vector[i] + 140127
  } else if (RRM2_vector[i] < 10160429){
    RRM2_vector_new[i] <- NA
  } else{
    RRM2_vector_new[i] <- RRM2_vector[i] + 140126
  }
}

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="RRM2")]<-RRM2_vector_new
  


# for DTL, the length of DTL is change compare hg38 (38202) and hg19 (69265)
# hg38 (212066811-212105013 ) 
# hg19 (212209090 (212240153)-212278355 ) we need add 173342

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="DTL")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="DTL")]) + 173342


# for MCM5, the length of MCM5 is change compare hg38 (54891) and hg19 (25291)
# hg38 (35400140-35455031 ) 
# hg19 (35796133-35821424 35851024) we need add 395993

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="MCM5")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="MCM5")]) + 395993


# for TOP2A, the length of TOP2A is change compare hg38 (26390) and hg19 (29371)
# hg38 (40391506-40417896) 
# hg19 (38544777(38547758)-38574148) we need MINUS 1843748

data_frame_site$start_hg19 [which(data_frame_site$gene_name=="TOP2A")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="TOP2A")]) -1843748


# for CCNB1, the length of CCNB1 is not change compare hg38 and hg19
# hg38 (69167150-69178245 ) hg19 (68462977-68474072 ) we need minus 704173
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="CCNB1")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="CCNB1")]) - 704173

# for AURKA, the length of AURKA is not change compare hg38 and hg19
# hg38 (56369390-56392215 ) hg19 (54944446-54967271 ) we need minus 1424944
data_frame_site$start_hg19 [which(data_frame_site$gene_name=="AURKA")]<- as.numeric(data_frame_site$start[which(data_frame_site$gene_name=="AURKA")]) - 1424944






TABLE_Bullseye <- iM6A_humanRRACH_gene_match[which(iM6A_humanRRACH_gene_match$End%in%data_frame_site$start_hg19),]
```





#  Regulatory inference from single-cell epitranscriptomics
## 5 GENE TPM
```{r,eval=FALSE}

expression_TPM <- readRDS("D:/data/SMART/preprocess2/expression_TPM.rds")
# METTL3 ENSG00000165819
# METTL14 ENSG00000145388
# WTAP ENSG00000146457
# FTO ENSG00000140718
# ALKBH5 ENSG00000091542
index <- which(rownames(expression_TPM)%in%c("ENSG00000165819","ENSG00000145388","ENSG00000146457","ENSG00000140718","ENSG00000091542"))



expression_5genes_TPM <- expression_TPM[index[c(1,4,5,2,3)],392:1382]

saveRDS(expression_5genes_TPM,"expression_TPM_5_GENE.rds")
```

## 9 gene site OR
```{r,eval=FALSE}


site_index_correct <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/site_index_correct.rds")

site_length_correct <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/site_length_correct.rds")

res_list <- readRDS("D:/data/SMART/results_ks/res_all_4_sim_SigRM/res_list.rds")

# all 780 sites
res_all_OR <- as.data.frame(lapply(res_list,function(df) df[[4]])) 
res_all_OR_index_site <- res_all_OR[site_index_correct,]

gene_from_780 <- rep(c("MCM6","PCNA","SLBP","RRM2","DTL","MCM5","TOP2A","CCNB1","AURKA"),unlist(site_length_correct))

# 185 sites using in case study 1

order_list_SigRM <- readRDS("D:/phd/project/SIGMR_single_cell/cell press review/real_data_ks/trajectory/order_list_SigRM.rds")

index_site <- site_index_correct[unlist(order_list_SigRM)]
gene_from_185 <- rep(c("MCM6","PCNA","SLBP","RRM2","DTL","MCM5","TOP2A","CCNB1","AURKA"),unlist(lapply(order_list_SigRM,length)))

res_all_OR_index_site <- res_all_OR[index_site,]

saveRDS(res_all_OR_index_site,"res_all_OR_185.rds")
saveRDS(gene_from_185,"gene_info_185.rds")

```



```{r,eval=FALSE}
library(rtracklayer)
library(GenomicAlignments)
```


```{r,eval=FALSE}
#read new updated data:OR

#185  OR>1
#check the files
order_list<-readRDS("~/GRN/revised/order_list_SigRM.rds")
head(order_list)
length(order_list) #9 gene 
length(unlist(order_list)) #185
gene_info_185<-readRDS("~/GRN/revised/gene_info_185.rds")
length(gene_info_185)  #185
unique(gene_info_185)  # nine genes
res_all_OR_185<-readRDS("~/GRN/revised/res_all_OR_185.rds")
res_all_OR_185%>%dim()  #185 991
colnames(res_all_OR_185)<-col_name
rownames(res_all_OR_185)<-paste0('site_',rownames(res_all_OR_185))
#make sure all the values larger than 0.
res_all_OR_185[res_all_OR_185<0]=0


#marker enzyme information
enzyme<-c("METTL3", "METTL14", "WTAP","FTO","ALKBH5")  
enzyme_ENSG <- bitr(enzyme, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")
#rownames:  METTL3  METTL14	 WTAP  FTO	 ALKBH5
enzyme_ENSG #check ensemble and gene names 

#read expression data
express_TPM_5gene<-readRDS("~/GRN/revised/expression_TPM_5_GENE.rds")
express_TPM_5gene
colnames(express_TPM_5gene)<-col_name

```
Revised Correlation:calculate correlation between gene expression level and site methylation level (185)
```{r,eval=FALSE}
#find correlation between gene expression level and site methylation level
library(psych)
type_enzyme<-c("writer","writer","writer","eraser","eraser")
annotation_enzyme = data.frame(
  geneType = factor(type_enzyme))
rownames(annotation_enzyme)<-t(express_TPM_5gene)%>%colnames()
cor.result.enzyme_re <- corr.test(t(res_all_OR_185),t(express_TPM_5gene),method = "spearman", adjust = "BH") 
pheatmap::pheatmap(na.omit(cor.result.enzyme_re$r),show_rownames = F,color = colorRampPalette(c("navy", "white", "firebrick3"))(50),annotation_col=annotation_enzyme)
r.enzyme.re <- cor.result.enzyme_re$r     #相关性


flattenCorrMatrix.enzyme<-function(cormat,pmat){
  
  df<-reshape2::melt(cormat)
  index<-order(df$value,decreasing = T)
  p_df<-reshape2::melt(pmat)
  df<-cbind(df,p_df$value)[index,]
  colnames(df)<-c('site','enzyme','cor','pval')
  return(df)
  }

res_cor_enzyme_re<-flattenCorrMatrix.enzyme(cor.result.enzyme_re$r,cor.result.enzyme_re$p)

res_cor_enzyme_re%>%arrange(enzyme,pval>0,desc(cor))

#subset(res_cor_enzyme_re,enzyme=="ENSG00000165819 "&pval<1e-5)
gene_df_enzyme_re<-data.frame(site=NA,enzyme=NA,cor=NA,pval=NA)
for(i in unique(res_cor_enzyme_re$enzyme)){
  res_df<-subset(res_cor_enzyme_re,enzyme==i&pval<0.01) #select p cutoff
  gene_df_enzyme_re<-rbind(gene_df_enzyme_re,res_df)
}
gene_df_enzyme_re<-gene_df_enzyme_re[-1,]
gene_df_enzyme_re
saveRDS(gene_df_enzyme_re,'~/GRN/revised/gene_df_enzyme_re.rds')
```




revised GENIE3 (185 sites)
```{r,eval=FALSE}
#deal with input OR matrix  185 (input in GENIE)
OR_enzyme_re_df<-rbind(res_all_OR_185,express_TPM_5gene)
tail(OR_enzyme_re_df)
OR_enzyme_re_df[is.na(OR_enzyme_re_df)]<-0
library(preprocessCore)
#quantile normization based on row (site level)
OR_enzyme_re_combine<-t(normalize.quantiles(t(x=as.matrix(OR_enzyme_re_df))))
colnames(OR_enzyme_re_combine)<-colnames(OR_enzyme_re_df)
rownames(OR_enzyme_re_combine)<-rownames(OR_enzyme_re_df)
saveRDS(OR_enzyme_re_combine,"~/GRN/revised/OR_enzyme_re_combine.rds")
#write.csv(OR_enzyme_re_combine,"~/GRN/revised/OR_enzyme_re_combine.csv")


library(GENIE3)
OR_genie_enz_re<-GENIE3(na.omit(OR_enzyme_re_combine))
re.enz.revised<-data.frame(OR_genie_enz_re[c(186:190),c(1:185)])

#result 
re.enz.revised%>%dim
saveRDS(re.enz.revised,"~/GRN/revised/re.enz.revised.rds")
write.csv(re.enz.revised,"~/GRN/revised/re.enz.revised.csv")

```

rank the GENIE3 results （185）
```{r,eval=FALSE}
re.enz.revised<-readRDS("~/GRN/revised/GENIE/re.enz.revised.rds")
# find the site with the highest weight 
# deal with row_handler
row_handler <- function(or.data,num){  
 # weight<-as.list(as.data.frame(t(or.data)))
  index<-order(or.data,decreasing = T)[1:num]
 # index <- which(or.data == max(or.data))  # 找出最大的元素的index
  #out <- paste0(names(or.data[index])," : ",or.data[index])  # 从index还原成列名
  site<-names(or.data[index])
  #out <- data.frame(or.data[index])
  out <- or.data[index]
 # names(out)<-names(or.data)[index]
#  colnames(out)<-rownames(or.data)[index]
  return(list(out))
}

# print for each row
res.lst<-function(res,num=10){
  re.out<-apply(res,1,row_handler,num)
  return(re.out)
}
re.out.enz.revised<-res.lst(re.enz.revised,num=15)

re.out.enz.revised

#change the result format
#make it accomadate to/from format 
get.link.list <- function(weight.lst, report.max=NULL) {
  # set negative weights (for permutation of OOB importance) to 0.0 
  # weight.lst[weight.lst < 0.0] <- 0.0
  num.sites<-  length(weight.lst[[1]][[1]])
  num.genes<-  length(weight.lst)
  sites.name <- sapply(weight.lst,function(x){x[[1]]%>%names()})
  wight<-sapply(weight.lst,function(x){x[[1]]})
   if (!is.null(report.max) && report.max < num.sites) {
    genes.name<-sapply(sapply(weight.lst,function(x){x[[1]]})%>%colnames,rep,report.max)
   }else{
    genes.name<-sapply(sapply(weight.lst,function(x){x[[1]]})%>%colnames,rep,num.sites)
  }
  list.length <- num.genes * num.sites
  if (!is.null(report.max) && report.max < num.sites) {
    list.length <- report.max*num.genes
  }
  # setup link list
  link.list <- data.frame(from.gene=rep("", list.length),
                          to.site=rep("", list.length),
                          im=rep(0.0, list.length),
                          stringsAsFactors=FALSE)
  sorted.indices <- sapply(weight.lst,function(x){x[[1]]})%>%order(decreasing=TRUE)
  # fill in link list
  index.number <- 1
  link.number <- 1
  while (index.number  <= list.length) {
    i <- sorted.indices[index.number]
    im <- wight[i]
    #row.col <- lin.to.square(i, num.sites)
    #row <- row.col[1]
    #col <- row.col[2]
    # Only process weights off-diagonal
    #if (row != col) {
      from.gene <- genes.name[i]
      to.site <- sites.name[i]
      link.list[link.number,] <- list(from.gene, to.site, im)
      link.number <- link.number + 1
      index.number <-index.number + 1
    #}
    #index.number <-index.number + 1
}
  return(link.list)
  }


res.enz.link.revised<-get.link.list(re.out.enz.revised)

res.enz.link.revised
```

Process the coupledNMF input data （185）
```{r,eval=FALSE}
##filter and build sparse matrix

#site index whole site
#find correlation between gene expression level and site methylation level/ select all sites
r.enzyme.re[is.na(r.enzyme.re)]<-0
r.enzyme.re[r.enzyme.re<0] <- 0
matrix_enzyme_re_S<-round(r.enzyme.re,4)
dim(matrix_enzyme_re_S) # 185 5
#for python cnmf
#matrix_re_enz_exp<-exp(matrix_enzyme_re_S)
write.table(t(matrix_enzyme_re_S),"~/GRN/revised/NMF/sparse_matrix_enz_re.txt",row.names = F, col.names = F, sep ="\t")

#name python/R
genenames_nmf_enz_re<-matrix_enzyme_re_S%>%colnames
sitenames_nmf_enz_re<-matrix_enzyme_re_S%>%rownames
write.table(genenames_nmf_enz_re,"~/GRN/revised/NMF/genenames_nmf_enz_re.txt")
write.table(sitenames_nmf_enz_re,"~/GRN/revised/NMF/sitenames_nmf_enz_re.txt")

#site OR matrix filtered site
OR_re_nmf<-(as.matrix(res_all_OR_185))
OR_re_nmf[is.na(OR_re_nmf)]<-0
OR_re_nmf<-round(OR_re_nmf,4)
OR_re_nmf[OR_re_nmf < 0]<-0
dim(OR_re_nmf)# 185 991
write.table(OR_re_nmf,"~/GRN/revised/NMF/OR_re_nmf.txt",row.names = F, col.names = F, sep ="\t")

#eexpress input
express_enzyme_re_nmf<-as.matrix(express_TPM_5gene)
express_enzyme_re_nmf[is.na(express_enzyme_re_nmf)]<-0
express_enzyme_re_nmf<-round(express_enzyme_re_nmf,4)
express_enzyme_re_nmf[express_enzyme_re_nmf < 0]<-0
write.table(express_enzyme_re_nmf,"~/GRN/revised/NMF/express_enzyme_re_nmf.txt",row.names = F, col.names = F, sep ="\t")
```

SELECT 185 OR
```{r,eval=FALSE}
##genie3
res.enz.link.revised

genie_site<-gene_info_185[match(res.enz.link.revised$to.site,as.factor(rownames(res_all_OR_185)))]
genie_enz<-enzyme_ENSG$SYMBOL[match(res.enz.link.revised$from.gene,enzyme_ENSG$ENSEMBL)]
res.enz.link.revised$gene_name<-genie_enz
res.enz.link.revised$site_name<-genie_site
as.data.frame(res.enz.link.revised)
write.table(as.data.frame(res.enz.link.revised),'~/GRN/revised/res.enz.link.revised.csv', row.names = FALSE,quote=F,sep = ",")
#cor
gene_df_enzyme_re
cor_site<-gene_info_185[match(gene_df_enzyme_re$site,as.factor(rownames(res_all_OR_185)))]
cor_enz<-enzyme_ENSG$SYMBOL[match(gene_df_enzyme_re$enzyme,enzyme_ENSG$ENSEMBL)]
gene_df_enzyme_re$gene_name<-cor_enz
gene_df_enzyme_re$site_name<-cor_site
write.table(as.data.frame(gene_df_enzyme_re),'~/GRN/revised/gene_df_enzyme.csv', row.names = FALSE,quote=F,sep = ",")

##cnmf
res_cnmf_re2_sel%>%dim

write.table(as.data.frame(res_cnmf_re2_sel),'~/GRN/revised/res_cnmf_re2_sel.csv', row.names = FALSE,quote=F,sep = ",")
#cnmf and genie
res_cnmf_re2_df[res_cnmf_re2_df$to.site%in%res.enz.link.revised$to.site,]
 res.enz.link.revised[res.enz.link.revised$to.site%in%res_cnmf_re2_df$to.site,]
res.enz.link.revised[which(res.enz.link.revised$gene_name=='METTL14'),]
#cnmf and cor
res_cnmf_re2_df[res_cnmf_re2_df$to.site%in%gene_df_enzyme_re$site,]
gene_df_enzyme_re[gene_df_enzyme_re$site%in%res_cnmf_re2_df$to.site,]

#GENIE and cor
res.enz.link.revised[res.enz.link.revised$to.site%in%gene_df_enzyme_re$site,]
gene_df_enzyme_re[gene_df_enzyme_re$site%in%res.enz.link.revised$to.site,]


#GENIE and cnmf
res.enz.link.revised[res.enz.link.revised$to.site%in%res_cnmf_re2_df$to.site,]
res_cnmf_re2_df[res_cnmf_re2_df$to.site%in%res.enz.link.revised$to.site,]

```


```{r,eval=FALSE}
##read cnmf results
res_cnmf_re2<-read.csv2('~/GRN/revised/NMF/results_180/cluster-specific-sites-genes-pairs.txt',sep="\t",header = F)
res_cnmf_re2
cluster_re2 <-lapply(strsplit(res_cnmf_re2$V1,":"), function(x) substr(x[1],1,9))
gene_res_re2<- lapply(strsplit(res_cnmf_re2$V1,":"), function(x) substr(x[2],4,18))
site_res_re2<-lapply(strsplit(res_cnmf_re2$V2," "), function(x) substr(x[2],1,15))
res_cnmf_re2_df<-data.frame(cluster=unlist(cluster_re2)%>%as.factor(),
                           from.gene=unlist(gene_res_re2)%>%as.factor(),
                           to.site=unlist(site_res_re2)%>%as.factor(),
                           gene_pval=unlist(res_cnmf_re2$V3)%>%as.numeric(),
                           site_pval=unlist(res_cnmf_re2$V4%>%as.numeric()))
res_cnmf_re2_df%>%head
idx_nmf_re2<-substr(res_cnmf_re2_df$to.site,6,12)%>%as.factor()

#res_cnmf_filter_df1<-data.frame(cluster=unlist(cluster_filter)%>%as.factor(),
#                               from.gene=unlist(gene_res_filter)%>%as.factor(),
#                              to.site=unlist(site_res_filter)%>%as.factor(),
#                             gene_pval=unlist(res_cnmf_filter$V3)%>%as.numeric(),
#                            site_pval=unlist(res_cnmf_filter$V4%>%as.numeric()))

##add gene name
cnmf_gene_re2<-gene_info_185[match(res_cnmf_re2_df$to.site,as.factor(rownames(res_all_OR_185)))]
cnmf_gene_re2%>%length
res_cnmf_re2_df$site_gene<-cnmf_gene_re2

res_cnmf_re2_df$to.site<-paste0(res_cnmf_re2_df$to.site,' ' ,cnmf_gene_re2)

res_cnmf_re2_df$from.geneID<-enzyme_ENSG$SYMBOL[match(res_cnmf_re2_df$from.gene,enzyme_ENSG$ENSEMBL)]
library("tidygraph")
library("ggraph")
library("igraph")

res_cnmf_re2_sel<-res_cnmf_re2_df[which(res_cnmf_re2_df$site_pval<1e-10),]
res_cnmf_re2_sel%>%dim
#  gene去重
gene_nmf_re2 <- res_cnmf_re2_sel %>%distinct(from.geneID) %>%rename(label = from.geneID)

#site去重
site_nmf_re2 <- res_cnmf_re2_sel %>%
  distinct(to.site) %>%
  rename(label = to.site)

## 合并数据并添加一列索引
nodes_nmf_re2 <- full_join(gene_nmf_re2, site_nmf_re2, by = "label") 
nodes_nmf_re2 <- nodes_nmf_re2 %>%
  mutate(id = 1:nrow(nodes_nmf_re2)) %>%
  select(id, everything())
head(nodes_nmf_re2, 3)

# Rename the n.call column to weight
#res_cnmf_re_sel <- res_cnmf_re_sel %>%
# rename(pval = pval)

# (a) Join nodes id for source column
edges_nmf_re2 <- res_cnmf_re2_sel %>% 
  left_join(nodes_nmf_re2, by = c("from.geneID" = "label")) %>% 
  rename(from = id)

# (b) Join nodes id for destination column
edges_nmf_re2 <- edges_nmf_re2 %>% 
  left_join(nodes_nmf_re2, by = c("to.site" = "label")) %>% 
  rename(to = id)

# (c) Select/keep only the columns from and to
edges_nmf_re2 <- select(edges_nmf_re2, from, to)

library(igraph)
set.seed(123)
net.igraph.nmf.re2 <- graph_from_data_frame(
  d = edges_nmf_re2, vertices = nodes_nmf_re2, 
  directed = TRUE
)

#在我们的过滤中，许多基因与其他基因没有任何的链接，可能是没有什么调控关系，我们想看哪些基因是有调控关系的。
plot(induced_subgraph(net.igraph.nmf.re2,  V(net.igraph.nmf.re2)[igraph::degree(net.igraph.nmf.re2)>0] ))
plot(net.igraph.nmf.re2, edge.arrow.size = 0.2,
     layout = layout_with_graphopt)
#调色

#other plot


edges_nmf_re2_view<-res_cnmf_re2_sel %>% 
  left_join(nodes_nmf_re2, by = c("from.geneID" = "label")) %>% 
  rename(from = id)%>% 
  left_join(nodes_nmf_re2, by = c("to.site" = "label")) %>% 
  rename(to = id)
edges_nmf_re2_view <- select(edges_nmf_re2_view, from, to, site_pval)
colnames(edges_nmf_re2_view)<-c('from','to','length')
edges_nmf_re2_view$length<- (-log2(edges_nmf_re2_view$length))*2

edges_nmf_re2_weight<-edges_nmf_re2_view%>%na.omit

nodes_nmf_re2_view<-nodes_nmf_re2
nodes_nmf_re2_view$label<-cnmf_gene_re2
nodes_nmf_re2_view$label
library(visNetwork)

nodes_nmf_re2_view$color<-c(rep(mycolor_nodes_re2[1],5),mycolor_nodes_re2[c(2:(length(nodes_nmf_re2$id)-5+1))])
nodes_nmf_re2_view$color<-colorRampPalette((pal_npg("nrc")(5)))(23)

mycolor_nodes_re2<-pal_npg("nrc")(5)
nodes_nmf_re2_view$color[1:5]<-mycolor_nodes_re2[1]
nodes_nmf_re2_view$color[c(6:(length(nodes_nmf_re2_view$id)))]<-NA
visNetwork(nodes_nmf_re2_view, edges_nmf_re2_view) 

#net plot
# 设置基因的颜色
#"#E64B35" "#4DBBD5" "#00A087" "#3C5488" "#FFCA28" "#8491B4"
V(net.igraph.nmf.re2)$color <- c("#E64B35","#6BA4B5", "#00A087", "#4DBBD5","#606286", rep("#FFCA28",23-5))

# 假设我们有基因之间的相关性，用于设置边的宽度
E(net.igraph.nmf.re2)$width <- abs(rnorm(n = ecount(net.igraph.nmf.re2)))
# 设置箭头大小和边的颜色
E(net.igraph.nmf.re2)$arrow.size <- .08
E(net.igraph.nmf.re2)$edge.color <- "gray80"
  # 设置布局
graph_attr(net.igraph.nmf.re2, "layout") <- layout_with_lgl
deg <-degree(net.igraph.nmf.re2, mode="all")
V(net.igraph.nmf.re2)$size <- (log(deg)+1) *6


plot(net.igraph.nmf.re2,layout = layout_on_grid)

#heat plot

cnmf_heat_re2<-ggplot(data = res_cnmf_re2_df, aes(x=from.gene, y=to.site, fill= site_pval))+
  geom_point(size = 3, shape = 21) +
  theme(panel.background =  element_blank())+scale_fill_gradient2(low="darkred",mid="white",
                                                                  high = 'darkgreen')

cnmf_heat_re2
#相关位点的共同markergene数目条形图
res_cnmf_re2_sel%>%group_by(from.gene)%>%count
cnmf_site_re2<-ggplot(res_cnmf_re2_sel) +
  geom_bar(aes(y = to.site, fill = from.gene)) +
  scale_x_continuous(position = "top", breaks = seq(0, 6, 2),
                     limits = c(0, 6),
                     expand = expansion(mult = 0, add = 0)) +
  theme(
    legend.position = "none",
    panel.background = element_blank(),
    axis.title = element_blank(),
    axis.ticks.y = element_blank(),
    axis.title.y = element_blank(),
    axis.text.y = element_blank(),
    axis.ticks.length.x.top = unit(.25, "cm"),
    axis.line.x.top = element_line(colour = "black")
  )+scale_fill_npg()

# marker gene的相关位点数目条形图

cnmf_gene_re2<-ggplot(res_cnmf_re2_sel
) +
  geom_bar(aes(x = from.gene,color=from.gene,fill=from.gene)) +
  scale_y_continuous(breaks = seq(0, 20, 5),
                     limits = c(0, 20),
                     expand = expansion(mult = 0, add = 0)) +
  theme(
    legend.position = "right",
    panel.background = element_blank(),
    axis.title = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.length.y.left = unit(.25, "cm"),
    axis.line.y.left = element_line(colour = "black"),
  )+scale_color_npg()+scale_fill_npg()
#
cnmf_gene_re2
library(cowplot)
library(ggsci)
p12_cnmf <- plot_grid(cnmf_gene_re2, cnmf_heat_re2, nrow = 2, align = "v",
                      rel_heights = c(2, 6))

pr_cnmf <- plot_grid(NULL, cnmf_site_re2, NULL, nrow = 3, align = "v",
                     rel_heights = c(4.8, 17, 1.05))

plot_grid(p12_cnmf, pr_cnmf, ncol = 2, align = "h",
          rel_widths = c(5, 1))

```
the results of the case study can be found in : https://drive.google.com/drive/folders/1D137_357nuVYkxYiwcxH8dNyKsEbYrUv?usp=sharing
