---
title: "<center> SigRM Vignette </center>"
output: 
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
abstract: |
      This Vignette to show the example to use the main function of the package, and demostrate the simulation part.\n <center> **Table of content** </center>
vignette: >
  % \VignetteIndexEntry{SigRM_vignette}
  % \VignetteEngine{knitr::rmarkdown}
  % \VignetteEncoding{UTF-8}
  % \VignetteDepends{pROC} 
  % \VignetteDepends{Seurat} 
  % \VignetteDepends{DESeq2} 
  % \VignetteDepends{edgeR} 
  % \VignetteDepends{latex2exp} 
  % \VignetteDepends{statmod}
---

```{r, include = FALSE}
knitr::opts_chunk$set(fig.width=6, fig.height=4,
  collapse = TRUE,
  comment = "#>"
)
```






# Install the package

At the beginning, we need to install the SigRM package and some other packages required for this R Markdown file.

## Install our package SigRM
```{r setup}
# install.packages("devtools")
# library("devtools")
# install_github("whz991026/SigRM")

library(SigRM)
```



## Install the library need for this r markdown file

Load the other packages:

* library(pROC)
* library(Seurat)
* library(ggplot2)
* library(DESeq2)
* library(edgeR)
* library(latex2exp)
```{r, include=FALSE}




# install.packages("pROC")
library(pROC)



#    install.packages("Seurat")

library(Seurat)


#    install.packages("ggplot2")

library(ggplot2)


#    if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("DESeq2")
library(DESeq2)


#    if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("edgeR")
library(edgeR)


library(latex2exp)
library(statmod)
```

# Quick start

It demonstrates how to utilize the main function, SigRMtest(), from the SigRM package. The primary input for the SigRMtest function is a matrix or data frame containing four sets of read counts, where rows represent genes, and columns represent single cells. These four sets of read counts are ordered as follows: methylation read counts in the control group, methylation read counts in the test group, unmethylation read counts in the control group, and unmethylation read counts in the test group. This data can be obtained by calling the simulateData() function.

The output of the SigRMtest function is a list with a length of 7:

* The first part consists of a data frame representing methylation proportions (with rows for genes and columns for single cells) in the test cells.
* The second part comprises a mean methylation proportion vector (with rows for genes and columns for single cells) in the control cells.
* The third part contains a data frame displaying log2 risk ratios for the test cells.
* The fourth part includes a data frame illustrating log2 odds ratios for the test cells.
* The fifth part encompasses a data frame containing p-values for the test cells.
* The sixth part represents estimated gene abundances.
* The seventh part is a data frame of adjusted p-values for the test cells. 


```{r,warning=FALSE}
set.seed(1)
# first simulate the data
data <- simulateData(n_Sites=2000,test_num = 10,control_num = 10)
# put into the main function SigRMtest
res <- SigRMtest(data[[1]],data[[2]],data[[3]],data[[4]])

# meth proportion data frame (row is gene and column is single cell) in the test cells. 
head(res[[1]])

# mean meth proportion vector (row is gene and column is single cell) in the control cells.
head(res[[2]])

# log2 risk ratio data frame for the test cells. 
head(res[[3]])

# log2 odds ratio data frame for the test cells. 
head(res[[4]])

# p value data frame for the test cells. 
head(res[[5]])

# estimated gene abundance
head(res[[6]])

# adjusted p value data frame for the test cells
head(res[[7]])
```


# Simulation Part

This section involves simulating data and evaluating the performance of the SigRM model compared to the original method. It also includes testing the performance of the SigRM_cluster_test() or SigRM_similarity_test() and SigRMtest() functions to determine whether selecting control cells based on the similarity of their expression profiles yields superior results.

For the simulation data:

* There are 5000 sites.
* The dataset comprises 100 control cells and 100 test cells.
* The logarithm of $q_{ij}$ follows a uniform distribution.
* The detection level $p_{ij}$ also follows a uniform distribution.
* For 50% of the background sites, the detection levels follow a beta distribution and the mean level is (0, 0.4), while the corresponding test data follow a $U(0.6, 0.9)$ distribution.
* For the remaining sites, their detection levels are similar, following a a beta distribution and the mean level is (0, 0.4). 

By the way, the simulateData() function is responsible for generating the dataset, as before, and provides an output list containing methylation read counts in the control group, methylation read counts in the test group, unmethylation read counts in the control group, and unmethylation read counts in the test group.


## Function for the simulation part
### Function in the ScDART paper written by Tegowski et al. 
The details can be found at https://www.cell.com/molecular-cell/pdf/S1097-2765(21)01143-6.pdf. In the paper, two platforms are discussed: the droplet-based 10x Genomics and the plate-based SMART-seq2 platforms. These platforms employ slightly different methods for detecting m6A modifications. Subsequently, we have developed two functions for these two methods, namely, single_test_10x and single_test_SMART.

The single_test_10x() function takes the following inputs: methylation read counts of the control data, methylation read counts of the test data, unmethylation read counts of the control data, and unmethylation read counts of the test data. This method is tailored for 10x Genomics sample preparation and involves the following criteria:

* The test data should have a minimum coverage of 3 reads.
* The detection level of the test data (t/T) should be greater than or equal to 10%.
* The detection level of the test data should be at least 1.5-fold higher than that of the control data.


Conversely, the single_test_SMART() function, designed for SMART-seq2 sample preparation and sequencing, takes similar inputs: methylation read counts of the control data, methylation read counts of the test data, unmethylation read counts of the control data, and unmethylation read counts of the test data. This method applies the following criteria:

* The test data should have a minimum coverage of 20 reads.
* The detection level of the test data (t/T) should be greater than or equal to 10%.
* The detection level of the test data should be at least 1.5-fold higher than that of the control data.
* The detected read counts should be greater than or equal to 2.



```{r,warning=FALSE}



single_test_10x <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  # combine as the total reads
  data1 <- meth_control+unmeth_control
  data2 <- meth_test+unmeth_test
  
  # calculate the mean p and the variance of the control data
  
  p_control <- apply((meth_control/data1) , 1, mean_na<- function(x) {mean(x,na.rm=TRUE)})
  var_control <- apply((meth_control/data1) , 1, var_na<- function(x){var(x,na.rm=TRUE)})
  
  # find the site satisfies the three conditions
  if (length(dim(meth_test)[2])==0){
     # first condition
      index_1 <- which(data2>=3)
      
      # second condition
      p_test <- meth_test/data2
      index_2 <- which(p_test>=0.1)
      index_2 <- index_2[index_2 %in% index_1] 
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index <- index_3 [index_3 %in% index_2]
  }else{
    index <- list()
    for (i in 1 : dim(meth_test)[2]){
      
      # first condition
      index_1 <- which(data2[,i]>=3)
      
      # second condition
      p_test <- meth_test[,i]/data2[,i]
      index_2 <- which(p_test>=0.1)
      index_2 <- index_2[index_2 %in% index_1]
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index[[i]] <- index_3[index_3 %in% index_2]
    }
    
  }
    
    
      return(index)
  
}
    


# put into the SMART model
single_test_SMART <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
  # combine as the total reads
  data1 <- meth_control+unmeth_control
  data2 <- meth_test+unmeth_test
  
  # calculate the mean p and the variance of the control data
  
  p_control <- apply((meth_control/data1) , 1, mean_na<- function(x) {mean(x,na.rm=TRUE)})
  var_control <- apply((meth_control/data1) , 1, var_na<- function(x){var(x,na.rm=TRUE)})
  
  # find the site satisfies the three conditions
  if (length(dim(meth_test)[2])==0){
     # first condition
      index_1 <- which(data2>=20)
      
      # second condition
      p_test <- meth_test/data2
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1] 
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test>=2)
      index <- index_4 [index_4 %in% index_3]
  }else{
    index <- list()
    for (i in 1 : dim(meth_test)[2]){
      
      # first condition
      index_1 <- which(data2[,i]>=20)
      
      # second condition
      p_test <- meth_test[,i]/data2[,i]
      index_2 <- which(p_test>=0.1&p_test<=0.95)
      index_2 <- index_2[index_2 %in% index_1]
      
      
      #third condition
      index_3 <- which(p_test>=1.5*p_control)
      index_3 <- index_3 [index_3 %in% index_2]
      
      #fourth condition
      index_4 <- which(meth_test[,i]>=2)
      index[[i]] <- index_4 [index_4 %in% index_3]
    }
    
    
    
  }
  
  
  return(index)
  
}





```

### DESeq2 function
DESeq2_test() function is the function apply DESeq2 package to detect the methylated sites.
```{r,warning=FALSE}
DESeq2_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  meth_control <- data.frame(meth_control)
  meth_test <- data.frame(meth_test)
  unmeth_control <- data.frame(unmeth_control)
  unmeth_test <- data.frame(unmeth_test)

   if (dim(meth_control)[2]<=20){
    num <- dim(meth_control)[2]
  } else {num <- 20}
  
  res <- list()
  for (i in 1:dim(meth_test)[2]){
    counts <- cbind(meth_control[,1:num],meth_test[,i],unmeth_control[,1:num],unmeth_test[,i])
    design = data.frame(
                    Trt = rep(c(rep("control", num),"test"),2), 
                    Meth = c(rep("meth",(num+1)),rep("unmeth", 
                            (num+1))))
    model = ~  Meth + Trt + Meth*Trt
    # 
    dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = design,
                                design = model)
    sf = colSums(cbind((meth_control[,1:num]+unmeth_control[,1:num]),
                       (meth_test[,i]+unmeth_test[,i])))
    sf = sf/median(sf)
    
    sizeFactors(dds) = rep(sf,2)
    

    
    dds <- DESeq(dds, test = "Wald")
    res[[i]] = DESeq2::results(dds, name = "Methunmeth.Trttest")
  }
  
  
  
  
 
  
  return(res)
}

```

### edgeR
```{r,warning=FALSE}


edgeR_test <- function(meth_control,meth_test,unmeth_control,unmeth_test){
  
   if (dim(meth_control)[2]<=20){
    num <- dim(meth_control)[2]
  } else {num <- 20}
  
  res <- list()
  
  for (i in 1: dim(meth_test)[2]){
       # create deglist
    countdata <- cbind(meth_control[,1:num],meth_test[,i],unmeth_control[,1:num],unmeth_test[,i])
   
    dgelist <- DGEList(counts = countdata)
    
    
    countdata_all <- cbind((meth_control[,1:num]+unmeth_control[,1:num]),(meth_test[,i]+unmeth_test[,i]))
   
    dgelist_all <- DGEList(counts = countdata_all)
    
    
    # normalize
    dgelist_norm_all <- calcNormFactors(dgelist_all, method = 'TMM')
    
    dgelist_norm <- calcNormFactors(dgelist, method = 'TMM')
    
    dgelist_norm@.Data[[2]] <- rbind(dgelist_norm_all@.Data[[2]],dgelist_norm_all@.Data[[2]])
    
    rownames(dgelist_norm@.Data[[2]]) <- paste0("Sample",1:(2*(num+1)))
    
    
    group1 <- factor(rep(c(rep("control",num),"test"),2))
    group2 <- factor(c(rep("meth",(num+1)),rep("unmeth",(num+1))))
    
    
    
    design <- model.matrix(~0+group1+group2+group1:group2)
    colnames(design) <- c("control", "test", "unmeth", "unmeth_and_test")
    
    dge <- estimateDisp(dgelist_norm, design, robust = TRUE)
    
    fit <- glmFit(dge, design, robust = TRUE)
    
    #contrast <- makeContrasts(test - control, levels = design)
    #lrt <- glmLRT(fit,contrast=contrast)@.Data
    lrt <- glmLRT(fit,coef=4)@.Data
    
    res[[i]] <- lrt[[14]]
  }
  
  return(res)
 
}



```

### FDR function

FDR_function(): This function calculates the FDR (false discovery rate).
```{r,include=FALSE,warning=FALSE}
FDR_function <- function(per_me,n_Sites,res,adjust=TRUE,threshold){
  if(adjust==TRUE){
     res <- p.adjust(res,method = "BH",n=length(res))
  }
 
  len_1 <- length(which((res[((round(n_Sites*per_me)+1):n_Sites)]<=threshold)))
  len <- length(which(res<=threshold))
  res <- len_1/(len)
  return(res)
}
```

### TPR function

TPR_function(): This function calculates the TPR (true positive rate).
```{r,include=FALSE,warning=FALSE}
TPR_function <- function(per_me,n_Sites,res,adjust=TRUE,threshold){
  if(adjust==TRUE){
     res <- p.adjust(res,method = "BH",n=length(res))
  }
 
  len_1 <- length(which((res[(1:(round(n_Sites*per_me)))]<=threshold)))
  len <- (round(n_Sites*per_me))
  res <- len_1/(len)
  return(res)
}
```






### test_function()
In the following, test_function() is the function used to simulate data and obtain results from the SigRM, 10x, and SMART models. Since the 10x and SMART methods can only determine whether sites are m6A or not, whereas the SigRM provides p-values, we establish a threshold to ensure that the number of m6A sites detected by SigRM matches that of 10x or SMART for a fair comparison. We then evaluate and compare the performance based on AUC, FDR, and TPR. The output of test_function() is a list with seven components:

* AUC results for each method.
* FDR results for each method.
* TPR results for each method.
* Simulated data.
* Results obtained from SigRM.
* Results from the 10x method.
* Results from the SMART method."
* Results from the DESeq2 method."



```{r,include=FALSE,warning=FALSE}
test_function <- function(n_Sites = 1000, test_num = 200, control_num = 300, 
     min_expression = 0, max_expression = 3, expression_p = 10,per_me =  per_me,
     prange1=c(0,0.4),prange2=c(0.6,0.9),prange3=c(0,0.4),shape2=10,
      remove.false = TRUE){

options (warn = - 1)
# simulate data
data <- simulateData(n_Sites = n_Sites,test_num = test_num,control_num = control_num,
          min_expression=min_expression,max_expression=max_expression,
          expression_p =expression_p ,per_me=per_me,
          prange1=prange1,prange2=prange2,prange3=prange3,shape2 = shape2)

# test

res1 = SigRMtest(data[[1]],data[[2]],data[[3]],data[[4]],method_dispersion = "locfit")
res2 = single_test_10x(data[[1]],data[[2]],data[[3]],data[[4]])
res3 = single_test_SMART(data[[1]],data[[2]],data[[3]],data[[4]])
res4 = DESeq2_test(data[[1]],data[[2]],data[[3]],data[[4]])
res5 = edgeR_test(data[[1]],data[[2]],data[[3]],data[[4]])

# set parameter

label=c(rep(0,round(n_Sites*(per_me))),rep(1,round(n_Sites*(1-per_me))))


# calculate auc, fdr, TPR
  

# 10x
list_create <- as.list( 1 : length(res2) )
list_function <- function (x) {
  l1 <- length(res2[[x]])
  
  threshold1 <- res1[[5]][,x][order(res1[[5]][,x])][l1]
  p <-rep(1,n_Sites)
  p[res2[[x]]]<-0
  auc_single_10x <- auc(roc(label,p))
  fdr_single_10x <-FDR_function(per_me,n_Sites,p,threshold=threshold1)
  TPR_single_10x <-TPR_function(per_me,n_Sites,p,threshold=threshold1)
   return(cbind(auc_single_10x,fdr_single_10x,TPR_single_10x))
}
test_results_10x  <- lapply(list_create,list_function)
auc_single_10x <- unlist(lapply(test_results_10x,function(df)df[,1]))
fdr_single_10x <- unlist(lapply(test_results_10x,function(df)df[,2]))
TPR_single_10x <- unlist(lapply(test_results_10x,function(df)df[,3]))

# SMART
list_create <- as.list( 1 : length(res2) )
list_function <- function (x) {
   
  l2 <- length(res3[[x]])
  
  threshold2 <-  res1[[5]][,x][order(res1[[5]][,x])][l2]
  p <-rep(1,n_Sites)
  p[res3[[x]]]<-0
  auc_single_SMART <- auc(roc(label,p))
  fdr_single_SMART <-FDR_function(per_me,n_Sites,p,threshold=threshold2)
  TPR_single_SMART<-TPR_function(per_me,n_Sites,p,threshold=threshold2)
  return(cbind(auc_single_SMART,fdr_single_SMART,TPR_single_SMART))
}
test_results_SMART  <- lapply(list_create,list_function)
auc_single_SMART <- unlist(lapply(test_results_SMART,function(df)df[,1]))
fdr_single_SMART <- unlist(lapply(test_results_SMART,function(df)df[,2]))
TPR_single_SMART <- unlist(lapply(test_results_SMART,function(df)df[,3]))


# SigRM
list_create <- as.list( 1 : length(res2) )
list_function <- function (x) {
  l1 <- length(res2[[x]])
  l2 <- length(res3[[x]])
  
  threshold1 <- res1[[5]][,x][order(res1[[5]][,x])][l1]
  threshold2 <-  res1[[5]][,x][order(res1[[5]][,x])][l2]
  

  p2 <-rep(1,length(res1[[5]][,x])) 
  p2[which(res1[[5]][,x] <=threshold1) ] <-0 
  auc_SigRM_10x<- auc(roc(label,p2))
  fdr_SigRM_10x<-FDR_function(per_me,n_Sites,p2,threshold=threshold1)
  TPR_SigRM_10x<-TPR_function(per_me,n_Sites,p2,threshold=threshold1)
  
  p2 <-rep(1,length(res1[[5]][,x])) 
  p2[which(res1[[5]][,x] <=threshold2) ] <-0 
  auc_SigRM_SMART <- auc(roc(label,p2))
  fdr_SigRM_SMART<-FDR_function(per_me,n_Sites,p2,threshold=threshold2)
  TPR_SigRM_SMART<-TPR_function(per_me,n_Sites,p2,threshold=threshold2)
  
  return(cbind(auc_SigRM_10x,fdr_SigRM_10x,TPR_SigRM_10x,
               auc_SigRM_SMART,fdr_SigRM_SMART,TPR_SigRM_SMART))
}
test_results_SigRM  <- lapply(list_create,list_function)
auc_SigRM_10x <- unlist(lapply(test_results_SigRM,function(df)df[,1]))
fdr_SigRM_10x <- unlist(lapply(test_results_SigRM,function(df)df[,2]))
TPR_SigRM_10x <- unlist(lapply(test_results_SigRM,function(df)df[,3]))
auc_SigRM_SMART <- unlist(lapply(test_results_SigRM,function(df)df[,4]))
fdr_SigRM_SMART <- unlist(lapply(test_results_SigRM,function(df)df[,5]))
TPR_SigRM_SMART <- unlist(lapply(test_results_SigRM,function(df)df[,6]))





# DESeq2
list_create <- as.list( 1 : length(res2) )
list_function <- function (x) {
  l1 <- length(res2[[x]])
  l2 <- length(res3[[x]])
  
  threshold1 <- res4[[x]][,5][order(res4[[x]][,5])][l1]
  threshold2 <- res4[[x]][,5][order(res4[[x]][,5])][l2]
  

  p2 <-rep(1,length(res4[[x]][,5])) 
  p2[which(res4[[x]][,5]<=threshold1) ] <-0 
  auc_DESeq2_10x<- auc(roc(label,p2))
  fdr_DESeq2_10x<-FDR_function(per_me,n_Sites,p2,threshold=threshold1)
  TPR_DESeq2_10x<-TPR_function(per_me,n_Sites,p2,threshold=threshold1)
  
  p2 <-rep(1,length(res4[[x]][,5])) 
  p2[which(res4[[x]][,5]<=threshold2) ] <-0 
  auc_DESeq2_SMART<- auc(roc(label,p2))
  fdr_DESeq2_SMART<-FDR_function(per_me,n_Sites,p2,threshold=threshold2)
  TPR_DESeq2_SMART<-TPR_function(per_me,n_Sites,p2,threshold=threshold2)
  
  return(cbind(auc_DESeq2_10x,fdr_DESeq2_10x,TPR_DESeq2_10x,
               auc_DESeq2_SMART,fdr_DESeq2_SMART,TPR_DESeq2_SMART))
}
test_results_DESeq2  <- lapply(list_create,list_function)
auc_DESeq2_10x <- unlist(lapply(test_results_DESeq2,function(df)df[,1]))
fdr_DESeq2_10x <- unlist(lapply(test_results_DESeq2,function(df)df[,2]))
TPR_DESeq2_10x <- unlist(lapply(test_results_DESeq2,function(df)df[,3]))
auc_DESeq2_SMART <- unlist(lapply(test_results_DESeq2,function(df)df[,4]))
fdr_DESeq2_SMART <- unlist(lapply(test_results_DESeq2,function(df)df[,5]))
TPR_DESeq2_SMART <- unlist(lapply(test_results_DESeq2,function(df)df[,6]))


# edgeR
list_create <- as.list( 1 : length(res2) )
list_function <- function (x) {
  l1 <- length(res2[[x]])
  l2 <- length(res3[[x]])
  
  threshold1 <- res5[[x]][,4][order(res5[[x]][,4])][l1]
  threshold2 <- res5[[x]][,4][order(res5[[x]][,4])][l2]
  

  p2 <-rep(1,length(res5[[x]][,4])) 
  p2[which(res5[[x]][,4]<=threshold1) ] <-0 
  auc_edgeR_10x<- auc(roc(label,p2))
  fdr_edgeR_10x<-FDR_function(per_me,n_Sites,p2,threshold=threshold1)
  TPR_edgeR_10x<-TPR_function(per_me,n_Sites,p2,threshold=threshold1)
  
  p2 <-rep(1,length(res5[[x]][,4])) 
  p2[which(res5[[x]][,4]<=threshold2) ] <-0 
  auc_edgeR_SMART<- auc(roc(label,p2))
  fdr_edgeR_SMART<-FDR_function(per_me,n_Sites,p2,threshold=threshold2)
  TPR_edgeR_SMART<-TPR_function(per_me,n_Sites,p2,threshold=threshold2)
  
  return(cbind(auc_edgeR_10x,fdr_edgeR_10x,TPR_edgeR_10x,
               auc_edgeR_SMART,fdr_edgeR_SMART,TPR_edgeR_SMART))
}
test_results_edgeR  <- lapply(list_create,list_function)
auc_edgeR_10x <- unlist(lapply(test_results_edgeR,function(df)df[,1]))
fdr_edgeR_10x <- unlist(lapply(test_results_edgeR,function(df)df[,2]))
TPR_edgeR_10x <- unlist(lapply(test_results_edgeR,function(df)df[,3]))
auc_edgeR_SMART <- unlist(lapply(test_results_edgeR,function(df)df[,4]))
fdr_edgeR_SMART <- unlist(lapply(test_results_edgeR,function(df)df[,5]))
TPR_edgeR_SMART <- unlist(lapply(test_results_edgeR,function(df)df[,6]))


# set table
auc_table <- cbind(c(auc_SigRM_10x,auc_SigRM_SMART,auc_single_10x,auc_single_SMART,
                     auc_DESeq2_10x,auc_DESeq2_SMART,auc_edgeR_10x,auc_edgeR_SMART),
                   c(rep("SigRM_10x",length(res2)),rep("SigRM_SMART",length(res2)), rep("single_10x",length(res2)), rep("single_SMART",length(res2)),
                     rep("DESeq2_10x",length(res2)),rep("DESeq2_SMART",length(res2)),
                     rep("edgeR_10x",length(res2)),rep("edgeR_SMART",length(res2))))

fdr_table <- cbind(c(fdr_SigRM_10x,fdr_SigRM_SMART,fdr_single_10x,fdr_single_SMART,
                     fdr_DESeq2_10x,fdr_DESeq2_SMART,fdr_edgeR_10x,fdr_edgeR_SMART),
                   c(rep("SigRM_10x",length(res2)),rep("SigRM_SMART",length(res2)), rep("single_10x",length(res2)), rep("single_SMART",length(res2)),
                     rep("DESeq2_10x",length(res2)),rep("DESeq2_SMART",length(res2)),
                     rep("edgeR_10x",length(res2)),rep("edgeR_SMART",length(res2))))

TPR_table <- cbind(c(TPR_SigRM_10x,TPR_SigRM_SMART,TPR_single_10x,TPR_single_SMART,
                     TPR_DESeq2_10x,TPR_DESeq2_SMART,TPR_edgeR_10x,TPR_edgeR_SMART),
                   c(rep("SigRM_10x",length(res2)),rep("SigRM_SMART",length(res2)), rep("single_10x",length(res2)), rep("single_SMART",length(res2)),
                     rep("DESeq2_10x",length(res2)),rep("DESeq2_SMART",length(res2)),
                     rep("edgeR_10x",length(res2)),rep("edgeR_SMART",length(res2))))

# return list
list_return <- list()
  list_return[[1]]=auc_table
  list_return[[2]]=fdr_table
  list_return[[3]]=TPR_table
  list_return[[4]]=data
  list_return[[5]]=res1
  list_return[[6]]=res2
  list_return[[7]]=res3
  list_return[[8]]=res4
  list_return[[9]]=res5

return(list_return)

}

```







## Compare origion methods

This section involves comparing the performance of the SigRM model with the original method by invoking the test_function().

To reduce the running time, I change to 100 sites and 10 test and control cells. If you change to 5000 sites, and 100 test and control cells. You well get the exact plot in the paper.
```{r,include=FALSE,warning=FALSE}


set.seed(1)

n_Sites <- 100
per_me  <-0.5
test_num = 10
control_num = 10
shape2= 1
expression_p = 10
min_expression = 0
max_expression = 3

metric_1 <- test_function(n_Sites = n_Sites, test_num = test_num, control_num =
      control_num , min_expression = min_expression, max_expression =max_expression,per_me =  per_me ,prange1=c(0,0.4),prange2=c(0.6,0.9),prange3=c(0,0.4),shape2=shape2,expression_p =expression_p,remove.false = TRUE)

AUC_metric_1 <- metric_1[[1]]
colnames(AUC_metric_1) <- c("AUC","Method")
AUC_metric_1 <- as.data.frame(AUC_metric_1)
AUC_metric_1 $AUC <-as.numeric(AUC_metric_1 $AUC)
AUC_metric_1 $shape2 <-rep(shape2,dim(AUC_metric_1 )[1])

FDR_metric_1 <- metric_1[[2]]
colnames(FDR_metric_1) <- c("FDR","Method")
FDR_metric_1 <- as.data.frame(FDR_metric_1)
FDR_metric_1 $FDR <-as.numeric(FDR_metric_1 $FDR)
FDR_metric_1 $shape2 <-rep(shape2,dim(FDR_metric_1 )[1])

TPR_metric_1 <- metric_1[[3]]
colnames(TPR_metric_1) <- c("TPR","Method")
TPR_metric_1 <- as.data.frame(TPR_metric_1)
TPR_metric_1 $TPR <-as.numeric(TPR_metric_1$TPR)
TPR_metric_1 $shape2 <-rep(shape2,dim(TPR_metric_1 )[1])

set.seed(1)

shape2= 2

metric_2 <- test_function(n_Sites = n_Sites, test_num = test_num, control_num =
      control_num , min_expression = min_expression, max_expression =max_expression,per_me =  per_me ,prange1=c(0,0.4),prange2=c(0.6,0.9),prange3=c(0,0.4),shape2=shape2,expression_p =expression_p,remove.false = TRUE)

AUC_metric_2 <- metric_2[[1]]
colnames(AUC_metric_2) <- c("AUC","Method")
AUC_metric_2 <- as.data.frame(AUC_metric_2)
AUC_metric_2 $AUC <-as.numeric(AUC_metric_2 $AUC)
AUC_metric_2 $shape2 <-rep(shape2,dim(AUC_metric_2 )[1])

FDR_metric_2 <- metric_2[[2]]
colnames(FDR_metric_2) <- c("FDR","Method")
FDR_metric_2 <- as.data.frame(FDR_metric_2)
FDR_metric_2 $FDR <-as.numeric(FDR_metric_2 $FDR)
FDR_metric_2 $shape2 <-rep(shape2,dim(FDR_metric_2 )[1])

TPR_metric_2 <- metric_2[[3]]
colnames(TPR_metric_2) <- c("TPR","Method")
TPR_metric_2 <- as.data.frame(TPR_metric_2)
TPR_metric_2 $TPR <-as.numeric(TPR_metric_2$TPR)
TPR_metric_2 $shape2 <-rep(shape2,dim(TPR_metric_2 )[1])

set.seed(1)

shape2= 3

metric_3 <- test_function(n_Sites = n_Sites, test_num = test_num, control_num =
      control_num , min_expression = min_expression, max_expression =max_expression,per_me =  per_me ,prange1=c(0,0.4),prange2=c(0.6,0.9),prange3=c(0,0.4),shape2=shape2,expression_p =expression_p,remove.false = TRUE)

AUC_metric_3 <- metric_3[[1]]
colnames(AUC_metric_3) <- c("AUC","Method")
AUC_metric_3 <- as.data.frame(AUC_metric_3)
AUC_metric_3 $AUC <-as.numeric(AUC_metric_3 $AUC)
AUC_metric_3 $shape2 <-rep(shape2,dim(AUC_metric_3 )[1])

FDR_metric_3 <- metric_3[[2]]
colnames(FDR_metric_3) <- c("FDR","Method")
FDR_metric_3 <- as.data.frame(FDR_metric_3)
FDR_metric_3 $FDR <-as.numeric(FDR_metric_3 $FDR)
FDR_metric_3 $shape2 <-rep(shape2,dim(FDR_metric_3 )[1])

TPR_metric_3 <- metric_3[[3]]
colnames(TPR_metric_3) <- c("TPR","Method")
TPR_metric_3 <- as.data.frame(TPR_metric_3)
TPR_metric_3 $TPR <-as.numeric(TPR_metric_3$TPR)
TPR_metric_3 $shape2 <-rep(shape2,dim(TPR_metric_3 )[1])

set.seed(1)

shape2= 4

metric_4 <- test_function(n_Sites = n_Sites, test_num = test_num, control_num =
      control_num , min_expression = min_expression, max_expression =max_expression,per_me =  per_me ,prange1=c(0,0.4),prange2=c(0.6,0.9),prange3=c(0,0.4),shape2=shape2,expression_p =expression_p,remove.false = TRUE)

AUC_metric_4 <- metric_4[[1]]
colnames(AUC_metric_4) <- c("AUC","Method")
AUC_metric_4 <- as.data.frame(AUC_metric_4)
AUC_metric_4 $AUC <-as.numeric(AUC_metric_4 $AUC)
AUC_metric_4 $shape2 <-rep(shape2,dim(AUC_metric_4 )[1])

FDR_metric_4 <- metric_4[[2]]
colnames(FDR_metric_4) <- c("FDR","Method")
FDR_metric_4 <- as.data.frame(FDR_metric_4)
FDR_metric_4 $FDR <-as.numeric(FDR_metric_4 $FDR)
FDR_metric_4 $shape2 <-rep(shape2,dim(FDR_metric_4 )[1])

TPR_metric_4 <- metric_4[[3]]
colnames(TPR_metric_4) <- c("TPR","Method")
TPR_metric_4 <- as.data.frame(TPR_metric_4)
TPR_metric_4 $TPR <-as.numeric(TPR_metric_4$TPR)
TPR_metric_4 $shape2 <-rep(shape2,dim(TPR_metric_4 )[1])


set.seed(1)

shape2= 5

metric_5 <- test_function(n_Sites = n_Sites, test_num = test_num, control_num =
      control_num , min_expression = min_expression, max_expression =max_expression,per_me =  per_me ,prange1=c(0,0.4),prange2=c(0.6,0.9),prange3=c(0,0.4),shape2=shape2,expression_p =expression_p,remove.false = TRUE)

AUC_metric_5 <- metric_5[[1]]
colnames(AUC_metric_5) <- c("AUC","Method")
AUC_metric_5 <- as.data.frame(AUC_metric_5)
AUC_metric_5 $AUC <-as.numeric(AUC_metric_5 $AUC)
AUC_metric_5 $shape2 <-rep(shape2,dim(AUC_metric_5 )[1])

FDR_metric_5 <- metric_5[[2]]
colnames(FDR_metric_5) <- c("FDR","Method")
FDR_metric_5 <- as.data.frame(FDR_metric_5)
FDR_metric_5 $FDR <-as.numeric(FDR_metric_5 $FDR)
FDR_metric_5 $shape2 <-rep(shape2,dim(FDR_metric_5 )[1])

TPR_metric_5 <- metric_5[[3]]
colnames(TPR_metric_5) <- c("TPR","Method")
TPR_metric_5 <- as.data.frame(TPR_metric_5)
TPR_metric_5 $TPR <-as.numeric(TPR_metric_5$TPR)
TPR_metric_5 $shape2 <-rep(shape2,dim(TPR_metric_5 )[1])







set.seed(1)


shape2= 6

metric_6 <- test_function(n_Sites = n_Sites, test_num = test_num, control_num =
      control_num , min_expression = min_expression, max_expression =max_expression,per_me =  per_me ,prange1=c(0,0.4),prange2=c(0.6,0.9),prange3=c(0,0.4),shape2=shape2,expression_p =expression_p,remove.false = TRUE)

AUC_metric_6 <- metric_6[[1]]
colnames(AUC_metric_6) <- c("AUC","Method")
AUC_metric_6 <- as.data.frame(AUC_metric_6)
AUC_metric_6 $AUC <-as.numeric(AUC_metric_6 $AUC)
AUC_metric_6 $shape2 <-rep(shape2,dim(AUC_metric_6 )[1])

FDR_metric_6 <- metric_6[[2]]
colnames(FDR_metric_6) <- c("FDR","Method")
FDR_metric_6 <- as.data.frame(FDR_metric_6)
FDR_metric_6 $FDR <-as.numeric(FDR_metric_6 $FDR)
FDR_metric_6 $shape2 <-rep(shape2,dim(FDR_metric_6 )[1])

TPR_metric_6 <- metric_6[[3]]
colnames(TPR_metric_6) <- c("TPR","Method")
TPR_metric_6 <- as.data.frame(TPR_metric_6)
TPR_metric_6 $TPR <-as.numeric(TPR_metric_6$TPR)
TPR_metric_6 $shape2 <-rep(shape2,dim(TPR_metric_6 )[1])


AUC_metric <- rbind(AUC_metric_1,AUC_metric_2,AUC_metric_3,AUC_metric_4,AUC_metric_5,AUC_metric_6)

AUC_metric$shape2 <- as.factor(AUC_metric$shape2)


FDR_metric <- rbind(FDR_metric_1,FDR_metric_2,FDR_metric_3,FDR_metric_4,FDR_metric_5,FDR_metric_6)

FDR_metric$shape2 <- as.factor(FDR_metric$shape2)

TPR_metric <- rbind(TPR_metric_1,TPR_metric_2,TPR_metric_3,TPR_metric_4,TPR_metric_5,TPR_metric_6)

TPR_metric$shape2 <- as.factor(TPR_metric$shape2)
```


# Cluster 
We aim to apply a clustering method in which cells with similar gene expression are grouped together. For the test cells, this method not only employs all control cells but also selects control cells with similar gene expression as the control group. We utilize the Seurat clustering method.


### SigRM_cluster_test function
Before comparing the performance of SigRM_cluster_test() and SigRMtest(), let's explore how to use the SigRM_cluster_test() function. The primary input for the SigRM_cluster_test function is the same as for the SigRMtest function: a four read counts matrix or data frame where rows represent genes and columns represent single cells. Additionally, it requires a factor variable named 'cluster variable,' indicating that you must cluster the cells before calling the SigRM_cluster_test function. For details on other parameters, please refer to the help page.

The outputs of the SigRM_cluster_test function also consist of a list with a length of 7. The key differences between the outputs of the SigRM_cluster_test function and the SigRMtest function are found in elements 2 and 6 of the list:

* The second part contains the mean methylation proportion data frame of the control group.
* The sixth part includes the estimated abundance data frame.


It's important to note that the mean methylation proportion of the control group and the estimated abundance can vary with different clusters, which is why they are represented as data frames rather than vectors.

### cluster by Seurat
```{r, include=FALSE,warning=FALSE}
set.seed(1)
data <- simulateData(n_Sites=2000,test_num = 10,control_num = 100)

# cluster the cells
data1 <- data[[1]]+data[[3]]
data2 <- data[[2]]+data[[4]]

l <-dim(data1)[2]


data_ <- cbind(data1,data2)
colnames(data_)<- paste0("cell",c(1:dim(data_)[2]))
rownames(data_)<- paste0("gene",c(1:dim(data_)[1]))


##  Seurat
pbmc <- CreateSeuratObject(counts = log2(data_+1), project = "scDART")


all.genes <- rownames(pbmc)
pbmc <- NormalizeData(pbmc, features = all.genes)
pbmc <- ScaleData(pbmc, features = all.genes)

pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)



pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))


pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)



JackStrawPlot(pbmc, dims = 1:20)

ElbowPlot(pbmc)


pbmc <- FindNeighbors(pbmc, dims = 1)
pbmc <- FindClusters(pbmc, resolution = 0.5)

cluster <- Idents(pbmc)




```



### call the SigRM_cluster_test() function
```{r,warning=FALSE}
res <- SigRM_cluster_test(data[[1]],data[[2]],data[[3]],data[[4]],cluster = cluster)

# meth proportion data frame (row is gene and column is single cell) in the test cells. 
head(res[[1]])

# mean meth proportion vector (row is gene and column is single cell) in the control cells.
head(res[[2]])

# log2 risk ratio data frame for the test cells. 
head(res[[3]])

# log2 odds ratio data frame for the test cells. 
head(res[[4]])

# p value data frame for the test cells. 
head(res[[5]])

# estimated gene abundance
head(res[[6]])

# adjusted p value data frame for the test cells
head(res[[7]])
```






# similarity
The cluster method may result in clusters that do not contain any control cells, leaving the test cells without a control group. In such cases, we can employ control cells with the most similar gene expression.

### SigRM_similarity_test function

Let's take a look at how to use the SigRM_similarity_test() function. The primary input for the SigRM_similarity_test() function is the same as that for the SigRMtest function: a four read counts matrix or data frame where rows represent genes and columns represent single cells. The 'methods' parameter is set to the default "Pearson," but it can also be customized from the vector c("Euclidean distance," "Manhattan distance," "Maximum distance," "Pearson," "Spearman").The 'num_cluster' parameter determines how many similar cells are selected from the control cells to form the control group.

Regarding the outputs, they are identical to those of the SigRM_cluster_test function.

These revisions should make the text clearer and more grammatically sound.

```{r,warning=FALSE}

set.seed(1)
data <- simulateData(n_Sites=2000,test_num = 10,control_num = 100)
res = SigRM_similarity_test(data[[1]],data[[2]],data[[3]],data[[4]],
            method="Pearson",num_control=10)


# meth proportion data frame (row is gene and column is single cell) in the test cells. 
head(res[[1]])

# mean meth proportion vector (row is gene and column is single cell) in the control cells.
head(res[[2]])

# log2 risk ratio data frame for the test cells. 
head(res[[3]])

# log2 odds ratio data frame for the test cells. 
head(res[[4]])

# p value data frame for the test cells. 
head(res[[5]])

# estimated gene abundance
head(res[[6]])

# adjusted p value data frame for the test cells
head(res[[7]])

```



